// @microsoft/power-apps-actions/dist/ReferenceStores/connectionReferences.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { equalsIgnoreCase } from '@microsoft/power-apps-common/utils';
import { LOGIC_FLOWS_API_ID } from '../Actions/flowConstants.js';
import { getSanitizedUniqueName } from '../CodeGen/shared/nameUtils.js';
// UUID generation utility
function generateUUID() {
    const seed = new Date().getTime();
    const uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (replacer) => {
        // eslint-disable-next-line no-bitwise
        const rnd = ((seed + Math.random() * 16) % 16) | 0;
        // eslint-disable-next-line no-bitwise
        return (replacer === 'x' ? rnd : (rnd & 0x3) | 0x8).toString(16);
    });
    return uuid;
}
/**
 * Upserts a connection reference in the provided collection.
 * @param connectionReferences - The collection of connection references.
 * @param params - Parameters for the connection reference.
 * @param params.apiId - The API ID of the connection.
 * @param params.displayName - The display name of the connection.
 * @param params.dataSourceName - The name of the data source.
 * @param params.authenticationType - The authentication type of the connection (optional).
 * @param params.sharedConnectionId - The shared connection ID (optional).
 * @param params.dataverseDataSourceNames - The names of the Dataverse data sources (optional).
 * @returns The disambiguated data source name.
 */
export function upsertConnectionReference(connectionReferences, params) {
    const { apiId, apiName, connectorDisplayName, tabularMetadata, authenticationType, sharedConnectionId, dataverseDataSourceNames, connectionReferenceLogicalName, dataSourceShareLink, siteId, } = params;
    // dataSourceName should be unique (aside from Stored Procedures, since they all live within the same entity)
    // We can have the upsertConnectionReference function ensure uniqueness and return the disambiguated name that we can later
    // use to save the metadata.
    const referenceId = tryMatchReferenceId(connectionReferences, apiId, authenticationType, sharedConnectionId) ||
        generateUUID();
    const desiredName = tabularMetadata?.friendlyName ?? apiName;
    const dataSourceName = getSanitizedUniqueName([
        ...(dataverseDataSourceNames ?? []),
        ...getExistingCdpNames(connectionReferences, {
            apiId,
            datasetName: tabularMetadata?.dataset,
        }),
    ], desiredName);
    if (connectionReferences[referenceId]) {
        // If a matching reference ID is found, use it
        connectionReferences[referenceId].dataSources ??= [];
        if (!connectionReferences[referenceId].dataSources.includes(dataSourceName)) {
            connectionReferences[referenceId].dataSources.push(dataSourceName);
            if (tabularMetadata) {
                connectionReferences[referenceId].dataSets ??= {};
                connectionReferences[referenceId].dataSets[tabularMetadata.dataset] ??= {
                    dataSources: {},
                };
                connectionReferences[referenceId].dataSets[tabularMetadata.dataset].dataSources[dataSourceName] = tabularMetadata.isProcedure
                    ? {}
                    : {
                        tableName: tabularMetadata.resourceName,
                    };
                // Set siteId if provided
                if (siteId) {
                    connectionReferences[referenceId].dataSets[tabularMetadata.dataset].siteId = siteId;
                }
            }
        }
        // Set dataSourceShareLink if provided
        if (dataSourceShareLink !== undefined) {
            connectionReferences[referenceId].dataSourcesShareLinks ??= {};
            connectionReferences[referenceId].dataSourcesShareLinks[dataSourceName] =
                dataSourceShareLink;
        }
    }
    else {
        // create a new reference ID
        const datasourceObject = tabularMetadata?.isProcedure ? {} : { tableName: tabularMetadata?.resourceName };
        if (tabularMetadata?.resourceNameOverride) {
            datasourceObject.tableNameOverride = {
                environmentVariableName: tabularMetadata.resourceNameOverride,
            };
        }
        const dataSetsObject = tabularMetadata
            ? {
                [tabularMetadata.dataset]: {
                    dataSources: {
                        [dataSourceName]: datasourceObject,
                    },
                    ...(siteId && { siteId }),
                },
            }
            : {};
        if (tabularMetadata?.datasetOverride) {
            dataSetsObject[tabularMetadata.dataset].datasetOverride = {
                environmentVariableName: tabularMetadata.datasetOverride,
            };
        }
        connectionReferences[referenceId] = {
            id: apiId,
            displayName: connectorDisplayName,
            dataSources: [dataSourceName],
            authenticationType: authenticationType || undefined,
            sharedConnectionId: sharedConnectionId || undefined,
            dataSets: dataSetsObject,
            xrmConnectionReferenceLogicalName: connectionReferenceLogicalName,
            dataSourcesShareLinks: dataSourceShareLink !== undefined ? { [dataSourceName]: dataSourceShareLink } : undefined,
        };
    }
    return tabularMetadata?.isProcedure
        ? { dataSourceName: tabularMetadata?.friendlyName ?? '', spEntity: dataSourceName }
        : { dataSourceName };
}
/**
 * Matches an existing reference ID by API ID, authentication type, shared connection ID, and xrmConnectionReferenceLogicalName.
 */
function tryMatchReferenceId(connectionReferences, apiId, authenticationType, sharedConnectionId, connectionReferenceLogicalName) {
    for (const [id, reference] of Object.entries(connectionReferences)) {
        // The logical name is only a tiebreaker when both a sharedConnectionId and a logical name are provided.
        const logicalNameMatches = sharedConnectionId?.trim() && connectionReferenceLogicalName?.trim()
            ? equalsIgnoreCase(reference.xrmConnectionReferenceLogicalName, connectionReferenceLogicalName)
            : true;
        if (equalsIgnoreCase(reference.id, apiId) &&
            equalsIgnoreCase(reference.authenticationType, authenticationType) &&
            equalsIgnoreCase(reference.sharedConnectionId, sharedConnectionId) &&
            logicalNameMatches) {
            return id;
        }
    }
    return undefined;
}
export function getExistingCdpNames(connectionReferences, exclusions) {
    if (!exclusions) {
        return Object.values(connectionReferences).flatMap((ref) => ref.dataSources ?? []);
    }
    return Object.values(connectionReferences).flatMap((ref) => {
        if (ref.id !== exclusions.apiId) {
            return ref.dataSources ?? [];
        }
        else if (exclusions.datasetName) {
            return Object.entries(ref.dataSets ?? {}).flatMap(([dsKey, dsValue]) => {
                if (dsKey !== exclusions.datasetName) {
                    return Object.keys(dsValue.dataSources);
                }
                return [];
            });
        }
        return [];
    });
}
/**
 * Upserts connection references for a flow and its connector dependencies.
 *
 * Creates or updates:
 * - A `shared_logicflows` connection reference for the flow itself (with `workflowDetails`).
 * - A connection reference for each connector the flow depends on, including
 *   `authenticationType`, `sharedConnectionId`, and `xrmConnectionReferenceLogicalName`.
 *
 * Existing entries are matched by `workflowEntityId` (for the flow) or by connector `id`
 * (for dependencies), so re-adding the same flow is idempotent.
 */
export function upsertFlowConnectionReferences(connectionReferences, params) {
    const { flowDisplayName, workflowEntityId, flowDependencies, workflowName } = params;
    // Step 1: Find or create UUIDs for each dependency connector.
    const dependenciesMap = {};
    for (const [flowRefName, flowConnectionRef] of Object.entries(flowDependencies)) {
        const existingId = tryMatchReferenceId(connectionReferences, flowConnectionRef.id.toLowerCase(), flowConnectionRef.authenticationType, flowConnectionRef.sharedConnectionId, flowConnectionRef.connectionReferenceLogicalName);
        const depUuid = existingId ?? generateUUID();
        if (!existingId) {
            connectionReferences[depUuid] = {
                id: flowConnectionRef.id.toLowerCase(),
                displayName: flowConnectionRef.displayName,
                authenticationType: flowConnectionRef.authenticationType || undefined,
                sharedConnectionId: flowConnectionRef.sharedConnectionId || undefined,
                // Power Apps only stores xrmConnectionReferenceLogicalName for shared connections
                xrmConnectionReferenceLogicalName: flowConnectionRef.sharedConnectionId
                    ? flowConnectionRef.connectionReferenceLogicalName || undefined
                    : undefined,
            };
        }
        dependenciesMap[flowRefName] = depUuid;
    }
    // Step 2: Find or create the flow's own connection reference.
    const existingFlowId = findFlowConnectionReferenceId(connectionReferences, workflowEntityId);
    const flowUuid = existingFlowId ?? generateUUID();
    // Compute a unique, sanitized data source name — same approach as upsertConnectionReference.
    // Exclude the current flow's own entry when updating so it doesn't conflict with itself.
    const existingDataSourceNames = Object.entries(connectionReferences)
        .filter(([uuid]) => uuid !== existingFlowId)
        .flatMap(([, ref]) => ref.dataSources ?? []);
    const dataSourceName = getSanitizedUniqueName(existingDataSourceNames, flowDisplayName);
    connectionReferences[flowUuid] = {
        id: LOGIC_FLOWS_API_ID,
        displayName: 'Logic flows',
        dataSources: [dataSourceName],
        workflowDetails: {
            workflowEntityId,
            workflowDisplayName: flowDisplayName,
            workflowName,
            ...(Object.keys(dependenciesMap).length > 0 && { dependencies: dependenciesMap }),
        },
    };
}
/**
 * Finds the UUID of a flow's connection reference by its Dataverse workflow entity ID.
 */
function findFlowConnectionReferenceId(connectionReferences, workflowEntityId) {
    for (const [uuid, ref] of Object.entries(connectionReferences)) {
        if (ref.workflowDetails?.workflowEntityId === workflowEntityId) {
            return uuid;
        }
    }
    return undefined;
}
