// @microsoft/power-apps-common/dist/http/HttpClient.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { parseCaeChallenge } from './CaeUtils.js';
import { HttpError } from './HttpTypes.js';
/** Generic HTTP client with auth, service name routing, interceptors, and CAE support. */
export class HttpClient {
    defaultHeaders;
    defaultTimeoutMs;
    fetchFn;
    requestInterceptors = [];
    responseInterceptors = [];
    authProvider;
    logger;
    endpoints = new Map();
    constructor(options = {}) {
        this.defaultHeaders = options.defaultHeaders ?? {};
        this.defaultTimeoutMs = options.defaultTimeoutMs ?? 30_000;
        this.fetchFn = options.fetch ?? globalThis.fetch.bind(globalThis);
        this.authProvider = options.authProvider;
        this.logger = options.logger;
    }
    /** Register a named service endpoint for service-name URL resolution. Returns the registered name. */
    registerEndpoint(name, endpoint) {
        this.endpoints.set(name, endpoint);
        return name;
    }
    /** Add a request interceptor. Returns a function to remove it. */
    addRequestInterceptor(interceptor) {
        this.requestInterceptors.push(interceptor);
        return () => {
            const index = this.requestInterceptors.indexOf(interceptor);
            if (index !== -1) {
                this.requestInterceptors.splice(index, 1);
            }
        };
    }
    /** Add a response interceptor. Returns a function to remove it. */
    addResponseInterceptor(interceptor) {
        this.responseInterceptors.push(interceptor);
        return () => {
            const index = this.responseInterceptors.indexOf(interceptor);
            if (index !== -1) {
                this.responseInterceptors.splice(index, 1);
            }
        };
    }
    /** Send a GET request. */
    async get(url, config) {
        return this.request({ ...config, url, method: 'GET' });
    }
    /** Send a POST request. */
    async post(url, body, config) {
        return this.request({ ...config, url, method: 'POST', body });
    }
    /** Send a PUT request. */
    async put(url, body, config) {
        return this.request({ ...config, url, method: 'PUT', body });
    }
    /** Send a PATCH request. */
    async patch(url, body, config) {
        return this.request({ ...config, url, method: 'PATCH', body });
    }
    /** Send a DELETE request. */
    async delete(url, config) {
        return this.request({ ...config, url, method: 'DELETE' });
    }
    /** Send an HTTP request with the given configuration. */
    async request(config) {
        let processedConfig = { ...config };
        // Apply request interceptors
        for (const interceptor of this.requestInterceptors) {
            processedConfig = await interceptor(processedConfig);
        }
        // Resolve service-name URLs and determine auth resource + endpoint metadata
        const { resolvedUrl, resource, endpointHeaders, extractTelemetryProperties } = this.resolveUrl(processedConfig.url);
        processedConfig = { ...processedConfig, url: resolvedUrl };
        // Resolve endpoint headers (evaluating functions) and merge under per-request headers
        let resolvedEndpointHeaders;
        if (endpointHeaders) {
            resolvedEndpointHeaders = {};
            for (const [name, value] of Object.entries(endpointHeaders)) {
                resolvedEndpointHeaders[name] = typeof value === 'function' ? value() : value;
            }
            processedConfig = {
                ...processedConfig,
                headers: { ...resolvedEndpointHeaders, ...processedConfig.headers },
            };
        }
        // Stamp auth token if provider is configured and no Authorization header is already set
        if (this.authProvider && resource && !processedConfig.headers?.['Authorization']) {
            const token = await this.authProvider.getToken({ resource });
            processedConfig = {
                ...processedConfig,
                headers: { ...processedConfig.headers, Authorization: `Bearer ${token.accessToken}` },
            };
        }
        // Execute with timing and telemetry
        const startTime = this.logger ? performance.now() : 0;
        try {
            const { response, isRetried } = await this.executeWithCaeRetry(processedConfig, resource);
            if (this.logger) {
                this.logRequest(config, processedConfig, response, startTime, {
                    resource,
                    resolvedEndpointHeaders,
                    isRetried,
                    extractTelemetryProperties,
                });
            }
            return response;
        }
        catch (error) {
            if (this.logger) {
                const status = error instanceof HttpError ? error.status : undefined;
                const responseHeaders = error instanceof HttpError ? error.responseHeaders : undefined;
                this.logRequest(config, processedConfig, { status, headers: responseHeaders }, startTime, {
                    resource,
                    resolvedEndpointHeaders,
                    isRetried: false,
                    extractTelemetryProperties,
                });
            }
            throw error;
        }
    }
    /** Execute fetch with CAE retry logic. Returns the response and whether a retry occurred. */
    async executeWithCaeRetry(processedConfig, resource) {
        try {
            const response = await this.executeFetch(processedConfig);
            return { response, isRetried: false };
        }
        catch (error) {
            // CAE retry: on 401 with claims challenge, re-acquire token and retry once
            if (this.authProvider && resource && error instanceof HttpError && error.status === 401) {
                const claims = parseCaeChallenge(error.responseHeaders?.['www-authenticate']);
                if (claims) {
                    const freshToken = await this.authProvider.getToken({
                        resource,
                        claims,
                        bypassCache: true,
                    });
                    const retryConfig = {
                        ...processedConfig,
                        headers: {
                            ...processedConfig.headers,
                            Authorization: `Bearer ${freshToken.accessToken}`,
                        },
                    };
                    const response = await this.executeFetch(retryConfig);
                    return { response, isRetried: true };
                }
            }
            throw error;
        }
    }
    /**
     * Execute the raw fetch, parse the response, throw HttpError on !ok,
     * and apply response interceptors.
     */
    async executeFetch(processedConfig) {
        const method = processedConfig.method ?? 'GET';
        const headers = {
            ...this.defaultHeaders,
            ...processedConfig.headers,
        };
        // Auto-set Content-Type for object bodies
        if (processedConfig.body !== undefined &&
            typeof processedConfig.body === 'object' &&
            !headers['Content-Type']) {
            headers['Content-Type'] = 'application/json';
        }
        const timeoutMs = processedConfig.timeoutMs ?? this.defaultTimeoutMs;
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), timeoutMs);
        // Compose signals if caller provided one
        const signal = processedConfig.signal
            ? this.composeAbortSignals(processedConfig.signal, controller.signal)
            : controller.signal;
        try {
            const response = await this.fetchFn(processedConfig.url, {
                method,
                headers,
                body: processedConfig.body !== undefined
                    ? typeof processedConfig.body === 'string'
                        ? processedConfig.body
                        : JSON.stringify(processedConfig.body)
                    : undefined,
                signal,
            });
            const responseHeaders = {};
            response.headers.forEach((value, key) => {
                responseHeaders[key] = value;
            });
            let data;
            const contentType = response.headers.get('content-type');
            if (contentType?.includes('application/json')) {
                data = (await response.json());
            }
            else {
                data = (await response.text());
            }
            if (!response.ok) {
                throw new HttpError(`HTTP ${response.status}: ${response.statusText}`, processedConfig, response.status, data, responseHeaders);
            }
            let result = {
                status: response.status,
                statusText: response.statusText,
                data,
                headers: responseHeaders,
            };
            // Apply response interceptors
            for (const interceptor of this.responseInterceptors) {
                result = await interceptor(result);
            }
            return result;
        }
        catch (error) {
            if (error instanceof HttpError) {
                throw error;
            }
            if (error instanceof Error && error.name === 'AbortError') {
                throw new HttpError('Request timed out', processedConfig);
            }
            throw new HttpError(error instanceof Error ? error.message : 'Unknown request error', processedConfig);
        }
        finally {
            clearTimeout(timeoutId);
        }
    }
    /**
     * Resolve a URL to its final form and determine the auth resource.
     *
     * 1. Absolute URL — used as-is; resource resolved by matching registered
     *    endpoint baseUrls (or undefined if no match).
     * 2. Service name prefix — first path segment matches a registered endpoint.
     * 3. Unresolved — passed through as-is; no resource.
     */
    resolveUrl(url) {
        // Absolute URL
        if (url.startsWith('http://') || url.startsWith('https://')) {
            // Check if URL starts with any registered endpoint's baseUrl
            for (const endpoint of this.endpoints.values()) {
                if (url.startsWith(endpoint.baseUrl)) {
                    return {
                        resolvedUrl: url,
                        resource: endpoint.resource,
                        endpointHeaders: endpoint.headers,
                        extractTelemetryProperties: endpoint.extractTelemetryProperties,
                    };
                }
            }
            return { resolvedUrl: url };
        }
        // Service name prefix: check if the first path segment matches a registered endpoint
        const slashIndex = url.indexOf('/');
        const firstSegment = slashIndex === -1 ? url : url.substring(0, slashIndex);
        const endpoint = this.endpoints.get(firstSegment);
        if (endpoint) {
            const remainingPath = slashIndex === -1 ? '' : url.substring(slashIndex);
            const base = endpoint.baseUrl.replace(/\/$/, '');
            const resolvedUrl = remainingPath ? `${base}/${remainingPath.replace(/^\//, '')}` : base;
            return {
                resolvedUrl,
                resource: endpoint.resource,
                endpointHeaders: endpoint.headers,
                extractTelemetryProperties: endpoint.extractTelemetryProperties,
            };
        }
        return { resolvedUrl: url };
    }
    composeAbortSignals(userSignal, timeoutSignal) {
        const controller = new AbortController();
        const onAbort = () => controller.abort();
        userSignal.addEventListener('abort', onAbort, { once: true });
        timeoutSignal.addEventListener('abort', onAbort, { once: true });
        return controller.signal;
    }
    /** Assemble and emit the HTTP telemetry event. */
    logRequest(originalConfig, processedConfig, outcome, startTime, context) {
        const duration = performance.now() - startTime;
        const status = outcome.status ?? 0;
        const succeeded = status >= 200 && status < 300;
        const responseHeaders = outcome.headers;
        // Layer 1: Endpoint-extracted properties
        const endpointProps = context.extractTelemetryProperties
            ? context.extractTelemetryProperties(originalConfig)
            : {};
        // Layer 2: Per-request telemetry properties
        const requestProps = originalConfig.telemetryProperties ?? {};
        // Layer 3: Computed properties (wins over layers 1 & 2)
        const computedProps = {
            url: processedConfig.url,
            httpMethod: processedConfig.method ?? 'GET',
            httpStatus: status,
            duration,
            succeeded,
            isRetried: context.isRetried,
        };
        if (context.resource) {
            computedProps.resourceUrl = context.resource;
        }
        // Log all resolved endpoint headers as telemetry properties
        if (context.resolvedEndpointHeaders) {
            for (const [name, value] of Object.entries(context.resolvedEndpointHeaders)) {
                computedProps[`requestHeader.${name}`] = value;
            }
        }
        const authHeader = processedConfig.headers?.['Authorization'];
        if (authHeader) {
            computedProps.requestAuthorizationHeaderLength = authHeader.length;
        }
        if (responseHeaders) {
            const correlationRequestId = responseHeaders['x-ms-correlation-request-id'];
            if (correlationRequestId) {
                computedProps.correlationRequestId = correlationRequestId;
            }
            const serverRequestId = responseHeaders['x-ms-request-id'];
            if (serverRequestId) {
                computedProps.serverRequestId = serverRequestId;
            }
            const contentType = responseHeaders['content-type'];
            if (contentType) {
                computedProps.responseContentTypeHeader = contentType;
            }
            const contentLength = responseHeaders['content-length'];
            if (contentLength) {
                computedProps.contentLength = contentLength;
            }
            const serverLocation = responseHeaders['x-ms-server-location'];
            if (serverLocation) {
                computedProps.serverLocation = serverLocation;
            }
        }
        // Merge: endpoint → request → computed (later wins)
        const properties = { ...endpointProps, ...requestProps, ...computedProps };
        this.logger.trackActivityEvent(originalConfig.operationName ?? 'unknown', properties, 'HttpClient.SendHttpAsync');
    }
}
