// @microsoft/power-apps-actions/dist/services/dataverse/DataverseMetadataService.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { stringifyError } from '../../ErrorUtils.js';
import { parseCsdlMetadata } from './CsdlMetadataParser.js';
export class DataverseMetadataService {
    httpClient;
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    async getEntityDefinitionAsync(orgUrl, tableName, includeOptionSets = true, logger) {
        if (!orgUrl || !tableName) {
            throw new Error('orgUrl and tableName are required parameters');
        }
        const scenario = logger.trackScenario('DataverseMetadataService.GetEntityDefinition', {
            orgUrl,
            tableName,
            includeOptionSets,
        });
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        let additionalScenarioData = {};
        try {
            const baseUrl = orgUrl.endsWith('/') ? orgUrl.slice(0, -1) : orgUrl;
            const selectParam = 'MetadataId,LogicalName,DisplayCollectionName,DisplayName,EntitySetName,Description,HasNotes,IsActivity,IsIntersect,IsManaged,TableType,IsPrivate,IsLogicalEntity,PrimaryIdAttribute,PrimaryNameAttribute,ObjectTypeCode,OwnershipType,Privileges,IsAvailableOffline,IsOfflineInMobileClient';
            const expandParam = 'Attributes,ManyToOneRelationships,OneToManyRelationships,ManyToManyRelationships';
            const requestUri = `EntityDefinitions(LogicalName='${tableName}')` +
                `?$select=${selectParam}` +
                `&$expand=${expandParam}`;
            const fullUrl = `${baseUrl}/api/data/v9.0/${requestUri}`;
            const authResource = baseUrl + '/';
            additionalScenarioData = { url: fullUrl, authResource };
            const response = await this.httpClient.get(fullUrl, {
                authResource,
                headers: {
                    'User-Agent': 'PowerApps',
                    Consistency: 'Strong',
                    Accept: 'application/json',
                    'OData-MaxVersion': '4.0',
                    'OData-Version': '4.0',
                },
            });
            if (includeOptionSets && response.data.Attributes) {
                try {
                    await this.enrichWithOptionSetMetadata(baseUrl, tableName, response.data, authResource, logger);
                }
                catch (optionSetError) {
                    logger.trackErrorEvent('DataverseMetadataService.GetEntityDefinition.OptionsetRetrievalFailed', {
                        tableName,
                        error: stringifyError(optionSetError),
                    });
                }
            }
            if (response.status === 200) {
                scenario.complete(additionalScenarioData);
            }
            else {
                additionalScenarioData.response = response.data;
                additionalScenarioData.status = response.status;
                scenario.completeWithError(additionalScenarioData);
            }
            return response.data;
        }
        catch (error) {
            scenario.failure({ error, ...additionalScenarioData });
            let errorMessage = `Failed to get entity definition for table '${tableName}' from organization '${orgUrl}'`;
            errorMessage += `: ${stringifyError(error)}`;
            throw new Error(errorMessage);
        }
    }
    /**
     * Fetches the OData $metadata CSDL document and returns all Action and Function definitions
     * whose name contains the given search term (case-insensitive).
     */
    async searchCsdlOperations(orgUrl, searchTerm, logger) {
        const scenario = logger.trackScenario('DataverseMetadataService.SearchCsdlOperations', {
            orgUrl,
            searchTerm,
        });
        let additionalScenarioData = {};
        try {
            const baseUrl = orgUrl.endsWith('/') ? orgUrl.slice(0, -1) : orgUrl;
            // v9.2 is required for full CSDL including custom operations (Actions/Functions)
            const fullUrl = `${baseUrl}/api/data/v9.2/$metadata`;
            const authResource = baseUrl + '/';
            additionalScenarioData = { url: fullUrl, authResource };
            const response = await this.httpClient.get(fullUrl, {
                responseType: 'text',
                authResource,
                headers: {
                    'User-Agent': 'PowerApps',
                    Accept: 'application/xml',
                },
            });
            const operations = parseCsdlMetadata(response.data, searchTerm);
            additionalScenarioData.resultCount = operations.length;
            scenario.complete(additionalScenarioData);
            return operations;
        }
        catch (error) {
            scenario.failure({ error, ...additionalScenarioData });
            throw new Error(`Failed to search CSDL operations in organization '${orgUrl}': ${stringifyError(error)}`);
        }
    }
    convertEntityDefinitionToSchema(entityDefinition) {
        const displayName = entityDefinition.DisplayCollectionName?.UserLocalizedLabel?.Label ||
            entityDefinition.DisplayName?.UserLocalizedLabel?.Label ||
            entityDefinition.LogicalName;
        const properties = {};
        const requiredFields = [];
        if (entityDefinition.Attributes) {
            for (const attribute of entityDefinition.Attributes) {
                const propertySchema = {
                    type: this.mapDataverseTypeToJsonSchemaType(attribute.AttributeTypeName?.Value),
                    title: attribute.DisplayName?.UserLocalizedLabel?.Label || attribute.LogicalName,
                    'x-ms-dataverse-attribute': attribute.LogicalName,
                    'x-ms-dataverse-type': attribute.AttributeTypeName?.Value,
                    'x-ms-schema-name': attribute.SchemaName,
                };
                if (attribute.IsValidForCreate === false && attribute.IsValidForUpdate === false) {
                    propertySchema['x-ms-read-only'] = true;
                }
                if (attribute.IsPrimaryId) {
                    propertySchema['x-ms-dataverse-primary-id'] = true;
                }
                if (attribute.IsPrimaryName) {
                    propertySchema['x-ms-dataverse-primary-name'] = true;
                }
                if (attribute.RequiredLevel?.Value === 'ApplicationRequired' ||
                    attribute.RequiredLevel?.Value === 'SystemRequired') {
                    propertySchema.required = true;
                    requiredFields.push(attribute.LogicalName);
                }
                if (attribute.MaxLength && attribute.MaxLength > 0) {
                    propertySchema.maxLength = attribute.MaxLength;
                }
                if (attribute.LogicalName === entityDefinition.PrimaryIdAttribute) {
                    propertySchema['x-ms-keyType'] = 'primary';
                }
                if (attribute.OptionSet) {
                    propertySchema.enum = [];
                    propertySchema['x-ms-enum-values'] = [];
                    propertySchema.Color = [];
                    this.enrichPropertyWithOptionSetData(propertySchema, attribute);
                }
                if (attribute.AttributeTypeName?.Value?.toLowerCase() === 'datetime') {
                    propertySchema.format = 'date-time';
                }
                properties[attribute.LogicalName] = propertySchema;
            }
        }
        return {
            name: entityDefinition.LogicalName,
            title: displayName,
            schema: {
                type: 'array',
                items: {
                    type: 'object',
                    'x-ms-dataverse-entity': true,
                    'x-ms-dataverse-entityset': entityDefinition.EntitySetName,
                    'x-ms-dataverse-primary-id': entityDefinition.PrimaryIdAttribute,
                    'x-ms-dataverse-primary-name': entityDefinition.PrimaryNameAttribute,
                    properties,
                    required: requiredFields.length > 0 ? requiredFields : undefined,
                },
            },
        };
    }
    mapDataverseTypeToJsonSchemaType(dataverseType) {
        if (!dataverseType) {
            return 'string';
        }
        switch (dataverseType.toLowerCase()) {
            case 'string':
            case 'memo':
            case 'entityname':
                return 'string';
            case 'integer':
            case 'bigint':
                return 'integer';
            case 'decimal':
            case 'double':
            case 'money':
                return 'number';
            case 'boolean':
                return 'boolean';
            case 'datetime':
                return 'string';
            case 'uniqueidentifier':
                return 'string';
            case 'lookup':
            case 'owner':
            case 'customer':
                return 'object';
            case 'picklist':
            case 'state':
            case 'status':
                return 'object';
            default:
                return 'string';
        }
    }
    async enrichWithOptionSetMetadata(baseUrl, tableName, entityDefinition, authResource, logger) {
        const scenario = logger.trackScenario('DataverseMetadataService.EnrichWithOptionSetMetadata', {
            tableName,
        });
        try {
            const optionSetRequests = this.createOptionSetRequests(baseUrl, tableName, authResource, logger);
            const optionSetResponses = await Promise.allSettled(optionSetRequests);
            this.mergeOptionSetData(entityDefinition, optionSetResponses);
            scenario.complete();
        }
        catch (error) {
            scenario.failure({ error });
        }
    }
    createOptionSetRequests(baseUrl, tableName, authResource, logger) {
        const optionSetRequestTemplate = `${baseUrl}/api/data/v9.0/EntityDefinitions(LogicalName='${tableName}')/Attributes/Microsoft.Dynamics.CRM.{0}` +
            '?$select=MetadataId,LogicalName,SchemaName,DisplayName' +
            '&$expand=OptionSet';
        const requestHeaders = {
            'User-Agent': 'PowerApps',
            Consistency: 'Strong',
            Accept: 'application/json',
            'OData-MaxVersion': '4.0',
            'OData-Version': '4.0',
        };
        const optionSetTypes = [
            'PicklistAttributeMetadata',
            'MultiSelectPicklistAttributeMetadata',
            'StateAttributeMetadata',
            'StatusAttributeMetadata',
            'BooleanAttributeMetadata',
            'EntityNameAttributeMetadata',
        ];
        return optionSetTypes.map(async (type) => {
            logger.trackActivityEvent('DataverseMetadataService.CreateOptionSetRequest', {
                optionSetRequestTemplate: optionSetRequestTemplate.replace('{0}', type),
            });
            try {
                const response = await this.httpClient.get(optionSetRequestTemplate.replace('{0}', type), {
                    authResource,
                    headers: requestHeaders,
                });
                return { type, data: response.data };
            }
            catch (error) {
                logger.trackErrorEvent('DataverseMetadataService.CreateOptionSetRequest.Failure', {
                    optionSetRequestTemplate: optionSetRequestTemplate.replace('{0}', type),
                    error: stringifyError(error),
                });
                throw error;
            }
        });
    }
    mergeOptionSetData(entityDefinition, optionSetResponses) {
        if (!entityDefinition.Attributes) {
            return;
        }
        const optionSetAttributes = new Map();
        for (const result of optionSetResponses) {
            if (result.status === 'fulfilled' &&
                result.value?.data &&
                typeof result.value.data === 'object' &&
                result.value.data !== null &&
                'value' in result.value.data &&
                Array.isArray(result.value.data.value)) {
                for (const attribute of result.value.data.value) {
                    if (attribute && typeof attribute === 'object' && 'LogicalName' in attribute) {
                        optionSetAttributes.set(attribute.LogicalName, attribute);
                    }
                }
            }
        }
        for (const attribute of entityDefinition.Attributes) {
            const optionSetData = optionSetAttributes.get(attribute.LogicalName);
            if (optionSetData && optionSetData.OptionSet) {
                attribute.OptionSet = optionSetData.OptionSet;
            }
        }
    }
    enrichPropertyWithOptionSetData(propertySchema, attribute) {
        const optionSet = attribute.OptionSet;
        if (!optionSet) {
            return;
        }
        if (optionSet.MetadataId) {
            propertySchema['x-ms-optionsetmetadataid'] = optionSet.MetadataId;
        }
        if (optionSet.IsGlobal !== undefined) {
            propertySchema['x-ms-isGlobal'] = optionSet.IsGlobal;
        }
        if (optionSet.DisplayName?.UserLocalizedLabel?.Label) {
            propertySchema['x-ms-optionsetdisplayname'] = optionSet.DisplayName.UserLocalizedLabel.Label;
        }
        if (optionSet.OptionSetType) {
            propertySchema['x-ms-optionSetType'] = optionSet.OptionSetType;
        }
        if (optionSet &&
            typeof optionSet === 'object' &&
            'TrueOption' in optionSet &&
            'FalseOption' in optionSet) {
            const booleanOptionSet = optionSet;
            const labels = [
                booleanOptionSet.FalseOption?.Label?.UserLocalizedLabel?.Label || 'False',
                booleanOptionSet.TrueOption?.Label?.UserLocalizedLabel?.Label || 'True',
            ];
            const values = [
                booleanOptionSet.FalseOption?.Value ?? 0,
                booleanOptionSet.TrueOption?.Value ?? 1,
            ];
            const colors = [
                booleanOptionSet.FalseOption?.Color || null,
                booleanOptionSet.TrueOption?.Color || null,
            ];
            propertySchema.enum = labels;
            propertySchema['x-ms-enum-values'] = values;
            propertySchema.Color = colors;
        }
        else if (optionSet.Options && optionSet.Options.length > 0) {
            const labels = [];
            const values = [];
            const colors = [];
            for (const option of optionSet.Options) {
                if (option.Value !== undefined) {
                    const label = option.Label?.UserLocalizedLabel?.Label || option.Value.toString();
                    labels.push(label);
                    values.push(option.Value);
                    colors.push(option.Color || null);
                }
            }
            propertySchema.enum = labels;
            propertySchema['x-ms-enum-values'] = values;
            propertySchema.Color = colors;
        }
    }
}
