// @microsoft/power-apps-actions/dist/CodeGen/shared/dataverseUtils.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { convertToValidIdentifier } from './nameUtils.js';
import { isJsonObject } from './schemaUtils.js';
/** Returns true if the schema root represents a Dataverse entity table. */
export function isDataverseEntity(root) {
    return (isJsonObject(root) &&
        isJsonObject(root.schema) &&
        isJsonObject(root.schema.items) &&
        root.schema.items['x-ms-dataverse-entity'] === true);
}
/** Returns the Dataverse entity set name for a schema (used as the OData API path), or null if not defined. */
export function getDataverseEntitySetName(root) {
    if (isJsonObject(root) && isJsonObject(root.schema) && isJsonObject(root.schema.items)) {
        const entitySetName = root.schema.items['x-ms-dataverse-entityset'];
        if (typeof entitySetName === 'string' && entitySetName.trim()) {
            return entitySetName;
        }
    }
    return null;
}
/** Returns true if the property is a Dataverse lookup relationship to another table. */
export function isDataverseLookupProperty(propValue) {
    return propValue['x-ms-dataverse-type'] === 'LookupType';
}
/** Collects all lookup relationship columns from a properties node, keyed by property name. */
export function getDataverseLookupColumns(propertiesNode) {
    const lookupColumns = new Map();
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        if (isJsonObject(propValue) && isDataverseLookupProperty(propValue)) {
            lookupColumns.set(propName, propValue);
        }
    }
    return lookupColumns;
}
/** Returns true if the property is an auto-generated name field for a lookup column (e.g. `owneridname` for `ownerid`). */
export function isLookupRelatedProperty(propName, lookupColumns) {
    for (const lookupPropName of lookupColumns.keys()) {
        if (propName === `${lookupPropName}name`) {
            return true;
        }
    }
    return false;
}
/** Returns the write-form property name for a lookup column, used when creating or updating Dataverse records. */
export function generateODataBindPropertyName(schemaName) {
    return `${schemaName}@odata.bind`;
}
/** Returns the read-form property name for a lookup's resolved GUID value. */
export function generateLookupValuePropertyName(lookupPropertyName) {
    return `_${lookupPropertyName}_value`;
}
/**
 * Returns true if the property is a Dataverse boolean (yes/no) field.
 * These fields have enum/x-ms-enum-values in the schema but must be typed as `boolean`,
 * not as a numeric optionset, because the Dataverse Web API expects true/false.
 */
export function isDataverseBooleanProperty(propValue) {
    return (propValue['x-ms-dataverse-type'] === 'BooleanType' ||
        propValue['x-ms-optionSetType'] === 'Boolean');
}
/**
 * Returns true if the property is a Dataverse numeric field (integer, decimal, double, money, bigint).
 * These fields have `"type": "string"` in the schema but must be typed as `number` because
 * the Dataverse Web API sends and receives them as JSON numbers.
 */
export function isDataverseNumericProperty(propValue) {
    const dvType = typeof propValue['x-ms-dataverse-type'] === 'string'
        ? propValue['x-ms-dataverse-type'].toLowerCase()
        : undefined;
    switch (dvType) {
        case 'integer':
        case 'integertype':
        case 'bigint':
        case 'biginttype':
        case 'decimal':
        case 'decimaltype':
        case 'double':
        case 'doubletype':
        case 'money':
        case 'moneytype':
            return true;
        default:
            return false;
    }
}
/**
 * Returns true if the property is a Dataverse multi-select picklist (multi-choice) field.
 * These fields allow multiple selections and should be typed as an array of the enum type.
 */
export function isDataverseMultiSelectPicklistProperty(propValue) {
    return propValue['x-ms-dataverse-type'] === 'MultiSelectPicklistType';
}
/**
 * Returns the names of all MultiSelectPicklist (multi-choice) properties in a Dataverse entity
 * properties node — i.e., properties that have optionset data AND are typed as MultiSelectPicklistType.
 * These fields require serialization (array → comma-separated string) when writing to the Dataverse API
 * and deserialization (comma-separated string → array) when reading back from it.
 */
export function getDataverseMultiSelectPicklistFields(propertiesNode) {
    const fields = [];
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        if (isJsonObject(propValue) &&
            isDataverseMultiSelectPicklistProperty(propValue) &&
            Array.isArray(propValue.enum) &&
            propValue.enum.length > 0) {
            fields.push(propName);
        }
    }
    return fields;
}
/** Returns true if the property represents an optionset (choice) column with enumerable values. */
export function hasOptionsetData(propValue) {
    if (isDataverseBooleanProperty(propValue)) {
        return false;
    }
    return Array.isArray(propValue.enum) && propValue.enum.length > 0;
}
/** Produces the TypeScript identifier used for the enum or const generated from an optionset property. */
export function generateOptionsetEnumName(modelName, propName) {
    return convertToValidIdentifier(`${modelName}${propName}`);
}
/** Formats an optionset display label as a TypeScript string literal for use in union types and const declarations. */
export function formatEnumLabel(label) {
    if (!label || label.trim() === '') {
        return "''";
    }
    return `'${label.replace(/'/g, "\\'")}'`;
}
/** Yields each property in a Dataverse entity's properties node, tagged by its role and mutability, for use in interface generation. */
export function* iterateDataverseProperties(propertiesNode, lookupColumns) {
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        if (isJsonObject(propValue)) {
            const readOnly = !!propValue['x-ms-read-only'];
            if (lookupColumns.has(propName)) {
                const schemaNameRaw = propValue['x-ms-schema-name'];
                const schemaName = typeof schemaNameRaw === 'string' ? schemaNameRaw : propName;
                yield { type: 'lookup', readOnly, propName, propValue, schemaName };
            }
            else if (isLookupRelatedProperty(propName, lookupColumns)) {
                yield { type: 'lookupRelated', readOnly, propName, propValue };
            }
            else {
                yield { type: 'normal', readOnly, propName, propValue };
            }
        }
    }
}
