// @microsoft/power-apps-actions/dist/CodeGen/shared/nameUtils.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
// TypeScript reserved keywords and built-in types
const RESERVED_KEYWORDS = new Set([
    'break',
    'case',
    'catch',
    'class',
    'const',
    'continue',
    'debugger',
    'default',
    'delete',
    'do',
    'else',
    'enum',
    'export',
    'extends',
    'false',
    'finally',
    'for',
    'function',
    'if',
    'import',
    'in',
    'instanceof',
    'new',
    'null',
    'return',
    'super',
    'switch',
    'this',
    'throw',
    'true',
    'try',
    'typeof',
    'var',
    'void',
    'while',
    'with',
    'as',
    'implements',
    'interface',
    'let',
    'package',
    'private',
    'protected',
    'public',
    'static',
    'yield',
    'any',
    'boolean',
    'constructor',
    'declare',
    'get',
    'module',
    'require',
    'number',
    'set',
    'string',
    'symbol',
    'type',
    'from',
    'of',
]);
/** Converts an arbitrary string to a safe TypeScript identifier for use in generated code. */
export function convertToValidIdentifier(variableOrInterfaceName) {
    if (variableOrInterfaceName.indexOf('@odata') > -1) {
        // this an odata property, we can skip processing it further and return immediately, but just return a quoted version of it
        return `"${variableOrInterfaceName}"`;
    }
    return sanitizeName(variableOrInterfaceName, false);
}
/** Sanitizes a raw string to a valid TypeScript identifier, for use when generating type and variable names. */
export function sanitizeName(name, removeWhitespace = true, prefixReservedKeywords = true) {
    const originalName = name;
    // Remove any leading or trailing whitespace
    name = name.trim();
    if (!name) {
        throw new Error('Invalid identifier: empty string');
    }
    if (removeWhitespace) {
        name = name.replace(/\s/g, '');
    }
    // Remove any characters that are not alphanumeric or underscores
    name = name.replace(/[^a-zA-Z0-9_$]/g, '_');
    // Ensure the name does not start with a number
    if (/^\d/.test(name)) {
        name = `_${name}`;
    }
    // Ensure the name is not empty or all underscores
    if (name.length === 0 || /^_+$/.test(name)) {
        throw new Error(`Failed to sanitize string ${originalName}`);
    }
    if (prefixReservedKeywords && RESERVED_KEYWORDS.has(name)) {
        name = `_${name}`;
    }
    return name;
}
/** Sanitizes a property name for use in a TypeScript interface, without prefixing reserved keywords. */
export function sanitizePropertyName(name) {
    if (name.indexOf('@odata') > -1) {
        return `"${name}"`;
    }
    return sanitizeName(name, false, false);
}
/** Returns a unique sanitized name, appending a numeric suffix to avoid collisions with existing names. */
export function getSanitizedUniqueName(existingNames, desiredName) {
    // Sanitize the base name
    desiredName = sanitizeName(desiredName);
    const nameSet = new Set(existingNames);
    if (!nameSet.has(desiredName)) {
        return desiredName.toLowerCase();
    }
    let maxSuffix = 0;
    // Non-literal regex shoud be safe here since baseName is sanitized
    // and should not contain any special characters that could lead to regex injection.
    const regex = new RegExp(`^${desiredName}_(\\d+)$`);
    for (const name of existingNames) {
        const match = name.match(regex);
        if (match) {
            const num = parseInt(match[1], 10);
            if (num > maxSuffix) {
                maxSuffix = num;
            }
        }
    }
    return `${desiredName}_${maxSuffix + 1}`.toLowerCase();
}
/** Normalizes a type name containing generic bracket syntax into a form suitable as a TypeScript identifier. */
export function normalizeGenericTypeName(typeName) {
    return typeName.replace(/\[/g, '_').replace(/\]/g, '');
}
/**
 * Extracts a clean identifier from a potentially schema-qualified resource name,
 * for use in generated TypeScript identifiers.
 *
 * Handles bracket-qualified names used by SQL, Oracle, and other tabular connectors
 * (e.g. `[dbo].[TableName]` or `[OWNER].[TABLE_NAME]`), as well as plain names.
 * Always takes the last segment after the final dot, strips brackets, then sanitizes.
 */
export function sanitizeTabularResourceName(name) {
    const result = name.trim().replace(/\[|\]/g, '');
    const match = result.match(/\.?([^.]*)$/);
    if (match) {
        return sanitizeName(match[1]);
    }
    return sanitizeName(result);
}
/** Returns the data source name accessor for the given model's service class. */
export function getServiceName(modelName) {
    if (!modelName || !modelName.trim()) {
        throw new Error('Model name cannot be null or empty.');
    }
    return `${modelName}Service.dataSourceName`;
}
/** Applies well-known casing fixes to property names before they are emitted into generated types. */
export function normalizePropertyName(propertyName) {
    return propertyName.toLowerCase() === 'resultsets' ? 'ResultSets' : propertyName;
}
/** Converts a kebab-case string to camelCase (e.g., `common-models` → `commonModels`). */
export function kebabToCamelCase(name) {
    return name.replace(/-+(.)/g, (_, c) => c.toUpperCase()).replace(/-+$/, '');
}
/** Wraps a property name in quotes when it cannot appear as a bare TypeScript identifier. */
export function formatPropertyName(propName) {
    // Skip if already quoted
    if (propName.startsWith('"') && propName.endsWith('"')) {
        return propName;
    }
    return isValidPropertyName(propName) ? propName : `"${propName}"`;
}
/** Returns true if a property name can be used without quotes in a TypeScript type definition. */
export function isValidPropertyName(propName) {
    // A valid property name must start with a letter, underscore, or dollar sign
    // and can only contain letters, numbers, underscores, or dollar signs
    return /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(propName);
}
