// @microsoft/power-apps-common/dist/services/api/PowerAppsApi.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { createPowerPlatformApiEnvironmentEndpoint } from '../ServiceEndpoints.js';
const enc = encodeURIComponent;
/** Typed API client for Power Apps endpoints. No business logic. */
export class PowerAppsApi {
    httpClient;
    env;
    constructor(httpClient, clusterCategory, environmentName) {
        this.httpClient = httpClient;
        this.env = httpClient.registerEndpoint('powerapps:env', createPowerPlatformApiEnvironmentEndpoint(environmentName, clusterCategory));
    }
    /** Launch an app via the environment-scoped Power Platform API. */
    async launchApp(appName, options) {
        const bypassCache = options?.bypassCache ?? true;
        let url = `${this.env}/powerapps/apps/${enc(appName)}/launch?api-version=2`;
        if (bypassCache) {
            url += '&bypass-cache=true';
        }
        if (options?.record) {
            url += '&record=true';
        }
        const response = await this.httpClient.post(url);
        return response.data;
    }
    /** Fetch environment metadata by name. */
    async getEnvironment(environmentName) {
        const response = await this.httpClient.get(`${this.env}/powerapps/environment?api-version=1&%24filter=name+eq+%27${enc(environmentName)}%27`);
        return response.data;
    }
    /** Evaluate DLP policies for an app. */
    async evaluateDlpPolicies(app) {
        const response = await this.httpClient.post(`${this.env}/powerapps/evaluateDlpPoliciesForApp?api-version=1`, app);
        return response.data;
    }
}
