// @microsoft/power-apps-common/dist/services/ConnectionService.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
/** Default connection service that delegates HTTP calls to ConnectorsApi. */
export class ConnectionService {
    api;
    constructor(api) {
        this.api = api;
    }
    async saveUserConsents(appName, consents) {
        await this.api.saveUserConsents(appName, consents);
    }
    async getConnectionConsent(connection, options) {
        await this.api.requestConnectionConsent(connection.name, {
            disableManualLogin: options?.disableManualLogin,
        });
    }
    isSsoAuthValid(connection) {
        return connection.isOAuthConnection === true && !connection.isAuthenticated;
    }
    async getConnectionById(connectionId, _bypassCache) {
        try {
            return await this.api.getConnectionById(connectionId);
        }
        catch {
            return null;
        }
    }
    async testConnection(connection) {
        await this.api.testConnection(connection.name);
    }
    async getConnection(connectionName, _apiId) {
        return this.api.getConnection(connectionName);
    }
    async createCompositeConnection(apiId, environment, dependencyTable, parameterHints, connectionName) {
        return this.api.createCompositeConnection(apiId, {
            environment: environment.name,
            dependencyTable,
            parameterHints,
            connectionName,
        });
    }
    async saveConnectionUsage(apiShortName, connectionName, environmentName, reference, resourceId, appActions) {
        await this.api.saveConnectionUsage(apiShortName, connectionName, environmentName, {
            reference,
            resourceId,
            appActions,
        });
    }
    async runDisambiguationRules(missingConnections, _appMetadata, _appName) {
        // Default: return missing connections as-is (no auto-mapping).
        return missingConnections;
    }
}
