// @microsoft/power-apps-common/dist/services/PowerPlatformApiDiscovery.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
/**
 * Resolves Power Platform API endpoints for tenant- and environment-scoped URLs.
 *
 * Each cluster category maps to a hostname suffix (e.g. `api.powerplatform.com` for prod).
 * Tenant/environment endpoints encode the identifier into the hostname using hex-encoding
 * conventions defined by the Power Platform API infrastructure.
 */
export class PowerPlatformApiDiscovery {
    clusterCategory;
    constructor(clusterCategory) {
        this.clusterCategory = clusterCategory;
    }
    /** Token audience for the Power Platform API in this cluster. */
    getTokenAudience() {
        return `https://${this.getHostNameSuffix()}`;
    }
    /** Token endpoint host for the Power Platform API in this cluster. */
    getTokenEndpointHost() {
        return this.getHostNameSuffix();
    }
    /** Resolve a tenant-scoped endpoint URL. */
    getTenantEndpoint(tenantId) {
        return this.generateDomain(tenantId, 'tenant');
    }
    /** Resolve a tenant island-cluster endpoint URL. */
    getTenantIslandClusterEndpoint(tenantId) {
        return this.generateDomain(tenantId, 'tenant', 'il-');
    }
    /** Resolve an environment-scoped endpoint URL. */
    getEnvironmentEndpoint(environmentName) {
        return this.generateDomain(environmentName, 'environment');
    }
    // ---------------------------------------------------------------------------
    // Private helpers
    // ---------------------------------------------------------------------------
    generateDomain(hostNameIdentifier, hostNameInfix, hostNamePrefix = '') {
        if (!/^[a-zA-Z0-9-]+$/g.test(hostNameIdentifier)) {
            throw new Error(`Cannot generate Power Platform API endpoint because the ${hostNameInfix} identifier contains invalid host name characters, only alphanumeric and dash characters are expected: ${hostNameIdentifier}`);
        }
        const hexNameSuffixLength = this.getHexApiSuffixLength();
        const hexName = hostNameIdentifier.toLowerCase().replace(/-/g, '');
        if (hexNameSuffixLength >= hexName.length) {
            throw new Error(`Cannot generate Power Platform API endpoint because the normalized ${hostNameInfix} identifier must be at least ${hexNameSuffixLength + 1} characters in length: ${hexName}`);
        }
        const hexNameSuffix = hexName.substring(hexName.length - hexNameSuffixLength);
        const hexNamePrefix = hexName.substring(0, hexName.length - hexNameSuffixLength);
        const hostNameSuffix = this.getHostNameSuffix();
        return `${hostNamePrefix}${hexNamePrefix}.${hexNameSuffix}.${hostNameInfix}.${hostNameSuffix}`;
    }
    getHexApiSuffixLength() {
        switch (this.clusterCategory) {
            case 'firstrelease':
            case 'prod':
                return 2;
            default:
                return 1;
        }
    }
    getHostNameSuffix() {
        switch (this.clusterCategory) {
            case 'local':
                return 'api.powerplatform.localhost';
            case 'dev':
                return 'api.dev.powerplatform.com';
            case 'test':
                return 'api.test.powerplatform.com';
            case 'preprod':
                return 'api.preprod.powerplatform.com';
            case 'firstrelease':
            case 'prod':
                return 'api.powerplatform.com';
            case 'gov':
                return 'api.gov.powerplatform.microsoft.us';
            case 'high':
                return 'api.high.powerplatform.microsoft.us';
            case 'dod':
                return 'api.appsplatform.us';
            case 'mooncake':
                return 'api.powerplatform.partner.microsoftonline.cn';
            case 'ex':
                return 'api.powerplatform.eaglex.ic.gov';
            case 'rx':
                return 'api.powerplatform.microsoft.scloud';
        }
    }
}
