// @microsoft/power-apps-actions/dist/CodeGen/v2/Dataverse/generateDataverseService.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { Scope } from 'ts-morph';
import { convertToValidIdentifier } from '../../shared/nameUtils.js';
export async function generateDataverseService({ parentContext, primaryKey, hasLookupColumns, multiSelectPicklistFields, fileColumnNames, imageColumnNames, serviceFilePath, dataSourceInfoPath, }) {
    const { modelName, dataSourceName, project } = parentContext;
    const serviceName = `${convertToValidIdentifier(modelName)}Service`;
    // When there are lookup columns the model generator emits a separate ModelNameBase interface
    // for write operations (create/update) and ModelName extends it for read operations.
    // When there are none, a single ModelName interface covers all operations.
    const writeTypeName = hasLookupColumns
        ? `${convertToValidIdentifier(modelName)}Base`
        : convertToValidIdentifier(modelName);
    const readTypeName = convertToValidIdentifier(modelName);
    const id = convertToValidIdentifier(modelName);
    const fileColumnNameType = `${id}FileColumnName`;
    const imageColumnNameType = `${id}ImageColumnName`;
    const uploadColumnNameType = `${id}UploadColumnName`;
    const hasFileColumns = fileColumnNames.length > 0;
    const hasImageColumns = imageColumnNames.length > 0;
    const hasFileOrImageColumns = hasFileColumns || hasImageColumns;
    const hasMultiSelectFields = multiSelectPicklistFields.length > 0;
    const createParamType = primaryKey !== '' ? `Omit<${writeTypeName}, '${primaryKey}'>` : writeTypeName;
    const updateParamType = primaryKey !== '' ? `Partial<Omit<${writeTypeName}, '${primaryKey}'>>` : writeTypeName;
    const sourceFile = await project.createFile('service', serviceFilePath, '');
    // External package imports — fixMissingImports cannot resolve these.
    const dataValueImports = ['getClient'];
    if (hasMultiSelectFields) {
        dataValueImports.push('deserializeMultiSelectPicklistFields', 'serializeMultiSelectPicklistFields');
    }
    sourceFile.addImportDeclaration({
        moduleSpecifier: '@microsoft/power-apps/data',
        namedImports: dataValueImports,
    });
    sourceFile.addImportDeclaration({
        moduleSpecifier: '@microsoft/power-apps/data',
        namedImports: ['IOperationResult'],
        isTypeOnly: true,
    });
    sourceFile.addImportDeclaration({
        moduleSpecifier: '@microsoft/power-apps/data/metadata/dataverse',
        namedImports: ['EntityMetadata', 'GetEntityMetadataOptions'],
        isTypeOnly: true,
    });
    sourceFile.addImportDeclaration({
        moduleSpecifier: dataSourceInfoPath,
        namedImports: ['dataSourcesInfo'],
    });
    // Internal in-project type imports - fixMissingImports would auto-add these at commit time,
    // but it cannot detect that these are type-only imports. We add them explicitly to make them type-only.
    sourceFile.addImportDeclaration({
        moduleSpecifier: sourceFile.getRelativePathAsModuleSpecifierTo(project.getSourceFile('commonModel')),
        namedImports: ['IGetAllOptions', 'IGetOptions'],
        isTypeOnly: true,
    });
    const modelNamedImports = hasLookupColumns ? [writeTypeName, readTypeName] : [readTypeName];
    if (hasFileColumns) {
        modelNamedImports.push(fileColumnNameType);
    }
    if (hasImageColumns) {
        modelNamedImports.push(imageColumnNameType);
    }
    if (hasFileOrImageColumns) {
        modelNamedImports.push(uploadColumnNameType);
    }
    sourceFile.addImportDeclaration({
        moduleSpecifier: sourceFile.getRelativePathAsModuleSpecifierTo(project.getSourceFile('model')),
        namedImports: modelNamedImports,
        isTypeOnly: true,
    });
    const multiSelectFieldsInitializer = `[${multiSelectPicklistFields.map((f) => `'${f}'`).join(', ')}] as const`;
    sourceFile.addClass({
        name: serviceName,
        isExported: true,
        properties: [
            {
                name: 'dataSourceName',
                isStatic: true,
                isReadonly: true,
                scope: Scope.Private,
                initializer: `'${dataSourceName.toLowerCase()}'`,
            },
            {
                name: 'client',
                isStatic: true,
                isReadonly: true,
                scope: Scope.Private,
                initializer: 'getClient(dataSourcesInfo)',
            },
            // Emitted only when the entity has MultiSelectPicklist columns. Used as the field list
            // passed to serializeMultiSelectPicklistFields / deserializeMultiSelectPicklistFields.
            ...(hasMultiSelectFields
                ? [
                    {
                        name: 'multiSelectPicklistFields',
                        isStatic: true,
                        isReadonly: true,
                        scope: Scope.Private,
                        initializer: multiSelectFieldsInitializer,
                    },
                ]
                : []),
        ],
        methods: [
            {
                name: 'create',
                isStatic: true,
                isAsync: true,
                scope: Scope.Public,
                parameters: [{ name: 'record', type: createParamType }],
                returnType: `Promise<IOperationResult<${readTypeName}>>`,
                statements: hasMultiSelectFields
                    ? [
                        `const result = await ${serviceName}.client.createRecordAsync<Record<string, unknown>, ${readTypeName}>(${serviceName}.dataSourceName, serializeMultiSelectPicklistFields(record as Record<string, unknown>, ${serviceName}.multiSelectPicklistFields));`,
                        `if (result.data) { deserializeMultiSelectPicklistFields(result.data as Record<string, unknown>, ${serviceName}.multiSelectPicklistFields); }`,
                        'return result;',
                    ]
                    : [
                        `const result = await ${serviceName}.client.createRecordAsync<${createParamType}, ${readTypeName}>(${serviceName}.dataSourceName, record);`,
                        'return result;',
                    ],
            },
            {
                name: 'update',
                isStatic: true,
                isAsync: true,
                scope: Scope.Public,
                parameters: [
                    { name: 'id', type: 'string' },
                    { name: 'changedFields', type: updateParamType },
                ],
                returnType: `Promise<IOperationResult<${readTypeName}>>`,
                statements: hasMultiSelectFields
                    ? [
                        `const result = await ${serviceName}.client.updateRecordAsync<Record<string, unknown>, ${readTypeName}>(${serviceName}.dataSourceName, id, serializeMultiSelectPicklistFields(changedFields as Record<string, unknown>, ${serviceName}.multiSelectPicklistFields));`,
                        `if (result.data) { deserializeMultiSelectPicklistFields(result.data as Record<string, unknown>, ${serviceName}.multiSelectPicklistFields); }`,
                        'return result;',
                    ]
                    : [
                        `const result = await ${serviceName}.client.updateRecordAsync<${updateParamType}, ${readTypeName}>(${serviceName}.dataSourceName, id, changedFields);`,
                        'return result;',
                    ],
            },
            {
                name: 'delete',
                isStatic: true,
                isAsync: true,
                scope: Scope.Public,
                parameters: [{ name: 'id', type: 'string' }],
                returnType: 'Promise<void>',
                statements: [
                    `await ${serviceName}.client.deleteRecordAsync(${serviceName}.dataSourceName, id);`,
                ],
            },
            {
                name: 'get',
                isStatic: true,
                isAsync: true,
                scope: Scope.Public,
                parameters: [
                    { name: 'id', type: 'string' },
                    { name: 'options', type: 'IGetOptions', hasQuestionToken: true },
                ],
                returnType: `Promise<IOperationResult<${readTypeName}>>`,
                statements: hasMultiSelectFields
                    ? [
                        `const result = await ${serviceName}.client.retrieveRecordAsync<${readTypeName}>(${serviceName}.dataSourceName, id, options);`,
                        `if (result.data) { deserializeMultiSelectPicklistFields(result.data as Record<string, unknown>, ${serviceName}.multiSelectPicklistFields); }`,
                        'return result;',
                    ]
                    : [
                        `const result = await ${serviceName}.client.retrieveRecordAsync<${readTypeName}>(${serviceName}.dataSourceName, id, options);`,
                        'return result;',
                    ],
            },
            {
                name: 'getAll',
                isStatic: true,
                isAsync: true,
                scope: Scope.Public,
                parameters: [{ name: 'options', type: 'IGetAllOptions', hasQuestionToken: true }],
                returnType: `Promise<IOperationResult<${readTypeName}[]>>`,
                statements: hasMultiSelectFields
                    ? [
                        `const result = await ${serviceName}.client.retrieveMultipleRecordsAsync<${readTypeName}>(${serviceName}.dataSourceName, options);`,
                        `result.data?.forEach(record => deserializeMultiSelectPicklistFields(record as Record<string, unknown>, ${serviceName}.multiSelectPicklistFields));`,
                        'return result;',
                    ]
                    : [
                        `const result = await ${serviceName}.client.retrieveMultipleRecordsAsync<${readTypeName}>(${serviceName}.dataSourceName, options);`,
                        'return result;',
                    ],
            },
            {
                name: 'getMetadata',
                isStatic: true,
                scope: Scope.Public,
                parameters: [
                    {
                        name: 'options',
                        type: `GetEntityMetadataOptions<${readTypeName}>`,
                        initializer: '{}',
                    },
                ],
                returnType: `Promise<IOperationResult<Partial<EntityMetadata>>>`,
                statements: [
                    `return ${serviceName}.client.executeAsync({`,
                    `  dataverseRequest: {`,
                    `    action: 'getEntityMetadata',`,
                    `    parameters: {`,
                    `      tableName: ${serviceName}.dataSourceName,`,
                    `      options: options as GetEntityMetadataOptions,`,
                    `    },`,
                    `  },`,
                    `});`,
                ],
            },
            ...(hasFileOrImageColumns
                ? [
                    {
                        name: 'upload',
                        isStatic: true,
                        isAsync: true,
                        scope: Scope.Public,
                        parameters: [
                            { name: 'id', type: 'string' },
                            { name: 'columnName', type: uploadColumnNameType },
                            { name: 'file', type: 'File' },
                            { name: 'fileDisplayName', type: 'string', hasQuestionToken: true },
                        ],
                        returnType: 'Promise<IOperationResult<void>>',
                        statements: [
                            'const arrayBuffer = await file.arrayBuffer();',
                            'const data = new Uint8Array(arrayBuffer);',
                            `const result = await ${serviceName}.client.uploadFileToRecord(`,
                            `  ${serviceName}.dataSourceName,`,
                            `  id,`,
                            `  columnName,`,
                            `  fileDisplayName || file.name,`,
                            `  data,`,
                            `);`,
                            'return result;',
                        ],
                    },
                ]
                : []),
            ...(hasFileColumns
                ? [
                    {
                        name: 'downloadFile',
                        isStatic: true,
                        isAsync: true,
                        scope: Scope.Public,
                        parameters: [
                            { name: 'id', type: 'string' },
                            { name: 'columnName', type: fileColumnNameType },
                        ],
                        returnType: 'Promise<IOperationResult<Uint8Array>>',
                        statements: [
                            `const result = await ${serviceName}.client.downloadFileFromRecord(`,
                            `  ${serviceName}.dataSourceName,`,
                            `  id,`,
                            `  columnName,`,
                            `);`,
                            'return result;',
                        ],
                    },
                ]
                : []),
            ...(hasImageColumns
                ? [
                    {
                        name: 'downloadImage',
                        isStatic: true,
                        isAsync: true,
                        scope: Scope.Public,
                        parameters: [
                            { name: 'id', type: 'string' },
                            { name: 'columnName', type: imageColumnNameType },
                            { name: 'fullSize', type: 'boolean', initializer: 'false' },
                        ],
                        returnType: 'Promise<IOperationResult<Uint8Array>>',
                        statements: [
                            `const result = await ${serviceName}.client.downloadImageFromRecord(`,
                            `  ${serviceName}.dataSourceName,`,
                            `  id,`,
                            `  columnName,`,
                            `  fullSize,`,
                            `);`,
                            'return result;',
                        ],
                    },
                ]
                : []),
            ...(hasFileOrImageColumns
                ? [
                    {
                        name: 'deleteFileOrImage',
                        isStatic: true,
                        isAsync: true,
                        scope: Scope.Public,
                        parameters: [
                            { name: 'id', type: 'string' },
                            { name: 'columnName', type: uploadColumnNameType },
                        ],
                        returnType: 'Promise<IOperationResult<void>>',
                        statements: [
                            `const result = await ${serviceName}.client.deleteFileOrImageFromRecord(`,
                            `  ${serviceName}.dataSourceName,`,
                            `  id,`,
                            `  columnName,`,
                            `);`,
                            'return result;',
                        ],
                    },
                ]
                : []),
        ],
    });
}
export default generateDataverseService;
