// @microsoft/power-apps-actions/dist/Config/V1/Schema.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { Regions } from '@microsoft/power-apps-common/services';
import { z } from 'zod';
export const VERSION = '1.0';
export const ShareLinkDetailsSchema = z.object({
    linkKind: z.number(),
    shareId: z.string(),
    url: z.string(),
});
export const DataSetSchema = z.object({
    dataSources: z.record(z.string(), z.object({
        tableName: z.string().optional(),
    })),
    datasetOverride: z
        .object({
        environmentVariableName: z.string(),
    })
        .optional(),
    siteId: z.string().optional(),
});
export const WorkflowDetailsSchema = z.object({
    workflowEntityId: z.string(),
    workflowDisplayName: z.string(),
    workflowName: z.string(),
    // Maps flow connection reference (e.g. "shared_office365") to the UUID key of the
    // dependency connection reference in the top-level connectionReferences record.
    dependencies: z.record(z.string(), z.string()).optional(),
});
export const ConnectionReferenceSchema = z.object({
    id: z.string(),
    displayName: z.string(),
    dataSources: z.array(z.string()).optional(),
    authenticationType: z.string().nullable().optional(),
    sharedConnectionId: z.string().nullable().optional(),
    dataSets: z.record(z.string(), DataSetSchema).optional(),
    xrmConnectionReferenceLogicalName: z.string().optional(),
    dataSourcesShareLinks: z.record(z.string(), ShareLinkDetailsSchema).optional(),
    workflowDetails: WorkflowDetailsSchema.optional(),
});
export const DataverseDataSourceSchema = z.object({
    entitySetName: z.string(),
    logicalName: z.string(),
    isHidden: z.boolean().default(false),
});
export const DatabaseReferenceSchema = z.object({
    dataSources: z.record(z.string(), DataverseDataSourceSchema),
    environmentVariableName: z.string(),
});
export const ConfigSchemaV1 = z.object({
    // Marking as optional for backward compatibility
    version: z.literal(VERSION).optional(),
    appId: z.string().optional().nullable(),
    appDisplayName: z.string().optional(),
    description: z.string().nullable().optional(),
    environmentId: z.string().optional(),
    buildPath: z.string().optional(),
    buildEntryPoint: z.string().optional(),
    logoPath: z.string().optional(),
    localAppUrl: z.string().optional(),
    region: z.enum(Regions).optional(),
    connectionReferences: z.record(z.string(), ConnectionReferenceSchema).optional(),
    databaseReferences: z.record(z.string(), DatabaseReferenceSchema).optional(),
});
