// @microsoft/power-apps-actions/dist/Actions/ListFlows.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { DataverseSystemDataService, getEnvironmentByName, getPlayerServiceConfig, } from '../services/index.js';
/**
 * Maps a DataverseWorkflow to the public Workflow interface
 * @param dataverseWorkflow The internal Dataverse workflow object
 * @returns A Workflow object with cleaned up field names
 */
function mapToWorkflow(dataverseWorkflow) {
    return {
        name: dataverseWorkflow.name,
        modifiedOn: dataverseWorkflow.modifiedon,
        workflowId: dataverseWorkflow.resourceid,
        statecode: dataverseWorkflow.statecode,
    };
}
/**
 * Get all cloud flows from a Dataverse environment.
 * The org URL is derived from the environment ID in the context.
 *
 * @param {ListFlowsContext} context
 * @returns {Promise<Workflow[]>}
 */
export async function listFlowsAsync(context) {
    // Get environment metadata to retrieve the Dataverse org URL
    const metadata = await getEnvironmentByName(context.environmentName);
    const orgUrl = metadata?.properties?.linkedEnvironmentMetadata?.instanceApiUrl;
    if (!orgUrl) {
        throw new Error('Unable to determine the Dataverse organization URL. Ensure the environment has a linked Dataverse instance.');
    }
    const { httpClient } = getPlayerServiceConfig();
    const dataService = new DataverseSystemDataService(httpClient);
    const dataverseWorkflows = (await dataService.getWorkflows(orgUrl, context.logger, context.actionsParams.search)) || [];
    return dataverseWorkflows.map(mapToWorkflow);
}
