// @microsoft/power-apps-actions/dist/Actions/RemoveFlow.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { equalsIgnoreCase } from '@microsoft/power-apps-common/utils';
import { processDataSourceInfo } from '../CodeGen/dataSourceInfoProcessor.js';
import { clearAndRegenerateModelService } from '../CodeGen/modelServiceGenerator.js';
import { sanitizeName } from '../CodeGen/shared/nameUtils.js';
import { readRepoConfig, writeRepoConfig } from '../Config/ConfigManager.js';
import { initializePlayerActions } from '../initializePlayerActions.js';
import { getVfs } from '../VfsManager.js';
import { normalizeParam } from './DataUtils.js';
import { FLOW_API_NAME, LOGIC_FLOWS_API_ID } from './flowConstants.js';
/**
 * Removes a flow from the current Power Apps code app by undoing what addFlowAsync performed:
 * - Removes the flow's connection reference from power.config
 * - Removes stub dependency connection references no longer referenced by any other flow
 * - Deletes the flow's schema file and cleans up the empty directory
 * - Regenerates models and services
 *
 * @param {RemoveFlowContext} context
 * @returns {Promise<RemoveFlowResult>}
 */
export async function removeFlowAsync(context) {
    const flowDataSourceName = await normalizeParam(context.actionsParams.flowDataSourceName);
    const flowId = await normalizeParam(context.actionsParams.flowId);
    if (!flowDataSourceName && !flowId) {
        return {
            success: false,
            found: false,
            error: 'Either --flow-name or --flow-id must be provided.',
        };
    }
    initializePlayerActions({
        vfs: context.vfs,
        authProvider: context.authProvider,
        region: context.region,
        environmentName: context.environmentName,
        logger: context.logger,
    });
    try {
        const { powerConfigPath, schemaPath, codeGenPath } = context.localFilePaths;
        const vfs = getVfs();
        if (!(await vfs.exists(powerConfigPath))) {
            throw new Error(`Configuration file not found at ${powerConfigPath}`);
        }
        const config = await readRepoConfig(powerConfigPath);
        if (!config.connectionReferences) {
            return { success: true, found: false };
        }
        const flowEntryId = flowId
            ? findFlowById(config.connectionReferences, flowId)
            : findFlowByDataSourceName(config.connectionReferences, flowDataSourceName);
        if (!flowEntryId) {
            return { success: true, found: false };
        }
        const flowEntry = config.connectionReferences[flowEntryId];
        // Validate required fields before making any mutations.
        const flowDisplayName = flowEntry.workflowDetails?.workflowDisplayName;
        if (!flowDisplayName) {
            throw new Error(`Flow entry ${flowEntryId} is missing workflowDetails.workflowDisplayName`);
        }
        const dependencyUuids = Object.values(flowEntry.workflowDetails?.dependencies ?? {});
        // Remove the flow's own connection reference.
        delete config.connectionReferences[flowEntryId];
        // Remove stub dependency connection references that are no longer referenced by any flow.
        for (const depUuid of dependencyUuids) {
            const dep = config.connectionReferences[depUuid];
            if (!dep)
                continue;
            const isStub = !dep.dataSources || dep.dataSources.length === 0;
            const isReferencedByOtherFlow = isFlowDependency(config.connectionReferences, depUuid);
            if (isStub && !isReferencedByOtherFlow) {
                delete config.connectionReferences[depUuid];
            }
        }
        await writeRepoConfig(powerConfigPath, config);
        const directoryPath = vfs.join(schemaPath, FLOW_API_NAME);
        const schemaFilePath = vfs.join(directoryPath, `${sanitizeName(flowDisplayName)}.Schema.json`);
        if (await vfs.exists(schemaFilePath)) {
            await vfs.unlink(schemaFilePath);
            // Remove the directory if it is now empty.
            if (await vfs.exists(directoryPath)) {
                const files = await vfs.readdir(directoryPath);
                if (files.length === 0) {
                    try {
                        await vfs.rmdir(directoryPath);
                    }
                    catch {
                        // Not critical if cleanup fails.
                    }
                }
            }
        }
        await processDataSourceInfo(schemaPath, context.logger);
        await clearAndRegenerateModelService({
            schemaFolderPath: schemaPath,
            codeGenPath,
            logger: context.logger,
        });
        return { success: true, found: true };
    }
    catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        return { success: false, found: false, error: errorMessage };
    }
}
/**
 * Finds the UUID of a flow's connection reference whose dataSources array contains
 * the given flow data source name (case-insensitive).
 */
function findFlowByDataSourceName(connectionReferences, flowDataSourceName) {
    const normalizedName = flowDataSourceName.toLowerCase();
    for (const [uuid, ref] of Object.entries(connectionReferences)) {
        if (equalsIgnoreCase(ref.id, LOGIC_FLOWS_API_ID) &&
            ref.dataSources?.some((ds) => ds.toLowerCase() === normalizedName)) {
            return uuid;
        }
    }
    return undefined;
}
/**
 * Finds the UUID of a flow's connection reference whose workflowDetails.workflowName matches
 * the given flow resource ID (case-insensitive).
 */
function findFlowById(connectionReferences, flowId) {
    const normalizedId = flowId.toLowerCase();
    for (const [uuid, ref] of Object.entries(connectionReferences)) {
        if (equalsIgnoreCase(ref.id, LOGIC_FLOWS_API_ID) &&
            ref.workflowDetails?.workflowName?.toLowerCase() === normalizedId) {
            return uuid;
        }
    }
    return undefined;
}
/**
 * Returns true if the given UUID appears as a dependency value in any remaining flow entry.
 */
function isFlowDependency(connectionReferences, uuid) {
    return Object.values(connectionReferences).some((ref) => ref.workflowDetails?.dependencies !== undefined &&
        Object.values(ref.workflowDetails.dependencies).includes(uuid));
}
