// @microsoft/power-apps-actions/dist/Common/HttpClient.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { randomUUID } from 'crypto';
const MS_CLIENT_REQUESTID = 'x-ms-client-request-id';
export class ActionsHttpClient {
    _authenticationProvider;
    constructor(_authenticationProvider) {
        this._authenticationProvider = _authenticationProvider;
    }
    async get(url, config) {
        return this._sendRequestAsync(url, 'GET', config);
    }
    async post(url, config) {
        return this._sendRequestAsync(url, 'POST', config);
    }
    async patch(url, config) {
        return this._sendRequestAsync(url, 'PATCH', config);
    }
    async put(url, config) {
        return this._sendRequestAsync(url, 'PUT', config);
    }
    async _sendRequestAsync(url, method, config) {
        const token = await this._authenticationProvider.getAccessTokenForResource(config?.authResource || '');
        const headers = {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${token}`,
            ...config?.headers,
        };
        headers[MS_CLIENT_REQUESTID] = randomUUID();
        const requestOptions = {
            method,
            headers,
            // Only include body if it exists
            body: config?.body ? JSON.stringify(config.body) : undefined,
        };
        const response = await fetch(url, requestOptions);
        if (!response.ok) {
            throw new Error(`HTTP error status: ${response.status} for ${method} ${url}`);
        }
        return {
            data: (config?.responseType === 'text' ? await response.text() : await response.json()),
            status: response.status,
            headers: this._transformHeaders(response.headers),
        };
    }
    _transformHeaders(headers) {
        const result = {};
        headers.forEach((value, key) => {
            result[key] = value;
        });
        return result;
    }
}
