// @microsoft/power-apps-actions/dist/services/index.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
// Core
export { getPlayerServiceConfig, initializePlayerServices, updateEnvironmentName, updateRegion, } from './initializePlayerServices.js';
export { getDatasetsAsync, getSqlStoredProcedureMetadataAsync, getSqlStoredProceduresAsync, getTableMetadataAsync, getTablesAsync, getTabularConnectorVersion, } from './connectivity/ApimService.js';
export { getConnectionByNameAsync, getConnectorAsync, isConnectionShareable, listConnectionsForConnectorAsync, } from './connectivity/ConnectivityService.js';
// Dataverse
export { DataverseMetadataService } from './dataverse/index.js';
export { DataverseSystemDataService } from './dataverse/DataverseSystemDataService.js';
// PowerApps RP
export { createApp, endSession, generateResourceStorage, getEnvironmentByName, listAppsAsync, makeAppSolutionAware, publishApp, saveApp, startSession, } from './powerAppsRp/PowerAppsRpService.js';
// Power Automate RP
export { getFlowMetadataAsync, getFlowSwaggerAsync, } from './powerAutomateRp/PowerAutomateRpService.js';
