// @microsoft/power-apps-actions/dist/services/initializePlayerServices.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
let _playerServicesConfig;
export function initializePlayerServices(options) {
    const { logger, httpClient, region, environmentName } = options;
    _playerServicesConfig = {
        logger,
        httpClient,
        region,
        environmentName,
        featureGates: {
            usePowerPlatformAPI: options.featureGates?.usePowerPlatformAPI || false,
        },
        apiVersions: {
            bapRp: options.apiVersions?.bapRp || '2021-04-01',
            launchService: options.apiVersions?.launchService || '2',
            launchOperations: options.apiVersions?.launchOperations || '1',
            monitoringHubLogAppOpen: options.apiVersions?.monitoringHubLogAppOpen || '2',
            monitoringHubLogAppNavigate: options.apiVersions?.monitoringHubLogAppNavigate || '1',
            monitoringHubLogDataRequests: options.apiVersions?.monitoringHubLogDataRequests || '1',
            requestAccess: options.apiVersions?.requestAccess || '1',
            permissions: options.apiVersions?.permissions || '1',
            modifyPermissions: options.apiVersions?.modifyPermissions || '1',
            evaluateEntitlements: options.apiVersions?.evaluateEntitlements || '1',
            appExtendedMetadata: options.apiVersions?.appExtendedMetadata || '1',
            getPreloadMetadata: options.apiVersions?.getPreloadMetadata || '1',
            getPublishMetadata: options.apiVersions?.getPublishMetadata || '3',
            copilotStatus: options.apiVersions?.copilotStatus || '1',
            neptune: options.apiVersions?.neptune || '1',
            environment: options.apiVersions?.environment || '1',
        },
    };
}
export function getPlayerServiceConfig() {
    return _playerServicesConfig;
}
export function updateEnvironmentName(environmentName) {
    _playerServicesConfig.environmentName = environmentName;
}
export function updateRegion(region) {
    _playerServicesConfig.region = region;
}
export function updateLogger(logger) {
    _playerServicesConfig.logger = logger;
}
