// @microsoft/power-apps-actions/dist/services/PowerPlatformApiUtils.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { PowerPlatformApiDiscovery, } from '@microsoft/power-apps-common/services';
import { getPlayerServiceConfig } from './initializePlayerServices.js';
import { getServiceConfig } from './ServiceUtils.js';
export const getAuthResource = () => {
    return {
        test: new PowerPlatformApiDiscovery('test').getTokenAudience(),
        preprod: new PowerPlatformApiDiscovery('preprod').getTokenAudience(),
        preview: new PowerPlatformApiDiscovery('firstrelease').getTokenAudience(),
        prod: new PowerPlatformApiDiscovery('prod').getTokenAudience(),
        gccmoderate: new PowerPlatformApiDiscovery('gov').getTokenAudience(),
        gcchigh: new PowerPlatformApiDiscovery('high').getTokenAudience(),
        dod: new PowerPlatformApiDiscovery('dod').getTokenAudience(),
        mooncake: new PowerPlatformApiDiscovery('mooncake').getTokenAudience(),
        ex: new PowerPlatformApiDiscovery('ex').getTokenAudience(),
        rx: new PowerPlatformApiDiscovery('rx').getTokenAudience(),
    };
};
export const getBapRPAuthResource = () => {
    return {
        test: 'https://service.powerapps.com/',
        preprod: 'https://service.powerapps.com/',
        preview: 'https://service.powerapps.com/',
        prod: 'https://service.powerapps.com/',
        gccmoderate: 'https://gov.service.powerapps.us/',
        gcchigh: 'https://high.service.powerapps.us/',
        dod: 'https://service.apps.appsplatform.us/',
        mooncake: 'https://service.powerapps.cn/',
        ex: 'https://service.powerapps.eaglex.ic.gov/',
        rx: 'https://service.powerapps.microsoft.scloud/',
    };
};
export function getPowerPlatformApiEnvironmentConfig(authResource) {
    const { environmentName } = getPlayerServiceConfig();
    return getServiceConfig({
        endpoint: {
            test: new PowerPlatformApiDiscovery('test').getEnvironmentEndpoint(environmentName),
            preprod: new PowerPlatformApiDiscovery('preprod').getEnvironmentEndpoint(environmentName),
            preview: new PowerPlatformApiDiscovery('firstrelease').getEnvironmentEndpoint(environmentName),
            prod: new PowerPlatformApiDiscovery('prod').getEnvironmentEndpoint(environmentName),
            gccmoderate: new PowerPlatformApiDiscovery('gov').getEnvironmentEndpoint(environmentName),
            gcchigh: new PowerPlatformApiDiscovery('high').getEnvironmentEndpoint(environmentName),
            dod: new PowerPlatformApiDiscovery('dod').getEnvironmentEndpoint(environmentName),
            mooncake: new PowerPlatformApiDiscovery('mooncake').getEnvironmentEndpoint(environmentName),
            ex: new PowerPlatformApiDiscovery('ex').getEnvironmentEndpoint(environmentName),
            rx: new PowerPlatformApiDiscovery('rx').getEnvironmentEndpoint(environmentName),
        },
        authResource: authResource || getAuthResource(),
    });
}
