// @microsoft/power-apps-actions/dist/CodeGen/dataSourceInfoProcessor.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { getVfs } from '../VfsManager.js';
import { isDataverseEntity } from './shared/dataverseUtils.js';
import { extractModelName } from './shared/modelUtils.js';
import { getAllJsonFiles, isJsonObject } from './shared/schemaUtils.js';
import { getSqlStoredProcApiKey } from './shared/sqlUtils.js';
/** Regex to match file extension (e.g., ".json", ".ts") */
const FILE_EXTENSION_PATTERN = /\.[^/.]+$/;
/**
 * Processes all JSON files in the specified folder and generates a TypeScript file with data source information.
 * @param schemaFolderPath The path to the folder containing JSON files.
 */
export async function processDataSourceInfo(schemaFolderPath, logger) {
    const scenario = logger.trackScenario('ProcessDataSourceInfo', { schemaFolderPath });
    const vfs = getVfs();
    try {
        const jsonFiles = await getAllJsonFiles(schemaFolderPath);
        const resultJson = {};
        for (const filePath of jsonFiles) {
            try {
                const jsonContent = await vfs.readFile(filePath, 'utf-8');
                const jsonDocument = JSON.parse(jsonContent);
                const dataSourceInfo = processJson(jsonDocument);
                // note: this should use the same function as used in modelServiceGenerator to avoid mismatches
                const propertyKey = determinePropertyKey(jsonDocument, filePath, vfs, logger);
                mergeDataSourceInfo(resultJson, propertyKey, dataSourceInfo);
            }
            catch (err) {
                const error = err;
                throw new Error(`Error processing file '${filePath}': ${error.message}`);
            }
        }
        const outputJson = JSON.stringify(resultJson, null, 2); // pretty print
        const finalOutput = `/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 * This file is auto-generated. Do not modify it manually.
 * Changes to this file may be overwritten.
 */

export const dataSourcesInfo = ${outputJson};
`;
        const outputDir = vfs.join(schemaFolderPath, 'appschemas');
        const outputFilePath = vfs.join(outputDir, 'dataSourcesInfo.ts');
        // Create the appschemas directory if it doesn't exist
        if (!(await vfs.exists(outputDir))) {
            await vfs.mkdir(outputDir, { recursive: true });
        }
        await vfs.writeFile(outputFilePath, finalOutput);
        scenario.complete();
    }
    catch (error) {
        scenario.failure({ error });
        throw new Error(`An unexpected error occurred: ${error.message}`);
    }
}
/**
 * Merges new data source information into the result JSON object.
 * If the property key already exists, it merges the APIs, avoiding duplicates.
 * @param resultJson The result JSON object to merge into.
 * @param propertyKey The key under which to store the data source information.
 * @param newDataSourceInfo The new data source information to merge.
 */
function mergeDataSourceInfo(resultJson, propertyKey, newDataSourceInfo) {
    if (Object.prototype.hasOwnProperty.call(resultJson, propertyKey)) {
        const existingDataSource = resultJson[propertyKey];
        if (existingDataSource && existingDataSource.apis) {
            for (const apiKey in newDataSourceInfo.apis) {
                if (!Object.prototype.hasOwnProperty.call(existingDataSource.apis, apiKey)) {
                    existingDataSource.apis[apiKey] = newDataSourceInfo.apis[apiKey];
                }
            }
        }
        else {
            throw new Error(`The existing data source for key '${propertyKey}' is not of type DataSourceInfo.`);
        }
    }
    else {
        resultJson[propertyKey] = newDataSourceInfo;
    }
}
/**
 * Determines the property key for a data source, falling back to filename if extraction fails.
 * @param jsonDocument The parsed JSON document.
 * @param filePath The path to the JSON file.
 * @param vfs The virtual file system instance.
 * @param logger The logger instance for telemetry.
 * @returns The property key to use for the data source.
 */
function determinePropertyKey(jsonDocument, filePath, vfs, logger) {
    try {
        return extractModelName(jsonDocument, filePath, vfs).dataSourceName.toLowerCase();
    }
    catch (error) {
        logger.trackActivityEvent('DataSourceInfoProcessor_FallbackName', {
            filePath,
            error: error instanceof Error ? error.message : String(error),
        });
        // fall back to using the file name without extension
        const fileName = vfs.basename(filePath);
        return fileName.replace(FILE_EXTENSION_PATTERN, '').toLowerCase();
    }
}
/**
 * Processes the JSON document and extracts relevant data source information.
 * @param json The JSON document to process.
 * @param existingDataSourceInfo Optional existing data source information to merge with.
 * @returns The processed data source information.
 */
function processJson(json, existingDataSourceInfo) {
    const root = json;
    if (typeof root.type === 'string' && root.type === 'Microsoft.PowerApps/apis') {
        const dataSourceInfo = {
            tableId: '',
            version: '',
            primaryKey: '',
            dataSourceType: 'Connector',
            apis: {},
        };
        const properties = isJsonObject(root.properties) ? root.properties : null;
        const swagger = properties && isJsonObject(properties.swagger) ? properties.swagger : null;
        const paths = swagger && isJsonObject(swagger.paths) ? swagger.paths : null;
        if (paths) {
            for (const [pathKey, pathValue] of Object.entries(paths)) {
                if (isJsonObject(pathValue)) {
                    for (const [methodKey, methodValue] of Object.entries(pathValue)) {
                        if (isJsonObject(methodValue)) {
                            const operation = methodValue;
                            const operationId = operation.operationId;
                            if (typeof operationId === 'string') {
                                const apiDefinition = {
                                    path: pathKey,
                                    method: methodKey.toUpperCase(),
                                };
                                apiDefinition.parameters =
                                    operation.parameters && Array.isArray(operation.parameters)
                                        ? processParameters(operation.parameters, swagger)
                                        : [];
                                if (operation.responses) {
                                    apiDefinition.responseInfo = processResponse(operation.responses, swagger);
                                }
                                else {
                                    apiDefinition.responseInfo = {
                                        default: { type: 'void', format: undefined },
                                    };
                                }
                                dataSourceInfo.apis[operationId] = apiDefinition;
                            }
                        }
                    }
                }
            }
        }
        return dataSourceInfo;
    }
    else if (isJsonObject(root.schema)) {
        const schema = root.schema;
        if (isDataverseEntity(root)) {
            const items = schema.items;
            // This is a Dataverse entity schema
            const primaryKey = typeof items['x-ms-dataverse-primary-id'] === 'string'
                ? items['x-ms-dataverse-primary-id']
                : '';
            return {
                tableId: '',
                version: '',
                primaryKey,
                dataSourceType: 'Dataverse',
                apis: {},
            };
        }
        if (typeof schema.procedureresultschema !== 'undefined') {
            const dataSourceInfo = existingDataSourceInfo ?? {
                tableId: '',
                version: 'v2',
                primaryKey: '',
                dataSourceType: 'Connector',
                apis: {},
            };
            const procedureName = typeof root.name === 'string' ? root.name : '';
            const cleanedProcedureName = '/' + procedureName.replace(/\[|\]/g, '');
            const cleanedProcedureNameWithoutSchema = getSqlStoredProcApiKey(procedureName);
            let parameters = [];
            if (isJsonObject(schema.inputparameters) && isJsonObject(schema.inputparameters.properties)) {
                // Stored procedure parameters are wrapped under 'inputParameters' by the codegen.
                // Use a single body param with that name so _buildOperationBodyParam can find the value.
                parameters = [{ name: 'inputParameters', in: 'body', required: false, type: 'object' }];
            }
            const apiDefinition = {
                path: cleanedProcedureName,
                method: 'POST',
                parameters,
            };
            if (!dataSourceInfo.apis[cleanedProcedureNameWithoutSchema]) {
                dataSourceInfo.apis[cleanedProcedureNameWithoutSchema] = apiDefinition;
            }
            else {
                throw new Error(`Duplicate stored procedure detected: ${cleanedProcedureName}`);
            }
            return dataSourceInfo;
        }
        else if (typeof root.name === 'string' && root.name.match(/^\[[^\]]+\]\./)) {
            // Use the explicit version annotation injected at add-time when available.
            // Fall back to 'v2' for backward compatibility with schema files created before this annotation was added.
            const version = typeof root['x-ms-connection-version'] === 'string'
                ? root['x-ms-connection-version']
                : 'v2';
            return {
                tableId: root.name,
                version,
                primaryKey: getPrimaryKeyFromSchema(schema),
                dataSourceType: 'Connector',
                apis: {},
            };
        }
        else {
            // Sharepoint List
            const tableId = typeof root.name === 'string' && root.name !== null ? root.name : '';
            const dataSourceInfo = {
                tableId,
                version: '',
                primaryKey: getPrimaryKeyFromSchema(schema),
                dataSourceType: 'Connector',
                apis: {},
            };
            const referencedEntities = isJsonObject(root.referencedEntities)
                ? root.referencedEntities
                : null;
            if (referencedEntities) {
                for (const referencedEntityKey in referencedEntities) {
                    if (!Object.prototype.hasOwnProperty.call(referencedEntities, referencedEntityKey)) {
                        continue;
                    }
                    const referencedEntityValue = referencedEntities[referencedEntityKey];
                    if (isJsonObject(referencedEntityValue) &&
                        referencedEntityValue.lookupEndpoint !== null) {
                        const lookupEndpointProps = referencedEntityValue.lookupEndpoint;
                        if (isJsonObject(lookupEndpointProps) && lookupEndpointProps.path !== null) {
                            const lookupEndpointPath = lookupEndpointProps.path;
                            if (typeof lookupEndpointPath === 'string') {
                                const segments = lookupEndpointPath.split('/');
                                // Process the segments as needed
                                segments[segments.length - 1] = referencedEntityKey; // Replace last segment
                                // in the updated segment remove hyphens in the GUIDs
                                // This might be an SPO connector issue, but we need to work around it here.
                                // ex:  "/tables/05bfc50c-b51b-4c17-aac4-45bd5dbb9532/entities/Author",
                                // becomes "/tables/05bfc50cb51b4c17aac445bd5dbb9532/entities/Author"
                                const guidRegex = /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/g;
                                const updatedSegment = segments
                                    .join('/')
                                    .replace(guidRegex, (match) => match.replace(/-/g, ''));
                                const finalPath = `/{connectionId}/datasets/{dataset}${updatedSegment}`;
                                const apiParameterList = new Array();
                                const defaultApiParameters = ['connectionId', 'dataset', 'table'];
                                const defaultQueryParameters = ['search'];
                                for (const param of defaultApiParameters) {
                                    const apiParameter = {
                                        name: param,
                                        in: 'path',
                                        required: true,
                                        type: 'string',
                                    };
                                    apiParameterList.push(apiParameter);
                                }
                                for (const param of defaultQueryParameters) {
                                    const apiParameter = {
                                        name: param,
                                        in: 'query',
                                        required: false,
                                        type: 'string',
                                    };
                                    apiParameterList.push(apiParameter);
                                }
                                const apiResponseInfo = {
                                    '200': { type: 'array' },
                                };
                                const apiDefinition = {
                                    path: finalPath,
                                    method: 'GET',
                                    parameters: apiParameterList,
                                    responseInfo: apiResponseInfo,
                                };
                                const referencedEntityKeyNormalized = normalizeHyphenatedName(referencedEntityKey);
                                const operationId = `Get${referencedEntityKeyNormalized}`;
                                if (dataSourceInfo.apis[operationId]) {
                                    throw new Error(`Duplicate operationId detected: ${operationId}`);
                                }
                                dataSourceInfo.apis[operationId] = apiDefinition;
                            }
                        }
                    }
                }
            }
            return dataSourceInfo;
        }
    }
    if (typeof root.type === 'string' && root.type === 'Microsoft.PowerApps/dataverseOperation') {
        return processDataverseOperation(root);
    }
    throw new Error('The JSON does not represent a valid data source.');
}
/**
 * Processes the responses node and extracts response information.
 * @param responsesNode The responses node from the Swagger document.
 * @param root The root of the Swagger document for reference.
 * @returns A record of response information keyed by status code.
 */
function processResponse(responses, root) {
    const result = {};
    // Type guard: ensure responsesNode is an object
    if (!isJsonObject(responses)) {
        return result;
    }
    for (const statusCode in responses) {
        if (!Object.prototype.hasOwnProperty.call(responses, statusCode)) {
            continue;
        }
        const respObj = responses[statusCode];
        // Type guard: ensure resp is an object
        if (isJsonObject(respObj)) {
            if (isJsonObject(respObj.schema)) {
                const schemaNode = respObj.schema;
                const type = typeof schemaNode.type === 'string' ? schemaNode.type : null;
                const format = typeof schemaNode.format === 'string' ? schemaNode.format : null;
                // If type is array, always return "array" as type
                if (type === 'array') {
                    result[statusCode] = { type: 'array', format: format || undefined };
                    continue;
                }
                // $ref (single object)
                // eslint-disable-next-line dot-notation
                if (schemaNode['$ref']) {
                    // eslint-disable-next-line dot-notation
                    const refValue = schemaNode['$ref'];
                    if (typeof refValue === 'string' && refValue.startsWith('#/definitions/')) {
                        const defName = refValue.substring('#/definitions/'.length);
                        let refType = 'object';
                        // Type guard for root to access definitions
                        if (isJsonObject(root)) {
                            if (isJsonObject(root.definitions)) {
                                const definitions = root.definitions;
                                const defNode = definitions[defName];
                                if (isJsonObject(defNode)) {
                                    refType = typeof defNode.type === 'string' ? defNode.type : 'object';
                                }
                            }
                        }
                        result[statusCode] = { type: refType, format: format || undefined };
                        continue;
                    }
                }
                // Primitive or object
                result[statusCode] = { type: type || 'object', format: format || undefined };
            }
            else {
                // No schema, but description exists
                result[statusCode] = { type: 'void', format: undefined };
            }
        }
        else {
            // resp is not an object, treat as void
            result[statusCode] = { type: 'void', format: undefined };
        }
    }
    // If no responses at all, return a default void
    if (Object.keys(result).length === 0) {
        result.default = { type: 'void', format: undefined };
    }
    return result;
}
/**
 * Retrieves the primary key from the schema node.
 * @param schemaNode The schema node to extract the primary key from.
 * @returns The primary key as a string, or an empty string if not found.
 */
function getPrimaryKeyFromSchema(schema) {
    // Type guard: ensure schemaNode. items exists and is an object. properties exists and is an object
    if (!isJsonObject(schema) ||
        !isJsonObject(schema.items) ||
        !isJsonObject(schema.items.properties)) {
        return '';
    }
    const propertiesNode = schema.items.properties;
    for (const propertyName in propertiesNode) {
        if (!Object.prototype.hasOwnProperty.call(propertiesNode, propertyName)) {
            continue;
        }
        const propertyValue = propertiesNode[propertyName];
        // Type guard: ensure propertyValue is an object
        if (isJsonObject(propertyValue)) {
            if (propertyValue['x-ms-keyType'] === 'primary') {
                return propertyName;
            }
        }
    }
    return '';
}
/**
 * Processes the parameters node and extracts parameter information.
 * @param parametersNode The parameters node from the Swagger document.
 * @param root The root of the Swagger document for reference.
 * @returns An array of ApiParameter objects representing the parameters.
 */
function processParameters(parametersNode, root) {
    const parameters = [];
    // Type guard: ensure parametersNode is an array
    if (!Array.isArray(parametersNode)) {
        return parameters;
    }
    for (const parameter of parametersNode) {
        // Type guard: ensure parameter is an object
        if (isJsonObject(parameter)) {
            if (parameter.$ref) {
                const refPath = parameter.$ref;
                if (typeof refPath === 'string') {
                    const resolvedParameter = resolveRef(refPath, root);
                    if (resolvedParameter) {
                        parameters.push(parseParameter(resolvedParameter));
                    }
                }
            }
            else {
                parameters.push(parseParameter(parameter));
            }
        }
    }
    // Stable sort: required=true first, then required=false
    return parameters.sort((a, b) => Number(b.required) - Number(a.required));
}
/**
 * Resolves a $ref reference in the Swagger document.
 * @param refPath The $ref path to resolve.
 * @param root The root of the Swagger document for reference.
 * @returns The resolved object or null if not found.
 */
function resolveRef(refPath, root) {
    // Type guard: ensure root is an object
    if (!isJsonObject(root)) {
        return null;
    }
    // Resolve $ref (e.g., "#/swagger/parameters/ParameterName")
    if (refPath.startsWith('#/')) {
        const parts = refPath.substring(2).split('/');
        let currentElement = root;
        for (const part of parts) {
            // Type guard: ensure currentElement is an object
            if (isJsonObject(currentElement)) {
                if (part in currentElement) {
                    currentElement = currentElement[part];
                }
                else {
                    return null; // Reference not found
                }
            }
            else {
                return null; // Current element is not an object
            }
        }
        // Type guard: ensure final result is an object
        if (isJsonObject(currentElement)) {
            return currentElement;
        }
    }
    return null;
}
/**
 * Parses a parameter object and extracts relevant information.
 * @param parameter The parameter object to parse.
 * @returns An ApiParameter object representing the parsed parameter.
 */
function parseParameter(parameter) {
    return {
        name: typeof parameter.name === 'string' ? parameter.name : '',
        in: typeof parameter.in === 'string' ? parameter.in : '',
        required: parameter.required === true,
        type: typeof parameter.type === 'string' ? parameter.type : 'object',
    };
}
/*
 * The normalizeHyphenatedName method normalizes hyphenated names.
 * It removes all hyphens from the name and returns the normalized name.
 */
function normalizeHyphenatedName(name) {
    return name.replace(/-/g, '');
}
/**
 * Processes a Microsoft.PowerApps/dataverseOperation schema into a DataSourceInfo entry.
 * Each operation is keyed by its lowercased operation name (e.g. `winopportunity`).
 *
 * Reads from the embedded Swagger 2.0 spec produced by `buildDataverseOperationSchema`.
 * The swagger is nested under `properties.swagger` (connector schema format) so that the
 * existing `handleSwaggerSchema` codegen path processes it without a separate handler.
 * All type mappings are already resolved in the swagger at write-time.
 */
function processDataverseOperation(root) {
    const props = isJsonObject(root.properties) ? root.properties : null;
    const swagger = props && isJsonObject(props.swagger) ? props.swagger : null;
    if (!swagger) {
        throw new Error('Dataverse operation schema is missing the required "properties.swagger" field.');
    }
    const basePath = typeof swagger.basePath === 'string' ? swagger.basePath : '';
    const paths = isJsonObject(swagger.paths) ? swagger.paths : null;
    if (!paths) {
        throw new Error('Dataverse operation schema swagger is missing "paths".');
    }
    const apis = {};
    for (const [pathKey, pathValue] of Object.entries(paths)) {
        if (!isJsonObject(pathValue))
            continue;
        for (const [methodKey, methodValue] of Object.entries(pathValue)) {
            if (!isJsonObject(methodValue))
                continue;
            const operationId = typeof methodValue.operationId === 'string' ? methodValue.operationId : '';
            if (!operationId)
                continue;
            const parameters = Array.isArray(methodValue.parameters)
                ? methodValue.parameters.filter(isJsonObject).map((param) => ({
                    name: typeof param.name === 'string' ? param.name : '',
                    in: typeof param.in === 'string' ? param.in : 'body',
                    required: param.required === true,
                    type: typeof param.type === 'string' ? param.type : 'object',
                }))
                : [];
            apis[operationId] = {
                path: basePath + pathKey,
                method: methodKey.toUpperCase(),
                parameters,
                responseInfo: processResponse(methodValue.responses, swagger),
            };
        }
    }
    return {
        tableId: '',
        version: '',
        primaryKey: '',
        dataSourceType: 'Dataverse',
        apis,
    };
}
