// @microsoft/power-apps-actions/dist/CodeGen/modelServiceGenerator.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { equalsIgnoreCase } from '@microsoft/power-apps-common/utils';
import * as path from 'path';
import { LOGIC_FLOWS_API_ID } from '../Actions/flowConstants.js';
import { ParameterPropertyInfo } from '../Common/parameterPropertyInfo.js';
import { getVfs } from '../VfsManager.js';
import { formatEnumLabel, generateOptionsetEnumName, getDataverseMultiSelectPicklistFields, hasOptionsetData, isDataverseEntity, } from './shared/dataverseUtils.js';
import { determineTypeScriptType, extractModelName } from './shared/modelUtils.js';
import { convertToValidIdentifier, formatPropertyName, getServiceName, isValidPropertyName, kebabToCamelCase, normalizeGenericTypeName, normalizePropertyName, sanitizeName, sanitizePropertyName, } from './shared/nameUtils.js';
import { CodegenV1Project } from './shared/project.js';
import { extractReferencedType, getAllJsonFiles, getNameNode, getPropertiesNode, getRequiredFields, isJsonObject, mapJsonTypeToTypeScript, } from './shared/schemaUtils.js';
import { isSharepointSchema } from './shared/sharepointUtils.js';
import { getSqlStoredProcApiKey, isSqlStoredProcedure } from './shared/sqlUtils.js';
import { isCodegenV2Context } from './shared/types.js';
import generateDataverseModelV2 from './v2/Dataverse/generateDataverseModel.js';
import generateDataverseServiceV2 from './v2/Dataverse/generateDataverseService.js';
import { CodegenV2Project } from './v2/project.js';
/**
 * TypeScript type constants
 */
const TypeScriptTypes = {
    STRING: 'string',
    NUMBER: 'number',
    BOOLEAN: 'boolean',
    DATE: 'Date',
    UNKNOWN: 'unknown',
    OBJECT: 'Record<string, unknown>',
};
// Configuration constants for output folders
const GENERATED_FOLDER = 'generated';
const MODELS_FOLDER = `${GENERATED_FOLDER}/models`;
const SERVICES_FOLDER = `${GENERATED_FOLDER}/services`;
const RELATIVE_MODEL_PATH = '../models';
/*
 * The main entry point for the model service generator.
 * It processes the schema files in the specified folder and generates TypeScript models and services.
 */
export async function generateModelService(options) {
    const { logger, schemaFolderPath, codeGenPath, schemaFilePath, useV2Codegen } = options;
    const scenario = logger.trackScenario('ModelServiceGenerator.GenerateModelService', options);
    try {
        const vfs = getVfs();
        if (!schemaFolderPath || !codeGenPath) {
            return;
        }
        await ensureOutputDirectoryExists(codeGenPath, vfs);
        let schemaFiles = [];
        if (schemaFilePath) {
            schemaFiles = [schemaFilePath];
        }
        else {
            schemaFiles = await getAllJsonFiles(schemaFolderPath);
        }
        const project = useV2Codegen
            ? new CodegenV2Project(vfs, logger)
            : new CodegenV1Project(vfs, logger);
        for (const schemaFile of schemaFiles) {
            try {
                await ensureOutputDirectoriesExist(codeGenPath, vfs);
                await processSchema(project, {
                    schemaFolder: schemaFolderPath,
                    schemaFile,
                    outputFolder: codeGenPath,
                    vfs: vfs,
                    logger,
                }, !!useV2Codegen);
            }
            catch (error) {
                throw new Error(`Error processing schema at path '${schemaFile}': ${error?.message}`);
            }
        }
        // Generate index.ts file
        await generateIndexFile(codeGenPath, project);
        // Commit all generated files
        await project.commit();
        scenario.complete();
    }
    catch (error) {
        scenario.failure({ error });
    }
}
/**
 * Clears and regenerates the model and service files.
 */
export async function clearAndRegenerateModelService(options) {
    const { codeGenPath } = options;
    const vfs = getVfs();
    const modelsDir = vfs.join(codeGenPath, MODELS_FOLDER);
    const servicesDir = vfs.join(codeGenPath, SERVICES_FOLDER);
    // Delete Models directory if it exists
    if (await vfs.exists(modelsDir)) {
        await vfs.rmdir(modelsDir);
    }
    // Delete Services directory if it exists
    if (await vfs.exists(servicesDir)) {
        await vfs.rmdir(servicesDir);
    }
    await generateModelService(options);
}
/*
 * The EnsureOutputDirectoryExists method ensures that the output directory exists.
 * It creates the directory if it does not exist.
 */
async function ensureOutputDirectoryExists(outputDir, vfs) {
    if (!(await vfs.exists(outputDir))) {
        await vfs.mkdir(outputDir, { recursive: true });
    }
}
/**
 * The processSchema method processes a single schema file.
 * It generates TypeScript models and services based on the schema.
 */
async function processSchema(project, context, isV2Enabled) {
    const { schemaFile, schemaFolder, outputFolder, logger, vfs } = context;
    // Destructure result from initializeSchemaProcessing (returns: schema, modelName, dataSourceName)
    const { schema, modelName, dataSourceName } = await readSchema(schemaFile, vfs);
    // Finished processing schema, create codegen context for codegen. No longer need Vfs, use project.
    const codegenContext = {
        schemaFile,
        schemaFolder,
        outputFolder,
        logger,
        schema,
        modelName,
        dataSourceName,
        project,
        isV2Enabled,
    };
    if (isSwaggerSchema(schema)) {
        await handleSwaggerSchema(codegenContext);
    }
    else if (isSqlStoredProcedure(schema)) {
        await handleSqlStoredProcedure(codegenContext);
    }
    else if (isSharepointSchema(schema)) {
        await handleSharepointSchema(codegenContext);
    }
    else {
        await handleStandardSchema(codegenContext);
    }
}
/**
 * The initializeSchemaProcessing method initializes the schema processing.
 * Returns an object containing the parsed schema, the schema JSON string, the output directory, model name, and data source name.
 */
async function readSchema(schemaPath, vfs) {
    const schema = await loadSchemaFromFile(schemaPath, vfs);
    // Ensure schema is a JsonObject for processing
    if (!isJsonObject(schema)) {
        throw new Error('Schema must be a valid JSON object');
    }
    // model name is used for the generated TypeScript interface and file names
    // dataSourceName is the logical name of the datasource ex:entity set name for Dataverse
    const { modelName, dataSourceName } = extractModelName(schema, schemaPath, vfs);
    return {
        schema,
        modelName,
        dataSourceName,
    };
}
/**
 * The loadSchemaFromFile method loads and parses a schema file.
 */
async function loadSchemaFromFile(schemaPath, vfs) {
    if (!(await vfs.exists(schemaPath))) {
        throw new Error(`Schema file not found at path: ${schemaPath}`);
    }
    const schemaJson = await vfs.readFile(schemaPath, 'utf-8');
    const schema = JSON.parse(schemaJson);
    return schema;
}
/**
 * The ensureOutputDirectoriesExist method ensures that the required output directories exist.
 * It creates the main output directory, plus 'generated/models' and 'generated/services' subfolders if needed.
 */
async function ensureOutputDirectoriesExist(outputDir, vfs) {
    await vfs.mkdir(outputDir, { recursive: true });
    await vfs.mkdir(vfs.join(outputDir, MODELS_FOLDER), { recursive: true });
    await vfs.mkdir(vfs.join(outputDir, SERVICES_FOLDER), {
        recursive: true,
    });
}
/**
 * The isSwaggerSchema method checks if the schema is a Swagger schema.
 * Returns true if schema.properties.swagger exists.
 */
function isSwaggerSchema(schema) {
    return (isJsonObject(schema) &&
        isJsonObject(schema.properties) &&
        Object.prototype.hasOwnProperty.call(schema.properties, 'swagger'));
}
/**
 * Checks if a property is a Dataverse lookup column.
 */
function isDataverseLookupProperty(propValue) {
    return propValue['x-ms-dataverse-type'] === 'LookupType';
}
/**
 * Checks if a property is a Dataverse file column.
 */
function isDataverseFileProperty(propValue) {
    return propValue['x-ms-dataverse-type'] === 'FileType';
}
/**
 * Checks if a property is a Dataverse image column.
 */
function isDataverseImageProperty(propValue) {
    return propValue['x-ms-dataverse-type'] === 'ImageType';
}
/**
 * Returns the names of all file columns in the properties node.
 * Compute once and pass the result down — do not re-traverse the node.
 */
function getFileColumnNames(propertiesNode) {
    const columnNames = [];
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        if (isJsonObject(propValue) && isDataverseFileProperty(propValue)) {
            columnNames.push(propName);
        }
    }
    return columnNames;
}
/**
 * Returns the names of all image columns in the properties node.
 * Compute once and pass the result down — do not re-traverse the node.
 */
function getImageColumnNames(propertiesNode) {
    const columnNames = [];
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        if (isJsonObject(propValue) && isDataverseImageProperty(propValue)) {
            columnNames.push(propName);
        }
    }
    return columnNames;
}
/**
 * Gets all Dataverse lookup columns from the properties object.
 */
function getDataverseLookupColumns(propertiesNode) {
    const lookupColumns = new Map();
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        if (isJsonObject(propValue) && isDataverseLookupProperty(propValue)) {
            lookupColumns.set(propName, propValue);
        }
    }
    return lookupColumns;
}
/**
 * Generates the OData bind property name for a lookup column.
 * Example: "cr160_lookupcol" -> "cr160_lookupcol@odata.bind"
 */
function generateODataBindPropertyName(schemaName) {
    return `${schemaName}@odata.bind`;
}
/**
 * Generates the value property name for a lookup column.
 * Example: "cr160_lookupcol" -> "_cr160_lookupcol_value"
 */
function generateLookupValuePropertyName(lookupPropertyName) {
    return `_${lookupPropertyName}_value`;
}
/**
 * The addCopyrightNotice method adds a copyright notice to the generated files.
 */
function addCopyrightNotice(lines) {
    lines.push('/*!');
    lines.push(' * Copyright (C) Microsoft Corporation. All rights reserved.');
    lines.push(' * This file is autogenerated. Do not edit this file directly.');
    lines.push(' */');
    lines.push('');
}
/**
 * Generates TypeScript interfaces for Dataverse entities with lookup columns.
 * Creates both a base interface (for PATCH/POST) and an extended interface (for GET).
 */
function generateDataverseInterfaces(modelName, definition, fileColumnNames, imageColumnNames) {
    const lines = [];
    addCopyrightNotice(lines);
    const propertiesNode = getPropertiesNode(definition);
    const lookupColumns = getDataverseLookupColumns(propertiesNode);
    const multiSelectPicklistFields = getDataverseMultiSelectPicklistFields(propertiesNode);
    const id = convertToValidIdentifier(modelName);
    const fileTypeAlias = fileColumnNames.length > 0
        ? `\nexport type ${id}FileColumnName = ${fileColumnNames.map((col) => `'${col}'`).join(' | ')};\n`
        : '';
    const imageTypeAlias = imageColumnNames.length > 0
        ? `\nexport type ${id}ImageColumnName = ${imageColumnNames.map((col) => `'${col}'`).join(' | ')};\n`
        : '';
    const uploadColumnNameTypeAlias = fileColumnNames.length > 0 || imageColumnNames.length > 0
        ? `\nexport type ${id}UploadColumnName = ${[
            fileColumnNames.length > 0 ? `${id}FileColumnName` : '',
            imageColumnNames.length > 0 ? `${id}ImageColumnName` : '',
        ]
            .filter(Boolean)
            .join(' | ')};\n`
        : '';
    const columnNameTypeAliases = fileTypeAlias + imageTypeAlias + uploadColumnNameTypeAlias;
    if (lookupColumns.size === 0) {
        // No lookup columns, generate single interface
        const singleInterface = generatePropertyLines(modelName, definition);
        return {
            baseInterface: lines.join('\n') + singleInterface + columnNameTypeAliases,
            extendedInterface: '',
            multiSelectPicklistFields,
        };
    }
    // Generate base interface for PATCH operations
    const baseInterfaceName = `${modelName}Base`;
    const baseInterfaceCode = generateDataverseBaseInterface(baseInterfaceName, modelName, definition, lookupColumns);
    // Generate extended interface for GET operations
    const extendedInterfaceName = modelName;
    const extendedInterfaceCode = generateDataverseExtendedInterface(extendedInterfaceName, baseInterfaceName, definition, lookupColumns);
    const baseInterface = lines.join('\n') + baseInterfaceCode;
    const extendedInterface = extendedInterfaceCode + columnNameTypeAliases;
    return { baseInterface, extendedInterface, multiSelectPicklistFields };
}
/**
 * Generates the base interface for Dataverse PATCH/POST operations.
 * Replaces lookup columns with @odata.bind properties.
 */
function generateDataverseBaseInterface(interfaceName, originalModelName, definition, lookupColumns) {
    const lines = [];
    const requiredFields = getRequiredFields(definition);
    const propertiesNode = getPropertiesNode(definition);
    // Generate optionset enums first - use original model name for enum generation
    const enumLines = generateOptionsetEnums(originalModelName, propertiesNode, '');
    if (enumLines.length > 0) {
        lines.push(...enumLines);
        lines.push('');
    }
    lines.push(`export interface ${convertToValidIdentifier(interfaceName)} {`);
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        const propValueAsObject = isJsonObject(propValue) ? propValue : undefined;
        // Skip read-only properties. They should not be included in PATCH/POST interfaces.
        if (propValueAsObject?.['x-ms-read-only']) {
            continue;
        }
        // Handle lookup columns and their related properties
        if (lookupColumns.has(propName) || isLookupRelatedProperty(propName, lookupColumns)) {
            if (lookupColumns.has(propName)) {
                // Add @odata.bind property instead
                const schemaName = propValueAsObject?.['x-ms-schema-name'];
                // Ensure schemaName is a string before passing to generateODataBindPropertyName
                const schemaNameStr = typeof schemaName === 'string' ? schemaName : propName;
                const odataBindProp = generateODataBindPropertyName(schemaNameStr);
                const formattedPropName = formatPropertyName(odataBindProp);
                const optionalMark = requiredFields.has(propName) ? '' : '?';
                lines.push(`  ${formattedPropName}${optionalMark}: string;`);
            }
            continue;
        }
        let formattedPropName = formatPropertyName(sanitizePropertyName(propName));
        formattedPropName = normalizePropertyName(formattedPropName);
        const tsType = determineTypeScriptType(originalModelName, propValueAsObject, '', propName);
        const optionalMark = requiredFields.has(propName) ? '' : '?';
        addPropertyComment(lines, propValueAsObject, '');
        lines.push(`  ${formattedPropName}${optionalMark}: ${tsType};`);
    }
    lines.push('}');
    return lines.join('\n');
}
/**
 * Generates the extended interface for Dataverse GET operations.
 * Extends the base interface and adds lookup object and value properties.
 */
function generateDataverseExtendedInterface(modelName, baseInterfaceName, definition, lookupColumns) {
    const lines = [];
    lines.push(`export interface ${convertToValidIdentifier(modelName)} extends ${convertToValidIdentifier(baseInterfaceName)} {`);
    const requiredFields = getRequiredFields(definition);
    const propertiesNode = getPropertiesNode(definition);
    // Add read-only properties (excluding lookup columns which are handled separately)
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        const propValueAsObject = isJsonObject(propValue) ? propValue : undefined;
        // Only include read-only properties in the extended interface
        if (!propValueAsObject?.['x-ms-read-only']) {
            continue;
        }
        // Skip lookup columns as they will be handled in the lookup-specific section
        if (lookupColumns.has(propName)) {
            continue;
        }
        let formattedPropName = formatPropertyName(sanitizePropertyName(propName));
        formattedPropName = normalizePropertyName(formattedPropName);
        const tsType = determineTypeScriptType(modelName, propValueAsObject, '', propName);
        const optionalMark = requiredFields.has(propName) ? '' : '?';
        addPropertyComment(lines, propValueAsObject, '');
        lines.push(`  ${formattedPropName}${optionalMark}: ${tsType};`);
    }
    // Add lookup-specific properties
    const lookupPropNames = Array.from(lookupColumns.keys());
    for (const lookupPropName of lookupPropNames) {
        // Add the lookup object property
        lines.push(`  ${formatPropertyName(sanitizePropertyName(lookupPropName))}?: object;`);
        // Add the lookup value property (_propname_value)
        const valuePropName = generateLookupValuePropertyName(lookupPropName);
        const formattedValuePropName = formatPropertyName(valuePropName);
        lines.push(`  ${formattedValuePropName}?: string;`);
    }
    lines.push('}');
    return lines.join('\n');
}
/**
 * Checks if a property is related to a lookup column (like name properties).
 */
function isLookupRelatedProperty(propName, lookupColumns) {
    // Check if this property is a name property for any lookup column
    const lookupPropNames = Array.from(lookupColumns.keys());
    for (const lookupPropName of lookupPropNames) {
        if (propName === `${lookupPropName}name`) {
            return true;
        }
    }
    return false;
}
/**
 * The handleSwaggerSchema method handles the Swagger schema.
 * It generates TypeScript models and services from the Swagger schema.
 */
async function handleSwaggerSchema(context) {
    const { modelName, schema, outputFolder, project } = context;
    const swaggerObject = getSwaggerObject(schema);
    const modelFilePath = project.join(outputFolder, MODELS_FOLDER, `${modelName}Model.ts`);
    const serviceFilePath = project.join(outputFolder, SERVICES_FOLDER, `${modelName}Service.ts`);
    const modelLines = [];
    addCopyrightNotice(modelLines);
    const definitionTypes = processDefinitions(swaggerObject, modelLines);
    const parameterTypes = processParameters(swaggerObject);
    if (definitionTypes.length > 0) {
        const modelCode = modelLines.join('\n');
        await project.createFile('model', modelFilePath, modelCode);
    }
    if ('paths' in swaggerObject && isJsonObject(swaggerObject.paths)) {
        // Dataverse operation schemas use `dataverseOperation`; all other swagger schemas use `connectorOperation`
        const operationCallType = isJsonObject(schema) && schema.type === 'Microsoft.PowerApps/dataverseOperation'
            ? 'dataverseOperation'
            : 'connectorOperation';
        // Build the logical-name → model-name map before generating the service so that
        // resolveEntityTypeFromCsdl can produce correct interface names and import paths.
        const entityModelMap = await buildEntityModelMap(swaggerObject, context);
        // Ensure entity model files exist before the service is written (imports must resolve).
        await ensureReferencedEntityModels(entityModelMap, context);
        const serviceCode = generateServiceFromPaths(swaggerObject.paths, parameterTypes, context, operationCallType, entityModelMap);
        await project.createFile('service', serviceFilePath, serviceCode);
    }
}
/**
 * Collects the entity logical name from a CSDL type string if it is a Dataverse entity reference.
 * Handles both `mscrm.{name}` and `Collection(mscrm.{name})` forms.
 */
function collectEntityName(csdlType, out) {
    if (typeof csdlType !== 'string')
        return;
    const m = csdlType.match(/^(?:Collection\()?mscrm\.(\w+)\)?$/);
    if (m)
        out.add(m[1]);
}
/**
 * Scans a swagger object for `x-ms-csdl-type` entity references and builds a map of
 * logical name → codegen model name, ensuring every referenced entity has a model file.
 *
 * Resolution order for each logical name:
 * 1. Entity schema file present (`{schemaFolder}/dataverse/{name}.Schema.json`): calls
 *    `extractModelName` to get the entity-set-based model name (e.g. `"Opportunities"`) and
 *    delegates to `ensureReferencedEntityModels` to generate the full model if absent.
 * 2. Schema absent but a previously-generated model file exists on disk: reuses it under the
 *    capitalised logical name (e.g. `"OpportunityModel.ts"`).
 * 3. Neither present: the entity is not added to the map. Callers that run
 *    `ensureDataverseEntityDataSource` for each referenced entity before invoking
 *    `generateModelService` will always land in case 1 or 2.
 *
 * Every entry added to the map is backed by an existing or newly-generated model file.
 */
async function buildEntityModelMap(swaggerObject, context) {
    const entityNames = new Set();
    if (isJsonObject(swaggerObject.paths)) {
        for (const pathItem of Object.values(swaggerObject.paths)) {
            if (!isJsonObject(pathItem))
                continue;
            for (const op of Object.values(pathItem)) {
                if (!isJsonObject(op))
                    continue;
                for (const param of Array.isArray(op.parameters) ? op.parameters : []) {
                    if (isJsonObject(param))
                        collectEntityName(param['x-ms-csdl-type'], entityNames);
                }
                const resp200 = isJsonObject(op.responses) && isJsonObject(op.responses['200'])
                    ? op.responses['200']
                    : null;
                if (resp200 && isJsonObject(resp200.schema)) {
                    collectEntityName(resp200.schema['x-ms-csdl-type'], entityNames);
                }
            }
        }
    }
    const map = new Map();
    const vfs = getVfs();
    for (const logicalName of entityNames) {
        const schemaFile = context.project.join(context.schemaFolder, 'dataverse', `${logicalName}.Schema.json`);
        if (await vfs.exists(schemaFile)) {
            try {
                const content = await vfs.readFile(schemaFile, 'utf-8');
                const schema = JSON.parse(content);
                const { modelName } = extractModelName(schema, schemaFile, vfs);
                map.set(logicalName, modelName);
                continue;
            }
            catch {
                // Schema unreadable — fall through to fallback.
            }
        }
        // Schema file absent or unreadable. Use the capitalised logical name as the model name.
        const fallbackModelName = toInterfaceName(logicalName);
        const fallbackModelFile = context.project.join(context.outputFolder, MODELS_FOLDER, `${fallbackModelName}Model.ts`);
        // Reuse an existing model if one is already on disk.
        if (await vfs.exists(fallbackModelFile)) {
            map.set(logicalName, fallbackModelName);
            continue;
        }
        // No entity schema and no existing model: skip this entity.
        // addDataverseApiAsync calls ensureDataverseEntityDataSource before generateModelService,
        // so this branch is only reached if an entity is referenced without being pre-populated.
        continue;
    }
    return map;
}
/**
 * For each entry in `entityModelMap`, ensures the corresponding model file exists in the
 * generated output. If the file is absent, finds the entity schema and runs codegen for it.
 *
 * Called before `generateServiceFromPaths` so that all referenced `*Base` interfaces are
 * on disk before the service file is written (necessary for downstream TypeScript compilation).
 */
async function ensureReferencedEntityModels(entityModelMap, context) {
    for (const [logicalName, modelName] of entityModelMap) {
        const modelFile = context.project.join(context.outputFolder, MODELS_FOLDER, `${modelName}Model.ts`);
        if (await context.project.fileExists(modelFile))
            continue;
        const entitySchemaFile = context.project.join(context.schemaFolder, 'dataverse', `${logicalName}.Schema.json`);
        // Use the VFS directly for the schema file check: project.fileExists checks the in-memory
        // TypeScript project (ts-morph), which does not track JSON schema files.
        if (!(await getVfs().exists(entitySchemaFile)))
            continue;
        await processSchema(context.project, {
            schemaFile: entitySchemaFile,
            schemaFolder: context.schemaFolder,
            outputFolder: context.outputFolder,
            logger: context.logger,
            vfs: getVfs(),
        }, context.isV2Enabled);
    }
}
/**
 * The getSwaggerObject method retrieves the Swagger object from the schema.
 * Throws an error if the schema is not valid Swagger format.
 */
function getSwaggerObject(schema) {
    if (isJsonObject(schema) && isJsonObject(schema.properties)) {
        const swaggerNode = schema.properties.swagger;
        if (!isJsonObject(swaggerNode)) {
            throw new Error("Invalid schema format: 'swagger' node is missing or malformed.");
        }
        return swaggerNode;
    }
    throw new Error('Schema must be a JSON object with properties to extract swagger object');
}
/*
 * The processDefinitions function processes the definitions node of the Swagger schema.
 * It collects model interface names and pushes generated code lines to modelLines.
 */
function processDefinitions(swaggerObject, modelLines) {
    const definitionTypes = [];
    if ('definitions' in swaggerObject && isJsonObject(swaggerObject.definitions)) {
        const definitionsObject = swaggerObject.definitions;
        for (const [interfaceName, definitionValue] of Object.entries(definitionsObject)) {
            if (isJsonObject(definitionValue)) {
                const modelCode = generateModelFromSchema(interfaceName, definitionValue);
                modelLines.push(modelCode);
                modelLines.push(''); // Add a blank line between models
                definitionTypes.push(interfaceName);
            }
        }
    }
    return definitionTypes;
}
/*
 * The processParameters function processes the parameters node of the Swagger schema.
 */
function processParameters(swaggerObject) {
    const parameterTypes = {};
    if ('parameters' in swaggerObject && isJsonObject(swaggerObject.parameters)) {
        const parametersObject = swaggerObject.parameters;
        for (const [interfaceName, parameterValue] of Object.entries(parametersObject)) {
            if (isJsonObject(parameterValue)) {
                const { propertyInfo } = generateModelFromParameters(interfaceName, parameterValue);
                // Optionally: modelLines.push(modelCode); // If you want to collect interface code
                parameterTypes[interfaceName] = propertyInfo;
            }
        }
    }
    return parameterTypes;
}
/**
 * The generateModelFromSchema method generates TypeScript models from the schema.
 * Returns a type alias or interface definition string.
 */
function generateModelFromSchema(modelName, definition) {
    modelName = convertToValidIdentifier(simplifyGenericInterfaceName(modelName));
    const lines = [];
    // If it's a simple type (e.g., string, number, boolean) and has no "properties" or "items", generate a type alias
    // Note: We exclude definitions with "items" because those are array types that need special handling
    if ('type' in definition &&
        definition.type &&
        !('properties' in definition) &&
        !('items' in definition)) {
        const tsType = mapJsonTypeToTypeScript(definition.type);
        lines.push(`export type ${modelName} = ${tsType};`);
        return lines.join('\n');
    }
    // If it's an empty object, generate a type alias for object
    const { isEmpty, itemsNode } = analyzeSchemaStructure(definition);
    if (isEmpty) {
        lines.push(generateTypeAliasForEmptyObject(modelName));
        return lines.join('\n');
    }
    // Check if this is an array type definition
    const isArrayType = definition.type === 'array' && itemsNode;
    if (isArrayType) {
        // If items is a $ref, use the referenced type directly instead of generating a new interface
        if ('$ref' in itemsNode && typeof itemsNode['$ref'] === 'string') {
            const refType = extractReferencedType(itemsNode['$ref']);
            lines.push(`export type ${modelName} = ${refType}[];`);
            return lines.join('\n');
        }
        // For array types with inline item schemas, generate an interface for the item type and a type alias for the array
        const itemInterfaceName = `${modelName}Item`;
        const itemModelCode = generatePropertyLines(itemInterfaceName, itemsNode);
        lines.push(itemModelCode);
        lines.push('');
        lines.push(`export type ${modelName} = ${itemInterfaceName}[];`);
        return lines.join('\n');
    }
    // If itemsNode exists but not an array type, pass that; otherwise, pass the full definition
    const modelCode = itemsNode
        ? generatePropertyLines(modelName, itemsNode)
        : generatePropertyLines(modelName, definition);
    lines.push(modelCode);
    return lines.join('\n');
}
/**
 * The simplifyGenericInterfaceName method simplifies the generic interface name.
 * It removes the generic parameters from the name.
 */
function simplifyGenericInterfaceName(modelName) {
    return normalizeGenericTypeName(modelName);
}
/**
 * Analyzes a JSON Schema definition to determine if it defines meaningful structure.
 *
 * @param definition - The JSON Schema definition to analyze
 * @returns Analysis results containing:
 *   - isEmpty: true if neither properties nor items are defined (or both are empty)
 *   - itemsNode: The items schema if present and valid, otherwise undefined
 */
function analyzeSchemaStructure(definition) {
    const itemsNode = 'items' in definition && isJsonObject(definition.items) ? definition.items : undefined;
    const hasProperties = 'properties' in definition &&
        isJsonObject(definition.properties) &&
        Object.keys(definition.properties).length > 0;
    const hasItems = itemsNode !== undefined && Object.keys(itemsNode).length > 0;
    return { isEmpty: !hasProperties && !hasItems, itemsNode };
}
/**
 * The generateTypeAliasForEmptyObject method generates a TypeScript type alias for an empty object.
 */
function generateTypeAliasForEmptyObject(modelName) {
    return `export type ${convertToValidIdentifier(modelName)} = object;`;
}
/*
 * The GeneratePropertyLines method generates TypeScript property lines from the schema.
 */
function generatePropertyLines(modelName, definition, addInterfaceName = true, padding = '') {
    if (!isJsonObject(definition)) {
        throw new Error('The definition must be a valid JSON object.');
    }
    const lines = [];
    const requiredFields = getRequiredFields(definition);
    const propertiesNode = getPropertiesNode(definition);
    // Generate optionset enums first (before the interface)
    // BUT only for top-level interfaces (addInterfaceName === true)
    // Nested inline objects should not generate enums to avoid exporting them mid-interface
    if (addInterfaceName) {
        // Collect all enums including from nested objects
        const allEnumLines = collectAllEnums(modelName, propertiesNode, padding);
        if (allEnumLines.length > 0) {
            lines.push(...allEnumLines);
            lines.push(''); // Add blank line between enums and interface
        }
    }
    addInterfaceDeclarationWithPadding(lines, modelName, addInterfaceName, padding);
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        let formattedPropName = formatPropertyName(sanitizePropertyName(propName));
        // Normalize specific property names (e.g., "resultsets" => "ResultSets")
        formattedPropName = normalizePropertyName(formattedPropName);
        const propValueAsObject = isJsonObject(propValue) ? propValue : undefined;
        const tsType = determineTypeScriptType(modelName, propValueAsObject, padding, propName);
        const optionalMark = requiredFields.has(propName) ? '' : '?';
        addPropertyComment(lines, propValueAsObject, padding);
        lines.push(`${padding}  ${formattedPropName}${optionalMark}: ${tsType};`);
    }
    closeInterfaceDeclarationWithPadding(lines, padding);
    return lines.join('\n');
}
/**
 * Recursively collects all enum declarations from properties, including nested objects.
 * This ensures enums are hoisted to the top level and not exported mid-interface.
 */
function collectAllEnums(modelName, propertiesNode, padding) {
    const enumLines = [];
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        const propValueAsObject = isJsonObject(propValue) ? propValue : undefined;
        if (!propValueAsObject) {
            continue;
        }
        // Check if this property has optionset data
        if (hasOptionsetData(propValueAsObject)) {
            const enumName = generateOptionsetEnumName(modelName, propName);
            const enumDeclaration = generateOptionsetConstDeclaration(enumName, propValueAsObject, padding);
            enumLines.push(enumDeclaration);
        }
        // Recursively collect enums from nested objects
        if (isJsonObject(propValueAsObject.properties)) {
            const nestedEnums = collectAllEnums(modelName, propValueAsObject.properties, padding);
            enumLines.push(...nestedEnums);
        }
        // Recursively collect enums from array items
        if (propValueAsObject.type === 'array' && isJsonObject(propValueAsObject.items)) {
            const itemsObj = propValueAsObject.items;
            if (isJsonObject(itemsObj.properties)) {
                const itemEnums = collectAllEnums(modelName, itemsObj.properties, padding);
                enumLines.push(...itemEnums);
            }
        }
    }
    return enumLines;
}
/**
 * The generateOptionsetEnums method generates TypeScript const objects for optionset properties.
 * These const objects are compatible with TypeScript's erasableSyntaxOnly flag and provide
 * type safety while preserving the original numeric values from the schema.
 * @returns An array of const object declaration strings.
 */
function generateOptionsetEnums(modelName, propertiesNode, padding) {
    const enumLines = [];
    for (const [propName, propValue] of Object.entries(propertiesNode)) {
        const propValueAsObject = isJsonObject(propValue) ? propValue : undefined;
        // Check if this property has optionset data
        if (propValueAsObject && hasOptionsetData(propValueAsObject)) {
            const enumName = generateOptionsetEnumName(modelName, propName);
            const enumDeclaration = generateOptionsetConstDeclaration(enumName, propValueAsObject, padding);
            enumLines.push(enumDeclaration);
        }
    }
    return enumLines;
}
/**
 * Generates const object declaration for an optionset property.
 * Creates a const object with typed properties that is compatible with erasableSyntaxOnly.
 * Also generates a corresponding type alias for type safety.
 */
function generateOptionsetConstDeclaration(enumName, propValue, padding) {
    // Check for x-ms-enum-values format (array of objects or numbers)
    const enumValues = propValue['x-ms-enum-values'];
    const enumLabels = propValue.enum;
    const lines = [];
    enumName = convertToValidIdentifier(enumName);
    // Check if x-ms-enum-values exists and is an array
    if (enumValues && Array.isArray(enumValues) && enumLabels && Array.isArray(enumLabels)) {
        // Check if x-ms-enum-values contains objects with numeric values (Dataverse format)
        const hasObjectFormat = enumValues.length > 0 && typeof enumValues[0] === 'object' && enumValues[0] !== null;
        if (hasObjectFormat) {
            // This is the Swagger format where x-ms-enum-values contains objects like {displayName, value}
            // Just use the enum array for string union type
            const labelStrings = enumLabels.map((label) => formatEnumLabel(label));
            lines.push(`${padding}export type ${enumName} = ${labelStrings.join('|')};`);
        }
        else {
            // This is the Dataverse format where x-ms-enum-values contains numeric values
            const enumNumericValues = enumValues;
            // Generate const object with numeric keys
            lines.push(`${padding}export const ${enumName} = {`);
            // Handle the format where enum contains display names and x-ms-enum-values contains numeric values
            for (let i = 0; i < enumLabels.length; i++) {
                const displayName = enumLabels[i]; // e.g., "Preferred Customer"
                const numericValue = enumNumericValues[i]; // e.g., 1
                // Use the display name directly as the string value
                let sanitizedLabel;
                if (!displayName || (typeof displayName === 'string' && displayName.trim() === '')) {
                    // Generate a placeholder name for empty labels
                    sanitizedLabel = `EmptyOption${i}`;
                }
                else {
                    sanitizedLabel = sanitizeName(displayName);
                }
                const comma = i < enumLabels.length - 1 ? ',' : '';
                lines.push(`${padding}  ${numericValue}: '${sanitizedLabel}'${comma}`);
            }
            lines.push(`${padding}} as const;`);
            // Generate type alias for the numeric values
            lines.push(`${padding}export type ${enumName} = keyof typeof ${enumName};`);
        }
    }
    else {
        // This is the path for swaggers where only enum array exists (no x-ms-enum-values)
        const labelStrings = enumLabels.map((label) => formatEnumLabel(label));
        lines.push(`${padding}export type ${enumName} = ${labelStrings.join('|')};`);
    }
    return lines.join('\n');
}
/*
 * The addInterfaceDeclarationWithPadding function adds the interface declaration to the lines.
 * It takes the lines to write to, model name, and padding as parameters.
 */
function addInterfaceDeclarationWithPadding(lines, modelName, addInterfaceName, padding) {
    if (addInterfaceName) {
        lines.push(`${padding}export interface ${convertToValidIdentifier(modelName)} {`);
    }
    else {
        lines.push(`{`);
    }
}
/*
 * The closeInterfaceDeclarationWithPadding function closes the interface declaration with padding.
 * It takes the lines to write to and padding as parameters.
 */
function closeInterfaceDeclarationWithPadding(lines, padding) {
    lines.push(`${padding}}`);
}
/*
 * The mapDataverseTypeToTypeScript method maps Dataverse types to TypeScript types.
 * Based on official Microsoft documentation: https://learn.microsoft.com/en-us/power-apps/maker/data-platform/types-of-fields
 */
function mapDataverseTypeToTypeScript(dataverseType) {
    const lowerType = dataverseType.toLowerCase();
    switch (lowerType) {
        // Text-based types - all map to string
        case 'string':
        case 'stringtype':
        case 'text':
        case 'textarea':
        case 'multilinetext':
        case 'memotype':
        case 'memo':
        case 'email':
        case 'phone':
        case 'url':
        case 'tickersymbol':
        case 'entityname':
        case 'uniqueidentifier':
        case 'uniqueidentifiertype':
            return TypeScriptTypes.STRING;
        // Choice/Picklist types - return string (option labels)
        case 'picklist':
        case 'picklisttype':
        case 'choice':
        case 'choicetype':
        case 'choices':
        case 'choicestype':
        case 'multiselectpicklisttype':
        case 'state':
        case 'statetype':
        case 'status':
        case 'statustype':
            return TypeScriptTypes.STRING;
        // Number types
        case 'integer':
        case 'integertype':
        case 'wholenumber':
        case 'duration':
        case 'language':
        case 'timezone':
            return TypeScriptTypes.NUMBER;
        case 'bigint':
        case 'biginttype':
        case 'big':
        case 'timestamp':
            return TypeScriptTypes.NUMBER;
        case 'decimal':
        case 'decimaltype':
        case 'decimalnumber':
            return TypeScriptTypes.NUMBER;
        case 'double':
        case 'doubletype':
        case 'floatingpointnumber':
            return TypeScriptTypes.NUMBER;
        case 'money':
        case 'moneytype':
        case 'currency':
            return TypeScriptTypes.NUMBER;
        // Boolean type
        case 'boolean':
        case 'booleantype':
        case 'yesno':
        case 'twooptions':
            return TypeScriptTypes.BOOLEAN;
        // Date/Time types
        case 'datetime':
        case 'datetimetype':
        case 'dateonly':
            return TypeScriptTypes.DATE;
        // Lookup types - return string (typically GUID or display name)
        case 'lookup':
        case 'lookuptype':
        case 'customer':
        case 'customertype':
        case 'owner':
        case 'ownertype':
            return TypeScriptTypes.STRING;
        // Binary/File types - return string (base64 or URL)
        case 'file':
        case 'filetype':
        case 'image':
        case 'imagetype':
            return TypeScriptTypes.STRING;
        // Special types
        case 'virtual':
        case 'virtualtype':
        case 'formula':
        case 'formulatype':
            return TypeScriptTypes.STRING;
        default:
            return TypeScriptTypes.STRING; // Default Dataverse fields to string
    }
}
/*
 * The addPropertyComment method adds a comment for the property.
 */
function addPropertyComment(lines, propValue, padding) {
    if (propValue?.description && typeof propValue.description === 'string') {
        lines.push(`${padding}  // ${propValue.description}`);
    }
}
/*
 * The generateModelFromParameters method generates TypeScript models from the parameters.
 */
function generateModelFromParameters(modelName, parameterObject) {
    const lines = [];
    addInterfaceDeclaration(lines, modelName);
    const propName = getParameterName(parameterObject, 'name');
    const tsType = determineParameterType(parameterObject);
    const optionalMark = isParameterRequired(parameterObject) ? '' : '?';
    const propertyDefinition = `${propName}${optionalMark}: ${tsType}`;
    lines.push(`  ${propertyDefinition};`);
    closeInterfaceDeclaration(lines);
    const propertyInfo = new ParameterPropertyInfo(propName, propertyDefinition);
    return { modelCode: lines.join('\n'), propertyInfo };
}
/*
 * The addInterfaceDeclaration method adds the interface declaration to the lines.
 * It takes the lines to write to and model name as parameters.
 */
function addInterfaceDeclaration(lines, modelName) {
    lines.push(`export interface ${convertToValidIdentifier(modelName)} {`);
}
/*
 * The closeInterfaceDeclaration method closes the interface declaration.
 */
function closeInterfaceDeclaration(lines) {
    lines.push('}');
}
/*
 * The determineParameterType method determines the type of the parameter.
 * It takes the parameter object as input and returns the type as a string.
 */
function determineParameterType(parameterObject) {
    if (Array.isArray(parameterObject.enum) && parameterObject.enum.length > 0) {
        return parameterObject.enum.map((v) => formatEnumLabel(String(v))).join(' | ');
    }
    if ('type' in parameterObject && typeof parameterObject.type === 'string') {
        return mapJsonTypeToTypeScript(parameterObject.type);
    }
    return 'unknown';
}
/*
 * The isParameterRequired method checks if the parameter is required.
 * It takes the parameter object as input and returns true if required, otherwise false.
 */
function isParameterRequired(parameterObject) {
    return ('required' in parameterObject &&
        (parameterObject.required === true || parameterObject.required === 'true'));
}
/*
 * The generateServiceFromPaths method generates the service class from the paths node.
 */
function generateServiceFromPaths(pathsNode, parameterTypes, context, operationCallType = 'connectorOperation', entityModelMap) {
    const { modelName, dataSourceName } = context;
    const lines = [];
    addCopyrightNotice(lines);
    const isLogicFlow = isJsonObject(context.schema) &&
        equalsIgnoreCase(context.schema.id, LOGIC_FLOWS_API_ID);
    const imports = collectStandardImports(context);
    const operations = processPaths(pathsNode, modelName, dataSourceName, imports, parameterTypes, operationCallType, entityModelMap, isLogicFlow);
    addImports(lines, imports);
    addClassDefinition(lines, modelName, dataSourceName, operations);
    return lines.join('\n');
}
function getPathToDatasourceInfo(context) {
    const { schemaFolder, outputFolder, project } = context;
    const servicesDir = project.join(outputFolder, SERVICES_FOLDER);
    // path to dataSourcesInfo.ts
    const dataSourceInfoPath = project.join(schemaFolder, 'appschemas', 'dataSourcesInfo');
    let relativePath = path.relative(servicesDir, dataSourceInfoPath).replace(/\\/g, '/');
    if (!relativePath.startsWith('.') && !relativePath.startsWith('/')) {
        relativePath = `./${relativePath}`;
    }
    return relativePath;
}
/*
 * The collectStandardImports method collects the standard imports for the service class.
 * It returns a Set of import statements.
 */
function collectStandardImports(context) {
    const relativePath = getPathToDatasourceInfo(context);
    // Base imports common to all service types
    const imports = new Set([
        `import { dataSourcesInfo } from '${relativePath}';`,
        "import type { IOperationResult } from '@microsoft/power-apps/data';",
        "import { getClient } from '@microsoft/power-apps/data';",
    ]);
    return imports;
}
/*
 * The processPaths method processes the paths node of the Swagger schema.
 * It extracts the operations and generates the methods for the service class.
 * It takes the paths node, model name, data source name, and imports set as parameters.
 * It returns a list of operations as strings.
 */
function processPaths(pathsNode, modelName, dataSourceName, imports, parameterTypes, operationCallType = 'connectorOperation', entityModelMap, isLogicFlow) {
    const operations = [];
    for (const pathKey in pathsNode) {
        if (!Object.prototype.hasOwnProperty.call(pathsNode, pathKey)) {
            continue;
        }
        const pathObject = pathsNode[pathKey];
        if (isJsonObject(pathObject)) {
            for (const operationKey in pathObject) {
                if (!Object.prototype.hasOwnProperty.call(pathObject, operationKey)) {
                    continue;
                }
                const operationObject = pathObject[operationKey];
                if (isJsonObject(operationObject)) {
                    processOperation(operationObject, modelName, dataSourceName, imports, operations, parameterTypes, operationCallType, entityModelMap, isLogicFlow);
                }
            }
        }
    }
    return operations;
}
/*
 * The processOperation method processes each operation in the path object.
 * It extracts the operation ID, summary, description, and parameters.
 * It generates the method for the service class.
 */
function processOperation(operationObject, modelName, dataSourceName, imports, operations, parameterTypes, operationCallType = 'connectorOperation', entityModelMap, isLogicFlow) {
    if (!('operationId' in operationObject)) {
        return;
    }
    const operationId = String(operationObject.operationId);
    const summary = 'summary' in operationObject ? String(operationObject.summary) : '';
    const description = 'description' in operationObject ? String(operationObject.description) : '';
    const outputModel = determineOutputModel(operationObject, modelName, imports, entityModelMap);
    const { parametersList, parametersProps, originalParamEntries, internalDefaults } = collectParameters(operationObject, modelName, imports, parameterTypes, isLogicFlow, entityModelMap);
    // Collect @param JSDoc lines for parameters that carry an x-ms-csdl-type annotation.
    const paramCsdlAnnotations = [];
    if ('parameters' in operationObject && Array.isArray(operationObject.parameters)) {
        for (const param of operationObject.parameters) {
            if (!isJsonObject(param))
                continue;
            const csdlType = typeof param['x-ms-csdl-type'] === 'string' ? param['x-ms-csdl-type'] : '';
            const name = typeof param.name === 'string' ? param.name : '';
            if (csdlType && name) {
                const displayType = csdlType.replace(/^mscrm\./, '').replace(/^Edm\./, '');
                paramCsdlAnnotations.push(`@param ${name} type: \`${displayType}\``);
            }
        }
    }
    generateOperationMethod(operations, operationId, summary, description, parametersList, parametersProps, originalParamEntries, outputModel, modelName, operationCallType, paramCsdlAnnotations, internalDefaults);
}
/*
 * The determineOutputModel method determines the output model for the operation.
 * It takes the operation object, model name, and imports set as parameters.
 * It returns the output model as a string.
 */
function determineOutputModel(operationObject, modelName, imports, entityModelMap) {
    if ('responses' in operationObject && isJsonObject(operationObject.responses)) {
        const responsesObj = operationObject.responses;
        if ('200' in responsesObj) {
            const response200Value = responsesObj['200'];
            if (isJsonObject(response200Value)) {
                if ('schema' in response200Value && isJsonObject(response200Value.schema)) {
                    const schemaObj = response200Value.schema;
                    return extractOutputModelFromSchema(schemaObj, modelName, imports, entityModelMap);
                }
            }
        }
    }
    return 'void';
}
/*
 * The collectParameters method collects the parameters from the operation object.
 * It takes the operation object, model name, and imports set as parameters.
 * It returns an object containing the parameters list and parameters properties.
 */
function collectParameters(operationObject, modelName, imports, parameterTypes, isLogicFlow, entityModelMap) {
    let parametersList = [];
    let parametersProps = [];
    const originalParamEntries = [];
    const internalDefaults = [];
    if ('parameters' in operationObject && Array.isArray(operationObject.parameters)) {
        const paramsArray = operationObject.parameters;
        const orderedParams = getOrderedPropertiesByRequired(paramsArray);
        for (const property of orderedParams) {
            const param = property.value;
            // For logicflow schemas only: skip internal params that have a default value — hardcode their default in executeAsync instead
            if (isLogicFlow &&
                isJsonObject(param) &&
                param['x-ms-visibility'] === 'internal' &&
                'default' in param) {
                const paramName = getParameterName(param, 'name');
                internalDefaults.push({ name: paramName, value: param['default'] });
                continue;
            }
            // property.Value in C# is property.value in TS - use property.value for the parameter object
            const result = processRequestParameters(param, modelName, imports, parametersList, parametersProps, 'name', parameterTypes, originalParamEntries, entityModelMap);
            parametersList = result[0];
            parametersProps = result[1];
        }
    }
    return { parametersList, parametersProps, originalParamEntries, internalDefaults };
}
/*
 * The addImports method adds the import statements to the lines.
 * It takes the lines to write to and the imports set as parameters.
 */
function addImports(lines, imports) {
    const sortedImports = Array.from(imports);
    sortedImports.sort();
    lines.push(...sortedImports);
}
/*
 * The addClassDefinition method adds the class definition to the lines.
 * It takes the lines to write to, model name, data source name, and operations as parameters.
 */
function addClassDefinition(lines, modelName, dataSourceName, operations) {
    addServiceClassHeader(lines, modelName, dataSourceName);
    lines.push(...operations);
    lines.push('}');
    lines.push('');
}
/*
 * The addServiceClassHeader method adds the service class header to the lines.
 * It takes the lines to write to, model name, and data source name as parameters.
 */
function addServiceClassHeader(lines, modelName, dataSourceName, additionalLine = true) {
    if (additionalLine) {
        lines.push('');
    }
    lines.push(`export class ${modelName}Service {`);
    addServiceFields(lines, dataSourceName);
}
/*
 * The extractOutputModelFromSchema method extracts the output model from the schema object.
 */
function extractOutputModelFromSchema(schemaObj, modelName, imports, entityModelMap) {
    // Resolve Dataverse entity refs (x-ms-csdl-type: mscrm.*) before generic type extraction.
    const entityType = resolveEntityTypeFromCsdl(schemaObj, imports, entityModelMap);
    if (entityType)
        return entityType;
    let outputModel = '';
    if (tryExtractArrayOutputModel(schemaObj, modelName, imports, (result) => {
        outputModel = result;
    })) {
        return outputModel;
    }
    if (tryExtractDirectRefOutputModel(schemaObj, modelName, imports, (result) => {
        outputModel = result;
    })) {
        return outputModel;
    }
    const [found, typeOutputModel] = tryExtractTypeOutputModel(schemaObj);
    if (found) {
        outputModel = typeOutputModel;
        return outputModel;
    }
    return 'void'; // Default to void if no specific type is defined
}
/*
 * The tryExtractArrayOutputModel method extracts the array output model from the schema object.
 * Returns true if type is successfully extracted, otherwise false.
 */
function tryExtractArrayOutputModel(schemaObj, modelName, imports, setOutputModel) {
    // First case: properties.value.type === 'array' && properties.value.items.$ref
    if ('properties' in schemaObj && isJsonObject(schemaObj.properties)) {
        const propertiesObj = schemaObj.properties;
        if ('value' in propertiesObj && isJsonObject(propertiesObj.value)) {
            const valueObj = propertiesObj.value;
            if ('type' in valueObj &&
                valueObj.type === 'array' &&
                'items' in valueObj &&
                isJsonObject(valueObj.items)) {
                const itemsObj = valueObj.items;
                if ('$ref' in itemsObj && typeof itemsObj.$ref === 'string') {
                    const refType = extractReferencedType(itemsObj.$ref);
                    setOutputModel(`{ value: ${refType}[] }`);
                    addImport(refType, modelName, imports);
                    return true;
                }
            }
        }
    }
    // Second case: type === 'array' && items.$ref
    if ('type' in schemaObj &&
        schemaObj.type === 'array' &&
        'items' in schemaObj &&
        isJsonObject(schemaObj.items)) {
        const itemsObj = schemaObj.items;
        if ('$ref' in itemsObj && typeof itemsObj.$ref === 'string') {
            const refType = extractReferencedType(itemsObj.$ref);
            setOutputModel(`${refType}[]`);
            addImport(refType, modelName, imports);
            return true;
        }
    }
    return false;
}
/*
 * The tryExtractDirectRefOutputModel method extracts the direct reference output model from the schema object.
 * Returns true if type is successfully extracted, otherwise false.
 */
function tryExtractDirectRefOutputModel(schemaObj, modelName, imports, setOutputModel) {
    // Case: items.$ref
    if ('items' in schemaObj && isJsonObject(schemaObj.items)) {
        const itemsObj = schemaObj.items;
        if ('$ref' in itemsObj && typeof itemsObj.$ref === 'string') {
            const outputModel = simplifyGenericTypeName(extractReferencedType(itemsObj.$ref));
            setOutputModel(outputModel);
            addImport(outputModel, modelName, imports);
            return true;
        }
    }
    // Case: $ref
    if ('$ref' in schemaObj && typeof schemaObj.$ref === 'string') {
        const outputModel = simplifyGenericTypeName(extractReferencedType(schemaObj.$ref));
        setOutputModel(outputModel);
        addImport(outputModel, modelName, imports);
        return true;
    }
    return false;
}
/*
 * The tryExtractTypeOutputModel method extracts the type output model from the schema object.
 * Returns a tuple: [found: boolean, outputModel: string].
 */
function tryExtractTypeOutputModel(schemaObj) {
    if ('type' in schemaObj && typeof schemaObj.type === 'string') {
        return [true, mapJsonTypeToTypeScript(schemaObj.type)];
    }
    return [false, ''];
}
/*
 * The addImport method adds an import statement to the imports set.
 * Takes the type name, model name, and imports set as parameters.
 */
function addImport(typeName, modelName, imports) {
    imports.add(`import type { ${typeName} } from '${RELATIVE_MODEL_PATH}/${modelName}Model';`);
}
/*
 * The simplifyGenericTypeName method simplifies the generic type name by removing the generic parameters.
 */
function simplifyGenericTypeName(typeName) {
    return normalizeGenericTypeName(typeName);
}
/*
 * The processRequestParameters method processes the parameters of the operation.
 * It extracts the parameter name and type from the schema and adds them to the lists.
 * It takes the parameter object, model name, imports set, parameters list, and parameters properties as parameters.
 */
function processRequestParameters(param, modelName, imports, parametersList, parametersProps, paramNodeName, parameterTypes, originalParamEntries, entityModelMap) {
    if (!isJsonObject(param)) {
        return [parametersList, parametersProps];
    }
    const paramName = getParameterName(param, paramNodeName);
    // Skip 'connectionId' parameter
    if (paramName.toLowerCase() === 'connectionid') {
        return [parametersList, parametersProps];
    }
    if (parameterTypes && parameterTypes[paramName]) {
        parametersList.push(`${parameterTypes[paramName].propertyDefinition}`);
        parametersProps.push(`${parameterTypes[paramName].propertyName}`);
        originalParamEntries?.push(`${parameterTypes[paramName].propertyDefinition}`);
    }
    else {
        // Only add to imports if not an override
        const paramType = getParameterType(param, modelName, imports, entityModelMap);
        const optionalMark = isParameterRequired(param) ? '' : '?';
        // Normalize parameter name (hyphens/dots → underscores for valid TS identifier)
        const normalizedParamName = paramName.replace(/[-.]/g, '_');
        parametersList.push(`${normalizedParamName}${optionalMark}: ${paramType}`);
        // When the name was hyphenated, emit an explicit key mapping in the object literal
        // e.g. "x-ms-conversation-id": x_ms_conversation_id
        const paramProp = normalizedParamName !== paramName
            ? `${formatPropertyName(paramName)}: ${normalizedParamName}`
            : normalizedParamName;
        parametersProps.push(paramProp);
        originalParamEntries?.push(`${formatPropertyName(paramName)}${optionalMark}: ${paramType}`);
    }
    return [parametersList, parametersProps];
}
/*
 * The getParameterName method extracts the name of the parameter from the schema.
 * It handles the case where the schema is an array or a reference.
 * It takes the parameter object and node name as parameters.
 */
function getParameterName(paramObj, paramNodeName) {
    const refNode = paramObj.$ref;
    if (refNode && typeof refNode === 'string') {
        const refValue = refNode;
        if (refValue.toLowerCase().startsWith('#/parameters/')) {
            const parameterName = refValue.substring('#/parameters/'.length);
            return parameterName;
        }
    }
    const nameNode = paramObj[paramNodeName];
    return nameNode ? String(nameNode) : 'param';
}
/** Capitalizes the first character of a logical name to form an interface prefix. */
function toInterfaceName(logicalName) {
    return logicalName.charAt(0).toUpperCase() + logicalName.slice(1);
}
/**
 * Resolves `x-ms-csdl-type` to a TypeScript entity interface type when the value is a Dataverse
 * entity reference (`mscrm.*`) or entity collection (`Collection(mscrm.*)`). Returns `null` when
 * the field is absent, not an entity reference, or the logical name is not present in
 * `entityModelMap` (meaning no model file can be guaranteed).
 *
 * `entityModelMap` is built by `buildEntityModelMap`, which guarantees that every entry is backed
 * by an existing or newly-generated model file. Callers that do not pass the map (e.g. non-Dataverse
 * swagger schemas) receive `null` and fall through to the standard JSON type mapping.
 *
 * Adds the required `import type { …Base }` statement to `imports`.
 */
function resolveEntityTypeFromCsdl(paramObj, imports, entityModelMap) {
    const csdlType = paramObj['x-ms-csdl-type'];
    if (typeof csdlType !== 'string')
        return null;
    const collectionMatch = csdlType.match(/^Collection\(mscrm\.(\w+)\)$/);
    if (collectionMatch) {
        const modelName = entityModelMap?.get(collectionMatch[1]);
        if (!modelName)
            return null;
        imports.add(`import type { ${modelName}Base } from '${RELATIVE_MODEL_PATH}/${modelName}Model';`);
        return `${modelName}Base[]`;
    }
    const entityMatch = csdlType.match(/^mscrm\.(\w+)$/);
    if (entityMatch) {
        const modelName = entityModelMap?.get(entityMatch[1]);
        if (!modelName)
            return null;
        imports.add(`import type { ${modelName}Base } from '${RELATIVE_MODEL_PATH}/${modelName}Model';`);
        return `${modelName}Base`;
    }
    return null;
}
/*
 * The getParameterType method extracts the type of the parameter from the schema.
 * It handles the case where the schema is an array or a reference.
 * It takes the parameter object, model name, and imports set as parameters.
 */
function getParameterType(paramObj, modelName, imports, entityModelMap) {
    // Resolve Dataverse entity refs (x-ms-csdl-type: mscrm.*) before falling back to JSON type.
    const entityType = resolveEntityTypeFromCsdl(paramObj, imports, entityModelMap);
    if (entityType)
        return entityType;
    if ('schema' in paramObj && isJsonObject(paramObj.schema)) {
        return getSchemaBasedType(paramObj.schema, modelName, imports);
    }
    if ('$ref' in paramObj && typeof paramObj.$ref === 'string') {
        return getRefBasedType(paramObj.$ref, modelName, imports);
    }
    if (Array.isArray(paramObj.enum) && paramObj.enum.length > 0) {
        return paramObj.enum.map((v) => formatEnumLabel(String(v))).join(' | ');
    }
    if ('type' in paramObj && typeof paramObj.type === 'string') {
        return mapJsonTypeToTypeScript(paramObj.type);
    }
    return 'unknown';
}
/*
 * The getSchemaBasedType method extracts the type from the schema object.
 * It handles the case where the schema is an array or a reference.
 * It takes the schema object, model name, and imports set as parameters.
 */
function getSchemaBasedType(schemaObj, modelName, imports) {
    if ('$ref' in schemaObj && typeof schemaObj.$ref === 'string') {
        return getRefBasedType(schemaObj.$ref, modelName, imports);
    }
    if ('type' in schemaObj && typeof schemaObj.type === 'string') {
        const typeValue = schemaObj.type;
        // Handle binary format as string
        if ('format' in schemaObj && schemaObj.format === 'binary' && typeValue === 'string') {
            return 'string';
        }
        return mapJsonTypeToTypeScript(typeValue);
    }
    return 'unknown';
}
/*
 * The getRefBasedType method extracts the referenced type from the ref value.
 * It adds the import statement for the referenced type to the imports set.
 * It takes the ref value, model name, and imports set as parameters.
 */
function getRefBasedType(refValue, modelName, imports) {
    const refType = extractReferencedType(refValue);
    addImport(refType, modelName, imports);
    return refType;
}
/*
 * The generateOperationMethod method generates a method for the service class.
 * It takes the operations list, operation ID, summary, description,
 * parameters list, parameters properties, output model, and data source name as parameters.
 */
function generateOperationMethod(operations, operationId, summary, description, parametersList, parametersProps, originalParamEntries, outputModel, modelName, operationCallType = 'connectorOperation', paramCsdlAnnotations = [], internalDefaults = []) {
    addMethodSummary(operations, summary, description, paramCsdlAnnotations);
    const operationFuncName = sanitizeName(operationId);
    const parameterSignature = getParameterSignature(parametersList);
    const parameterObject = getParameterObject(parametersProps);
    const parametersType = getParametersType(parametersList);
    const originalParamsType = getParametersType(originalParamEntries);
    if (parametersList.length === 0) {
        addMethodWithoutParameters(operations, operationFuncName, outputModel, operationId, modelName, operationCallType, internalDefaults);
    }
    else {
        addMethodWithParameters(operations, operationFuncName, parameterSignature, parametersType, originalParamsType, parameterObject, outputModel, operationId, modelName, operationCallType, internalDefaults);
    }
}
/*
 * The addMethodSummary method adds a summary and description to the method.
 * It takes the operations list, summary, and description as parameters.
 */
function addMethodSummary(operations, summary, description, paramCsdlAnnotations = []) {
    const hasContent = (summary && summary.trim().length > 0) || paramCsdlAnnotations.length > 0;
    if (hasContent) {
        operations.push('');
        operations.push('  /**');
        if (summary && summary.trim().length > 0) {
            operations.push(`   * ${summary}`);
            if (description && description.trim().length > 0) {
                operations.push(`   * ${description}`);
            }
        }
        for (const annotation of paramCsdlAnnotations) {
            operations.push(`   * ${annotation}`);
        }
        operations.push('   */');
    }
    else {
        operations.push('');
    }
}
/*
 * The getParameterSignature method generates the parameter signature for the service class.
 * It takes the parameters list as input and returns the parameter signature as a string.
 */
function getParameterSignature(parametersList) {
    return parametersList.length > 0 ? parametersList.join(', ') : '';
}
/*
 * The getParameterObject method generates the parameter object for the service class.
 * It takes the parameters properties list as input and returns the parameter object as a string.
 */
function getParameterObject(parametersProps) {
    return parametersProps.length > 0 ? parametersProps.join(', ') : '';
}
/*
 * The getParametersType method generates the parameters type for the service class.
 * It takes the parameters list as input and returns the parameters type as a string.
 */
function getParametersType(parametersList) {
    return parametersList.length > 0 ? `{ ${parametersList.join(', ')} }` : 'object';
}
/*
 * The buildInternalDefaultsLiteral method generates a string of key-value pairs
 * for parameters that are internal with default values, to be inlined into executeAsync.
 */
function buildInternalDefaultsLiteral(internalDefaults) {
    return internalDefaults
        .map(({ name, value }) => `${formatPropertyName(name)}: ${JSON.stringify(value)}`)
        .join(', ');
}
/*
 * The addMethodWithoutParameters method generates a method without parameters for the service class.
 * It takes the operations list, operation function name, output model, data source name,
 * and operation ID as parameters.
 */
function addMethodWithoutParameters(operations, operationFuncName, outputModel, operationId, modelName, operationCallType = 'connectorOperation', internalDefaults = []) {
    if (operationCallType === 'dataverseOperation') {
        operations.push(`  public static async ${operationFuncName}(): Promise<IOperationResult<${outputModel}>> {`, `    const result = await ${modelName}Service.client.executeAsync<void, ${outputModel}>(`, `      {`, `        dataverseRequest: {`, `          action: 'customapi',`, `          parameters: {`, `            operationName: '${operationId}',`, `            tableName: ${getServiceName(modelName)},`, `          }`, `        },`, `      });`, `    return result;`, `  }`);
    }
    else {
        const parametersLine = internalDefaults.length > 0
            ? [`          parameters: { ${buildInternalDefaultsLiteral(internalDefaults)} },`]
            : [];
        operations.push(`  public static async ${operationFuncName}(): Promise<IOperationResult<${outputModel}>> {`, `    const result = await ${modelName}Service.client.executeAsync<void, ${outputModel}>(`, `      {`, `        ${operationCallType}: {`, `          tableName: ${getServiceName(modelName)},`, `          operationName: '${operationId}',`, ...parametersLine, `        },`, `      });`, `    return result;`, `  }`);
    }
}
/*
 * The addMethodWithParameters method generates a method with parameters for the service class.
 * It takes the operations list, operation function name, parameter signature, parameters type,
 * parameter object, output model, data source name, and operation ID as parameters.
 */
function addMethodWithParameters(operations, operationFuncName, parameterSignature, parametersType, originalParamsType, parameterObject, outputModel, operationId, modelName, operationCallType = 'connectorOperation', internalDefaults = []) {
    if (operationCallType === 'dataverseOperation') {
        operations.push(`  public static async ${operationFuncName}(${parameterSignature}): Promise<IOperationResult<${outputModel}>> {`, `    const params: ${originalParamsType} = { ${parameterObject} };`, `    const result = await ${modelName}Service.client.executeAsync<${parametersType}, ${outputModel}>(`, `      {`, `        dataverseRequest: {`, `          action: 'customapi',`, `          parameters: {`, `            operationName: '${operationId}',`, `            tableName: ${getServiceName(modelName)},`, `            body: params`, `          }`, `        },`, `      });`, `    return result;`, `  }`);
    }
    else {
        const allParamsLine = internalDefaults.length > 0
            ? [
                `    const allParams = { ...params, ${buildInternalDefaultsLiteral(internalDefaults)} };`,
            ]
            : [];
        const parametersValue = internalDefaults.length > 0 ? 'allParams' : 'params';
        operations.push(`  public static async ${operationFuncName}(${parameterSignature}): Promise<IOperationResult<${outputModel}>> {`, `    const params: ${originalParamsType} = { ${parameterObject} };`, ...allParamsLine, `    const result = await ${modelName}Service.client.executeAsync<${originalParamsType}, ${outputModel}>(`, `      {`, `        ${operationCallType}: {`, `          tableName: ${getServiceName(modelName)},`, `          operationName: '${operationId}',`, `          parameters: ${parametersValue}`, `        },`, `      });`, `    return result;`, `  }`);
    }
}
/*
 * The handleSqlStoredProcedure method handles the SQL stored procedure generation.
 */
async function handleSqlStoredProcedure(context) {
    const { modelName, schema, outputFolder, project } = context;
    const { modelCode, responseModelName } = generateStoredProcedureModel(modelName, schema);
    const serviceCode = generateStoredProcedureService(responseModelName, context);
    await writeModelFile(outputFolder, modelName, modelCode, project);
    await writeServiceFile(outputFolder, modelName, serviceCode, project);
}
/*
 * The generateStoredProcedureModel method generates the model code for the stored procedure.
 * It takes the model name and schema as parameters and returns the generated model code.
 */
function generateStoredProcedureModel(modelName, schema) {
    const lines = [];
    addCopyrightNotice(lines);
    // Validate schema
    if (!isJsonObject(schema) || !('schema' in schema)) {
        throw new Error('Invalid schema format for stored procedure.');
    }
    const schemaNode = schema.schema;
    // Ensure schemaNode is a JsonObject
    if (!isJsonObject(schemaNode)) {
        throw new Error('Invalid schema node format for stored procedure.');
    }
    addInputParametersModel(lines, modelName, schemaNode);
    const responseModelName = addOutputParametersModel(lines, modelName, schemaNode);
    return { modelCode: lines.join('\n'), responseModelName };
}
/*
 * The addInputParametersModel method generates the input parameters model for the stored procedure.
 * It takes the lines to write to, model name, and schema node as parameters.
 */
function addInputParametersModel(lines, modelName, schemaNode) {
    if ('inputparameters' in schemaNode && isJsonObject(schemaNode.inputparameters)) {
        const requestModelName = `${modelName}Request`;
        const modelCode = generateModelFromSchema(requestModelName, schemaNode.inputparameters);
        lines.push(modelCode);
        lines.push(''); // Add a blank line between models
    }
}
/*
 * The addOutputParametersModel method generates the output parameters model for the stored procedure.
 * It takes the lines to write to, model name, and schema node as parameters.
 */
function addOutputParametersModel(lines, modelName, schemaNode) {
    if ('procedureresultschema' in schemaNode && isJsonObject(schemaNode.procedureresultschema)) {
        const responseModelName = `${modelName}Response`;
        const modelCode = generateModelFromSchema(responseModelName, schemaNode.procedureresultschema);
        lines.push(modelCode);
        lines.push(''); // Add a blank line between models
        return responseModelName;
    }
    return '';
}
/*
 * The generateStoredProcedureService method generates the service code for the stored procedure.
 * It takes the response model name and context as parameters.
 */
function generateStoredProcedureService(responseModelName, context) {
    const { modelName, dataSourceName, schema } = context;
    const lines = [];
    addCopyrightNotice(lines);
    addServiceImports(lines, responseModelName, context);
    addStoredProcedureServiceClassDefinition(lines, modelName, dataSourceName, schema, responseModelName);
    return lines.join('\n');
}
/*
 * The addServiceImports method adds necessary imports to the service file.
 * It takes the lines to write to, response model name, and context as parameters.
 */
function addServiceImports(lines, responseModelName, context) {
    const { modelName } = context;
    const standardImports = collectStandardImports(context);
    lines.push(...standardImports); // Add standard imports
    lines.push(`import type { ${modelName}Request, ${responseModelName} } from '${RELATIVE_MODEL_PATH}/${modelName}Model';`);
    lines.push('');
}
/*
 * The addServiceClassDefinition method generates the service class definition.
 * It takes the lines to write to, model name, data source name, schema, and response model name as parameters.
 */
function addStoredProcedureServiceClassDefinition(lines, modelName, dataSourceName, schema, responseModelName) {
    addServiceClassHeader(lines, modelName, dataSourceName, false);
    addStoredProcedureServiceMethods(lines, modelName, schema, responseModelName);
    lines.push('}');
    lines.push('');
}
/*
 * The addStoredProcedureServiceMethods method generates service methods for the stored procedure.
 * It takes the lines to write to, model name, schema, and response model name as parameters.
 */
function addStoredProcedureServiceMethods(lines, modelName, schema, responseModelName) {
    const requestModelName = `${modelName}Request`;
    const parametersList = [`inputParameters: ${requestModelName}`];
    const parametersProps = ['inputParameters'];
    // Derive operationId from the raw schema name to preserve original casing.
    // modelName has been capitalized for use as a TypeScript identifier, but operationName
    // must match the key stored in datasourcesinfo exactly (e.g. 'testProc', not 'TestProc').
    const operationId = isJsonObject(schema)
        ? getSqlStoredProcApiKey(getNameNode(schema))
        : modelName;
    generateOperationMethod(lines, operationId, '', '', parametersList, parametersProps, parametersList, responseModelName, modelName);
}
/*
 * The getOrderedPropertiesByRequired method sorts the properties by required property and original index.
 */
function getOrderedPropertiesByRequired(propertiesOrArray) {
    let propertiesObj;
    if (Array.isArray(propertiesOrArray)) {
        propertiesObj = {};
        for (let idx = 0; idx < propertiesOrArray.length; idx++) {
            propertiesObj[`param_${idx}`] = propertiesOrArray[idx];
        }
    }
    else if (isJsonObject(propertiesOrArray)) {
        propertiesObj = propertiesOrArray;
    }
    else {
        throw new Error('Input must be an object or array');
    }
    const propertyList = [];
    let index = 0;
    for (const key in propertiesObj) {
        if (!Object.prototype.hasOwnProperty.call(propertiesObj, key)) {
            continue;
        }
        const value = propertiesObj[key];
        let required = false;
        if (isJsonObject(value) && 'required' in value) {
            required = String(value.required) === 'true';
        }
        propertyList.push({ key, value, required, index });
        index++;
    }
    // Required first, then original order
    propertyList.sort((a, b) => {
        const reqCompare = Number(b.required) - Number(a.required);
        if (reqCompare !== 0) {
            return reqCompare;
        }
        return a.index - b.index;
    });
    return propertyList.map((p) => ({ key: p.key, value: p.value }));
}
async function writeModelFile(outputDir, modelName, modelCode, project) {
    const modelFilePath = project.join(outputDir, MODELS_FOLDER, `${modelName}Model.ts`);
    await project.createFile('model', modelFilePath, modelCode);
}
async function writeServiceFile(outputDir, modelName, serviceCode, project) {
    const serviceFilePath = project.join(outputDir, SERVICES_FOLDER, `${modelName}Service.ts`);
    await project.createFile('service', serviceFilePath, serviceCode);
}
/*
 * The handleStandardSchema method generates TypeScript models and services from any schema type.
 * It routes to the appropriate handler based on schema type.
 */
async function handleStandardSchema(context) {
    const { schema, outputFolder, project } = context;
    await ensureCommonModelFileExists(outputFolder, project);
    // Check if this is a Dataverse entity schema (has nested schema.items structure)
    if (isDataverseEntity(schema)) {
        await handleDataverseSchema(context);
    }
    else {
        await handleStandardSchemaOnly(context);
    }
}
/*
 * The handleDataverseSchema method generates TypeScript models and services specifically for Dataverse entities.
 */
async function handleDataverseSchema(context) {
    const { modelName, schema, outputFolder, project } = context;
    if (!isJsonObject(schema) || !isJsonObject(schema.schema)) {
        throw new Error(`Invalid Dataverse entity schema structure: missing 'schema' property for ${modelName}`);
    }
    const schemaProperty = schema.schema;
    if (!isJsonObject(schemaProperty.items)) {
        throw new Error(`Invalid Dataverse entity schema structure: missing 'schema.items' property for ${modelName}`);
    }
    const itemsNode = schemaProperty.items;
    let primaryKey = 'id'; // default
    // Extract primary key from Dataverse entity schema inline
    const properties = itemsNode.properties;
    if (properties) {
        for (const [propName, propValue] of Object.entries(properties)) {
            if (isJsonObject(propValue)) {
                if (propValue['x-ms-dataverse-primary-id'] === true ||
                    propValue['x-ms-keyType'] === 'primary') {
                    primaryKey = propName;
                    break;
                }
            }
        }
    }
    // For Dataverse entities, generate both base and extended interfaces for lookup columns
    const propertiesNode = isJsonObject(properties) ? properties : {};
    const fileColumnNames = getFileColumnNames(propertiesNode);
    const imageColumnNames = getImageColumnNames(propertiesNode);
    if (isCodegenV2Context(context)) {
        const { hasLookupColumns, multiSelectPicklistFields } = await generateDataverseModelV2({
            parentContext: context,
            schema: itemsNode,
            fileColumnNames,
            imageColumnNames,
            modelFilePath: project.join(outputFolder, MODELS_FOLDER, `${modelName}Model.ts`),
        });
        await generateDataverseServiceV2({
            parentContext: context,
            primaryKey,
            hasLookupColumns,
            multiSelectPicklistFields,
            fileColumnNames,
            imageColumnNames,
            serviceFilePath: project.join(outputFolder, SERVICES_FOLDER, `${modelName}Service.ts`),
            dataSourceInfoPath: getPathToDatasourceInfo(context),
        });
    }
    else {
        const { baseInterface, extendedInterface, multiSelectPicklistFields } = generateDataverseInterfaces(modelName, itemsNode, fileColumnNames, imageColumnNames);
        const hasLookupColumns = !!extendedInterface;
        if (hasLookupColumns) {
            // Generate both interfaces when there are lookup columns
            const combinedModelCode = baseInterface + '\n\n' + extendedInterface;
            await writeModelFile(outputFolder, modelName, combinedModelCode, project);
        }
        else {
            // Generate single interface when no lookup columns
            await writeModelFile(outputFolder, modelName, baseInterface, project);
        }
        const serviceCode = generateDataverseTypeScriptService(primaryKey, hasLookupColumns, multiSelectPicklistFields, fileColumnNames, imageColumnNames, context);
        await writeServiceFile(outputFolder, modelName, serviceCode, project);
    }
}
/*
 * The handleStandardSchemaOnly method generates TypeScript models and services for standard (non-Dataverse) schemas.
 */
async function handleStandardSchemaOnly(context) {
    const { modelName, outputFolder, schema, project } = context;
    const { modelCode, primaryKey } = generateTypeScriptModel(schema, modelName);
    const serviceCode = generateTypeScriptService(primaryKey, false, context);
    await writeModelFile(outputFolder, modelName, modelCode, project);
    await writeServiceFile(outputFolder, modelName, serviceCode, project);
}
/**
 * The handleSharepointSchema method handles SharePoint schema generation.
 * It generates TypeScript models and services specifically for SharePoint data sources.
 */
async function handleSharepointSchema(context) {
    const { modelName, outputFolder, schema, project } = context;
    await ensureCommonModelFileExists(outputFolder, project);
    // SharePoint-specific processing logic
    const { modelCode, primaryKey, hasReferenceEntities } = generateSharepointTypeScriptModel(schema, modelName);
    // Generate SharePoint-specific service code
    const serviceCode = generateSharepointService(primaryKey, hasReferenceEntities, context);
    await writeModelFile(outputFolder, modelName, modelCode, project);
    await writeServiceFile(outputFolder, modelName, serviceCode, project);
}
/**
 * Generates SharePoint-specific service code.
 */
function generateSharepointService(primaryKey, hasReferenceEntities, context) {
    const { modelName, dataSourceName, schema } = context;
    const lines = [];
    addCopyrightNotice(lines);
    addSharepointServiceImports(lines, hasReferenceEntities, context);
    addSharepointServiceClassDefinition(lines, modelName, dataSourceName, schema, primaryKey, hasReferenceEntities);
    return lines.join('\n');
}
/**
 * Adds the SharePoint service class definition to the generated code.
 */
function addSharepointServiceClassDefinition(lines, modelName, dataSourceName, schema, primaryKey, hasReferenceEntities) {
    addServiceClassHeader(lines, modelName, dataSourceName, true);
    // Add CRUD methods
    const methods = [];
    methods.push(generateSharepointCreateMethod(modelName, primaryKey, hasReferenceEntities));
    methods.push(generateSharepointUpdateMethod(modelName, primaryKey, hasReferenceEntities));
    methods.push(generateSharepointDeleteMethod(modelName));
    methods.push(generateSharepointGetMethod(modelName, hasReferenceEntities));
    methods.push(generateSharepointGetAllMethod(modelName, hasReferenceEntities));
    for (const method of methods) {
        lines.push('');
        lines.push(...method);
    }
    // Add SharePoint-specific methods
    addGetMethodForReferenceEntities(lines, modelName, dataSourceName, schema);
    lines.push('}');
    lines.push('');
}
/**
 * Adds the SharePoint-specific get method for reference entities.
 */
function addGetMethodForReferenceEntities(lines, modelName, dataSourceName, schema) {
    if (isJsonObject(schema) && isJsonObject(schema.referencedEntities)) {
        addGetMethodForReferenceEntitiesCore(lines, modelName);
    }
}
/*
 * The get method adds get function for sharepoint referenceEntities
 */
function addGetMethodForReferenceEntitiesCore(lines, modelName) {
    // Add method summary
    lines.push(`/**`, `* Utility method to get possible values of a referenced entity by its name.`, `*/`, `  public static async getReferencedEntity(search: string, referencedEntityKeyNormalized: string): Promise<IOperationResult<unknown>> {`, `    const params: {search: string} = { search };`, `    const result = await ${modelName}Service.client.executeAsync<{search: string}, unknown>(`, `      {`, `        connectorOperation: {`, `          tableName: ${getServiceName(modelName)},`, `          operationName: 'Get' + referencedEntityKeyNormalized,`, `          parameters: params`, `        }`, `      });`, `    return result;`, `  }`, ``);
}
async function ensureCommonModelFileExists(outputDir, project) {
    // Node.js built-in modules are statically imported at the top of the file
    const commonModelPath = project.join(outputDir, MODELS_FOLDER, 'CommonModels.ts');
    if (!(await project.fileExists(commonModelPath))) {
        const lines = [];
        addCopyrightNotice(lines);
        addCommonModelCode(lines);
        await project.createFile('commonModel', commonModelPath, lines.join('\n'));
    }
}
/*
 * The addCommonModelCode method generates common TypeScript model code for the CommonModels.ts file.
 * It includes interfaces for IGetOptions and IGetAllOptions.
 */
function addCommonModelCode(lines) {
    lines.push('export interface IGetOptions {');
    lines.push('  select?: string[];');
    lines.push('};');
    lines.push('');
    lines.push('export interface IGetAllOptions {');
    lines.push('  maxPageSize?: number;');
    lines.push('  select?: string[];');
    lines.push('  filter?: string;');
    lines.push('  orderBy?: string[];');
    lines.push('  top?: number;');
    lines.push('  skip?: number;');
    lines.push('  skipToken?: string;');
    lines.push('}');
    lines.push('');
}
/*
 * The generateTypeScriptModel method generates a TypeScript model from the JSON schema.
 * It returns the generated model code and the primary key of the model.
 */
function generateTypeScriptModel(schema, modelName) {
    const root = getItemsFromSchema(schema);
    const propertiesNode = getPropertiesNode(root);
    const requiredFields = getRequiredFieldsFromStandardSchema(root);
    const lines = [];
    addCopyrightNotice(lines);
    const primaryKeyObj = { value: '' };
    addModelInterfaceDeclaration(lines, modelName, propertiesNode, requiredFields, primaryKeyObj);
    return { modelCode: lines.join('\n'), primaryKey: primaryKeyObj.value };
}
function generateSharepointTypeScriptModel(schema, modelName) {
    const root = getItemsFromSchema(schema);
    const propertiesNode = getPropertiesNode(root);
    const requiredFields = getRequiredFieldsFromStandardSchema(root);
    const lines = [];
    addCopyrightNotice(lines);
    const primaryKeyObj = { value: '' };
    const { hasReferenceEntities } = addSPModelInterfaceDeclaration(lines, modelName, propertiesNode, requiredFields, primaryKeyObj, schema);
    return { modelCode: lines.join('\n'), primaryKey: primaryKeyObj.value, hasReferenceEntities };
}
/*
 * The validateSchemaJson method validates the JSON schema and retrieves the items node.
 * It throws an exception if the schema format is invalid.
 */
function getItemsFromSchema(root) {
    if (!isJsonObject(root) || !isJsonObject(root.schema) || !isJsonObject(root.schema.items)) {
        throw new Error('Schema items must be a valid object');
    }
    return root.schema.items;
}
/*
 * The getRequiredFieldsFromStandardSchema method retrieves the required fields from the JSON schema.
 * It returns a Set of required field names.
 */
function getRequiredFieldsFromStandardSchema(itemsNode) {
    if (!isJsonObject(itemsNode)) {
        return new Set();
    }
    if (!itemsNode.required || !Array.isArray(itemsNode.required)) {
        return new Set();
    }
    return new Set(itemsNode.required.map((r) => String(r)));
}
/*
 * The addSPModelInterfaceDeclaration method generates the TypeScript interface declaration for the model particular to Sharepoint.
 */
function addSPModelInterfaceDeclaration(lines, modelName, propertiesNode, requiredFields, primaryKeyObj, schema) {
    const result = addSPReferencedEntityModelInterfaceDeclaration(schema, lines);
    const referenceEntitiesObjectMapping = result.mapping;
    // lines are already updated by the method above
    // We will generate 3 interfaces:
    // A base one with properties that are not reference entities
    // A "read" model that extends the base one and adds reference entity properties with an object type (inferred from the referencedEntities section)
    // A "write" model that extends the base one and adds reference entities with the string/string[] type
    // All 3 interfaces will be exported
    const baseInterfaceLines = [];
    const readInterfaceLines = [];
    const writeInterfaceLines = [];
    // This will be true if there is at least one referenced entity
    // This flag is later used to determine if we need to use multiple interfaces in the service file
    let hasReferenceEntities = false;
    baseInterfaceLines.push(`export interface ${modelName}Base {`);
    readInterfaceLines.push(`export interface ${modelName}Read extends ${modelName}Base {`);
    writeInterfaceLines.push(`export interface ${modelName}Write extends ${modelName}Base {`);
    // lines.push(`export interface ${modelName} {`);
    for (const propName of Object.keys(propertiesNode)) {
        let propNameNormalized = propName;
        if (!isValidPropertyName(propName)) {
            propNameNormalized = `"${propName}"`; // Enclose invalid property names in double quotes
        }
        const isRequired = requiredFields.has(propName);
        const optionalMark = isRequired ? '' : '?';
        if (propName in referenceEntitiesObjectMapping) {
            if (isSharepointChoiceColumn(propertiesNode[propName])) {
                // Note: SharePoint choice/multi-choice columns are intentionally mapped to string/string[]
                // instead of the schema's { Value: string } object. This simplifies the SDK's public
                // surface and prevents recurring errors in AI agents (especially App Builder) that
                // frequently mishandled the complex object wrapper.
                if (propertiesNode[propName].type === 'array') {
                    writeInterfaceLines.push(`  ${propNameNormalized}${optionalMark}: string[];`);
                }
                else {
                    writeInterfaceLines.push(`  ${propNameNormalized}${optionalMark}: string;`);
                }
            }
            else {
                writeInterfaceLines.push(`  ${propNameNormalized}${optionalMark}: ${referenceEntitiesObjectMapping[propName]};`);
            }
            readInterfaceLines.push(`  ${propNameNormalized}${optionalMark}: ${referenceEntitiesObjectMapping[propName]};`);
            hasReferenceEntities = true;
            continue;
        }
        addPropertyDeclaration(baseInterfaceLines, propName, propertiesNode[propName], requiredFields, primaryKeyObj);
    }
    baseInterfaceLines.push('}');
    baseInterfaceLines.push('');
    lines.push(...baseInterfaceLines);
    if (hasReferenceEntities) {
        readInterfaceLines.push('}');
        readInterfaceLines.push('');
        lines.push(...readInterfaceLines);
    }
    if (hasReferenceEntities) {
        writeInterfaceLines.push('}');
        writeInterfaceLines.push('');
        lines.push(...writeInterfaceLines);
    }
    return { referenceEntitiesObjectMapping, lines, hasReferenceEntities };
}
function addSPReferencedEntityModelInterfaceDeclaration(schema, lines) {
    const referenceEntitiesObjectMapping = {};
    const items = getItemsFromSchema(schema);
    const propertiesNode = getPropertiesNode(items);
    if (isJsonObject(schema) && isJsonObject(schema.referencedEntities)) {
        const referencedEntities = schema.referencedEntities;
        for (const referencedEntityKey in referencedEntities) {
            if (!Object.prototype.hasOwnProperty.call(referencedEntities, referencedEntityKey)) {
                continue;
            }
            referenceEntitiesObjectMapping[referencedEntityKey] = referencedEntityKey + 'Value';
            if (!(referencedEntityKey in propertiesNode)) {
                continue;
            }
            lines.push(`export interface ${referenceEntitiesObjectMapping[referencedEntityKey]} {`);
            // Add actual declaration of the referenced entity
            addInnerPropertyDeclaration(lines, propertiesNode, referencedEntityKey);
            lines.push('}');
            lines.push('');
        }
    }
    return {
        mapping: referenceEntitiesObjectMapping,
        lines,
    };
}
function addInnerPropertyDeclaration(lines, propertiesNode, referencedEntityKey) {
    const propertyValue = propertiesNode[referencedEntityKey];
    if (isJsonObject(propertyValue)) {
        if (isJsonObject(propertyValue.properties)) {
            const properties = propertyValue.properties;
            for (const prop in properties) {
                if (!Object.prototype.hasOwnProperty.call(properties, prop)) {
                    continue;
                }
                const propValue = properties[prop];
                if (isJsonObject(propValue)) {
                    const normalizedProp = convertToValidIdentifier(prop);
                    lines.push(`  ${normalizedProp}: ${mapJsonTypeToTypeScript(propValue.type)};`);
                }
            }
        }
        else if ('items' in propertyValue && isJsonObject(propertyValue.items)) {
            const items = propertyValue.items;
            if ('properties' in items && isJsonObject(items.properties)) {
                const itemProperties = items.properties;
                for (const itemProp in itemProperties) {
                    if (!Object.prototype.hasOwnProperty.call(itemProperties, itemProp)) {
                        continue;
                    }
                    const itemPropValue = itemProperties[itemProp];
                    if (isJsonObject(itemPropValue)) {
                        const normalizedItemProp = convertToValidIdentifier(itemProp);
                        lines.push(`  ${normalizedItemProp}: ${mapJsonTypeToTypeScript(itemPropValue.type)};`);
                    }
                }
            }
        }
    }
}
/*
 * The addModelInterfaceDeclaration method generates the TypeScript interface declaration for the model.
 */
function addModelInterfaceDeclaration(lines, modelName, propertiesNode, requiredFields, primaryKey) {
    lines.push(`export interface ${convertToValidIdentifier(modelName)} {`);
    primaryKey.value = '';
    // propertiesNode is an object with keys as property names.
    for (const propName of Object.keys(propertiesNode)) {
        addPropertyDeclaration(lines, propName, propertiesNode[propName], requiredFields, primaryKey);
    }
    lines.push('}');
    lines.push('');
}
/*
 * The addPropertyDeclaration method generates the TypeScript property declaration for a given JSON property.
 * It checks if the property name is valid, maps the JSON type to TypeScript type,
 * and adds the property to the lines list.
 */
function addPropertyDeclaration(lines, propertyName, propertyValue, requiredFields, primaryKey) {
    let propName = propertyName;
    // Extract type information
    let propType = 'unknown';
    let dataverseType;
    if (isJsonObject(propertyValue)) {
        // Check for Dataverse-specific type first
        if ('x-ms-dataverse-type' in propertyValue &&
            typeof propertyValue['x-ms-dataverse-type'] === 'string') {
            dataverseType = propertyValue['x-ms-dataverse-type'];
            // Map Dataverse type to TypeScript type
            propType = mapDataverseTypeToTypeScript(dataverseType);
        }
        else if ('type' in propertyValue && typeof propertyValue.type === 'string') {
            // Map standard JSON type to TypeScript type
            propType = determineTypeScriptType(propertyName, propertyValue, '', propName);
        }
    }
    const keyType = isJsonObject(propertyValue) && 'x-ms-keyType' in propertyValue
        ? propertyValue['x-ms-keyType']
        : '';
    if (keyType === 'primary') {
        primaryKey.value = propName;
    }
    // Check if the property name is valid
    if (!isValidPropertyName(propName)) {
        propName = `"${propName}"`; // Enclose invalid property names in double quotes
    }
    const isRequired = requiredFields.has(propertyName);
    const optionalMark = isRequired ? '' : '?';
    lines.push(`  ${propName}${optionalMark}: ${propType};`);
}
/*
 * The generateTypeScriptService method generates a TypeScript service class for the model.
 * It includes methods for CRUD operations and uses the Power SDK for data access.
 */
function generateTypeScriptService(primaryKey, isADataverseEntity, context) {
    const { modelName, dataSourceName } = context;
    const lines = [];
    addCopyrightNotice(lines);
    addStandardServiceImports(lines, isADataverseEntity, context, false);
    addServiceClassDefinition(lines, modelName, dataSourceName, primaryKey, isADataverseEntity);
    return lines.join('\n');
}
/*
 * The generateDataverseTypeScriptService method generates a TypeScript service class specifically for Dataverse entities.
 * It handles lookup columns by using base interface for Create/Update and extended interface for Get/GetAll operations.
 */
function generateDataverseTypeScriptService(primaryKey, hasLookupColumns, multiSelectPicklistFields, fileColumnNames, imageColumnNames, context) {
    const { modelName, dataSourceName } = context;
    const lines = [];
    addCopyrightNotice(lines);
    addDataverseServiceImports(lines, hasLookupColumns, multiSelectPicklistFields, fileColumnNames, imageColumnNames, context);
    addDataverseServiceClassDefinition(lines, modelName, dataSourceName, primaryKey, hasLookupColumns, multiSelectPicklistFields, fileColumnNames, imageColumnNames);
    return lines.join('\n');
}
/*
 * The addStandardServiceImports method adds the necessary imports for the service class.
 */
function addStandardServiceImports(lines, isADataverseEntity, context, additionalLine = true) {
    const { modelName } = context;
    const imports = [...collectStandardImports(context)];
    imports.push(`import type { IGetOptions, IGetAllOptions } from '${RELATIVE_MODEL_PATH}/CommonModels';`, `import type { ${modelName} } from '${RELATIVE_MODEL_PATH}/${modelName}Model';`);
    if (isADataverseEntity) {
        imports.push("import type { GetEntityMetadataOptions, EntityMetadata } from '@microsoft/power-apps/data/metadata/dataverse';");
    }
    lines.push(...imports.sort());
    if (additionalLine) {
        lines.push('');
    }
}
/*
 * The addDataverseServiceImports method adds the necessary imports for Dataverse service classes with lookup columns.
 */
function addDataverseServiceImports(lines, hasLookupColumns, multiSelectPicklistFields, fileColumnNames, imageColumnNames, context) {
    const { modelName } = context;
    const id = convertToValidIdentifier(modelName);
    const imports = [...collectStandardImports(context)];
    if (multiSelectPicklistFields.length > 0) {
        imports.push("import { deserializeMultiSelectPicklistFields, serializeMultiSelectPicklistFields } from '@microsoft/power-apps/data';");
    }
    imports.push(`import type { IGetOptions, IGetAllOptions } from '${RELATIVE_MODEL_PATH}/CommonModels';`);
    // Build a single combined import from the model file for all required types
    const modelTypeImports = hasLookupColumns
        ? [`${modelName}Base`, modelName]
        : [modelName];
    if (fileColumnNames.length > 0) {
        modelTypeImports.push(`${id}FileColumnName`);
    }
    if (imageColumnNames.length > 0) {
        modelTypeImports.push(`${id}ImageColumnName`);
    }
    if (fileColumnNames.length > 0 || imageColumnNames.length > 0) {
        modelTypeImports.push(`${id}UploadColumnName`);
    }
    imports.push(`import type { ${modelTypeImports.join(', ')} } from '${RELATIVE_MODEL_PATH}/${modelName}Model';`);
    imports.push("import type { GetEntityMetadataOptions, EntityMetadata } from '@microsoft/power-apps/data/metadata/dataverse';");
    lines.push(...imports.sort());
    lines.push('');
}
function addSharepointServiceImports(lines, hasReferenceEntities, context) {
    const { modelName } = context;
    const imports = [...collectStandardImports(context)];
    imports.push(`import type { IGetOptions, IGetAllOptions } from '${RELATIVE_MODEL_PATH}/CommonModels';`);
    if (hasReferenceEntities) {
        // Import both base and read/write interfaces
        imports.push(`import type { ${modelName}Read, ${modelName}Write } from '${RELATIVE_MODEL_PATH}/${modelName}Model';`);
    }
    else {
        // Import single interface
        imports.push(`import type { ${modelName}Base } from '${RELATIVE_MODEL_PATH}/${modelName}Model';`);
    }
    lines.push(...imports.sort());
    lines.push('');
}
/*
 * The addServiceFields method adds fields to the service class.
 * It includes the Power SDK instance and a static instance of the service.
 */
function addServiceFields(lines, dataSourceName) {
    lines.push(`  private static readonly dataSourceName = '${dataSourceName.toLowerCase()}';`);
    lines.push(``);
    lines.push(`  private static readonly client = getClient(dataSourcesInfo);`);
}
/*
 * The AddServiceClassDefinition method generates the service class definition.
 * It takes the lines to write to, model name, data source name, schema, and response model name as parameters.
 */
function addServiceClassDefinition(lines, modelName, dataSourceName, primaryKey, isADataverseEntity) {
    addServiceClassHeader(lines, modelName, dataSourceName, true);
    addServiceMethods(lines, modelName, dataSourceName, primaryKey, isADataverseEntity);
    lines.push('}');
    lines.push('');
}
/*
 * The addDataverseServiceClassDefinition method generates the service class definition for Dataverse entities with lookup columns.
 */
function addDataverseServiceClassDefinition(lines, modelName, dataSourceName, primaryKey, hasLookupColumns, multiSelectPicklistFields, fileColumnNames, imageColumnNames) {
    addServiceClassHeader(lines, modelName, dataSourceName, true);
    if (multiSelectPicklistFields.length > 0) {
        const fieldList = multiSelectPicklistFields.map((f) => `'${f}'`).join(', ');
        lines.push(`  private static readonly multiSelectPicklistFields = [${fieldList}] as const;`);
    }
    addDataverseServiceMethods(lines, modelName, dataSourceName, primaryKey, hasLookupColumns, multiSelectPicklistFields, fileColumnNames, imageColumnNames);
    lines.push('}');
    lines.push('');
}
/*
 * The addServiceMethods method generates the CRUD methods for the service class.
 * It includes create, update, delete, get, and getAll methods.
 */
function addServiceMethods(lines, modelName, dataSourceName, primaryKey, isADataverseEntity) {
    const methods = [];
    methods.push(generateCreateMethod(modelName, primaryKey));
    methods.push(generateUpdateMethod(modelName, primaryKey));
    methods.push(generateDeleteMethod(modelName));
    methods.push(generateGetMethod(modelName));
    methods.push(generateGetAllMethod(modelName));
    if (isADataverseEntity) {
        methods.push(generateGetMetadataMethod(modelName));
    }
    for (const method of methods) {
        lines.push('');
        lines.push(...method);
    }
}
/*
 * The addDataverseServiceMethods method generates the CRUD methods for Dataverse service classes with lookup columns.
 * Uses base interface for Create/Update and extended interface for Get/GetAll operations.
 */
function addDataverseServiceMethods(lines, modelName, dataSourceName, primaryKey, hasLookupColumns, multiSelectPicklistFields, fileColumnNames, imageColumnNames) {
    const hasFileColumns = fileColumnNames.length > 0;
    const hasImageColumns = imageColumnNames.length > 0;
    const hasFileOrImageColumns = hasFileColumns || hasImageColumns;
    const methods = [];
    methods.push(generateDataverseCreateMethod(modelName, primaryKey, hasLookupColumns, multiSelectPicklistFields));
    methods.push(generateDataverseUpdateMethod(modelName, primaryKey, hasLookupColumns, multiSelectPicklistFields));
    methods.push(generateDeleteMethod(modelName));
    methods.push(generateDataverseGetMethod(modelName, multiSelectPicklistFields));
    methods.push(generateDataverseGetAllMethod(modelName, multiSelectPicklistFields));
    methods.push(generateDataverseGetMetadataMethod(modelName));
    if (hasFileOrImageColumns) {
        methods.push(generateDataverseUploadMethod(modelName));
    }
    if (hasFileColumns) {
        methods.push(generateDataverseDownloadFileMethod(modelName));
    }
    if (hasImageColumns) {
        methods.push(generateDataverseDownloadImageMethod(modelName));
    }
    if (hasFileOrImageColumns) {
        methods.push(generateDataverseDeleteFileOrImageMethod(modelName));
    }
    for (const method of methods) {
        lines.push('');
        lines.push(...method);
    }
}
/*
 * The create method adds a new record to the data source.
 */
function generateCreateMethod(modelName, primaryKey) {
    const datatypeForCreate = primaryKey !== '' ? `Omit<${modelName}, '${primaryKey}'>` : modelName;
    return [
        `  public static async create(record: ${datatypeForCreate}): Promise<IOperationResult<${modelName}>> {`,
        `    const result = await ${modelName}Service.client.createRecordAsync<${datatypeForCreate}, ${modelName}>(`,
        `      ${getServiceName(modelName)},`,
        `      record`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
/*
 * The update method modifies an existing record.
 */
function generateUpdateMethod(modelName, primaryKey) {
    const datatypeForUpdate = primaryKey !== '' ? `Partial<Omit<${modelName}, '${primaryKey}'>>` : modelName;
    return [
        `  public static async update(id: string, changedFields: ${datatypeForUpdate}): Promise<IOperationResult<${modelName}>> {`,
        `    const result = await ${modelName}Service.client.updateRecordAsync<${datatypeForUpdate}, ${modelName}>(`,
        `      ${getServiceName(modelName)},`,
        `      id,`,
        `      changedFields`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
/*
 * The delete method removes a record by its ID.
 */
function generateDeleteMethod(modelName) {
    return [
        `  public static async delete(id: string): Promise<void> {`,
        `    await ${modelName}Service.client.deleteRecordAsync(`,
        `      ${getServiceName(modelName)},`,
        `      id);`,
        `  }`,
    ];
}
/*
 * The get method retrieves a record by its ID.
 */
function generateGetMethod(modelName) {
    return [
        `  public static async get(id: string, options?: IGetOptions): Promise<IOperationResult<${modelName}>> {`,
        `    const result = await ${modelName}Service.client.retrieveRecordAsync<${modelName}>(`,
        `      ${getServiceName(modelName)},`,
        `      id,`,
        `      options`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
/*
 * The getAll method retrieves all records of the specified model.
 */
function generateGetAllMethod(modelName) {
    return [
        `  public static async getAll(options?: IGetAllOptions): Promise<IOperationResult<${modelName}[]>> {`,
        `    const result = await ${modelName}Service.client.retrieveMultipleRecordsAsync<${modelName}>(`,
        `      ${getServiceName(modelName)},`,
        `      options`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
function generateGetMetadataMethod(modelName) {
    return [
        `  public static getMetadata(`,
        `    options: GetEntityMetadataOptions<${modelName}> = {}`,
        `  ): Promise<IOperationResult<Partial<EntityMetadata>>> {`,
        `    return ${modelName}Service.client.executeAsync({`,
        `      dataverseRequest: {`,
        `        action: "getEntityMetadata",`,
        `        parameters: {`,
        `          tableName: ${getServiceName(modelName)},`,
        `          options: options as GetEntityMetadataOptions,`,
        `        },`,
        `      },`,
        `    });`,
        `  }`,
    ];
}
/*
 * The create method adds a new record to the data source.
 */
function generateSharepointCreateMethod(modelName, primaryKey, hasReferenceEntities) {
    /**
     * if hasReferenceEntities is true take in modelNameWrite-primaryKey as input and return modelNameRead
     * else take in modelNameBase-primaryKey as input and return modelNameBase
     */
    const datatypeForCreate = primaryKey !== ''
        ? `Omit<${hasReferenceEntities ? `${modelName}Write` : `${modelName}Base`}, '${primaryKey}'>`
        : hasReferenceEntities
            ? `${modelName}Write`
            : `${modelName}Base`;
    const returnType = hasReferenceEntities ? `${modelName}Read` : `${modelName}Base`;
    return [
        `  public static async create(record: ${datatypeForCreate}): Promise<IOperationResult<${returnType}>> {`,
        `    const result = await ${modelName}Service.client.createRecordAsync<${datatypeForCreate}, ${returnType}>(`,
        `      ${getServiceName(modelName)},`,
        `      record`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
/*
 * The update method modifies an existing record.
 */
function generateSharepointUpdateMethod(modelName, primaryKey, hasReferenceEntities) {
    /**
     * if hasReferenceEntities is true take in modelNameWrite-primaryKey as input and return modelNameRead
     * else take in modelNameBase-primaryKey as input and return modelNameBase
     */
    const datatypeForUpdate = primaryKey !== ''
        ? `Partial<Omit<${hasReferenceEntities ? `${modelName}Write` : `${modelName}Base`}, '${primaryKey}'>>`
        : hasReferenceEntities
            ? `${modelName}Write`
            : `${modelName}Base`;
    const returnType = hasReferenceEntities ? `${modelName}Read` : `${modelName}Base`;
    return [
        `  public static async update(id: string, changedFields: ${datatypeForUpdate}): Promise<IOperationResult<${returnType}>> {`,
        `    const result = await ${modelName}Service.client.updateRecordAsync<${datatypeForUpdate}, ${returnType}>(`,
        `      ${getServiceName(modelName)},`,
        `      id,`,
        `      changedFields`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
/*
 * The delete method removes a record by its ID.
 */
function generateSharepointDeleteMethod(modelName) {
    return [
        `  public static async delete(id: string): Promise<void> {`,
        `    await ${modelName}Service.client.deleteRecordAsync(`,
        `      ${getServiceName(modelName)},`,
        `      id);`,
        `  }`,
    ];
}
/*
 * The get method retrieves a record by its ID.
 */
function generateSharepointGetMethod(modelName, hasReferenceEntities) {
    /**
     * if referenceEntities is true return modelNameRead
     * else return modelNameBase
     */
    const returnType = hasReferenceEntities ? `${modelName}Read` : `${modelName}Base`;
    /**
     */
    return [
        `  public static async get(id: string, options?: IGetOptions): Promise<IOperationResult<${returnType}>> {`,
        `    const result = await ${modelName}Service.client.retrieveRecordAsync<${returnType}>(`,
        `      ${getServiceName(modelName)},`,
        `      id,`,
        `      options`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
/*
 * The getAll method retrieves all records of the specified model.
 */
function generateSharepointGetAllMethod(modelName, hasReferenceEntities) {
    const returnType = hasReferenceEntities ? `${modelName}Read` : `${modelName}Base`;
    return [
        `  public static async getAll(options?: IGetAllOptions): Promise<IOperationResult<${returnType}[]>> {`,
        `    const result = await ${modelName}Service.client.retrieveMultipleRecordsAsync<${returnType}>(`,
        `      ${getServiceName(modelName)},`,
        `      options`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
/*
 * The generateDataverseCreateMethod method adds a new record using the base interface for Dataverse entities.
 */
function generateDataverseCreateMethod(modelName, primaryKey, hasLookupColumns, multiSelectPicklistFields) {
    const baseInterfaceName = hasLookupColumns ? `${modelName}Base` : modelName;
    const extendedInterfaceName = modelName;
    const datatypeForCreate = primaryKey !== '' ? `Omit<${baseInterfaceName}, '${primaryKey}'>` : baseInterfaceName;
    const hasMultiSelect = multiSelectPicklistFields.length > 0;
    return hasMultiSelect
        ? [
            `  public static async create(record: ${datatypeForCreate}): Promise<IOperationResult<${extendedInterfaceName}>> {`,
            `    const result = await ${modelName}Service.client.createRecordAsync<Record<string, unknown>, ${extendedInterfaceName}>(`,
            `      ${getServiceName(modelName)},`,
            `      serializeMultiSelectPicklistFields(record as Record<string, unknown>, ${modelName}Service.multiSelectPicklistFields)`,
            `    );`,
            `    if (result.data) {`,
            `      deserializeMultiSelectPicklistFields(result.data as Record<string, unknown>, ${modelName}Service.multiSelectPicklistFields);`,
            `    }`,
            `    return result;`,
            `  }`,
        ]
        : [
            `  public static async create(record: ${datatypeForCreate}): Promise<IOperationResult<${extendedInterfaceName}>> {`,
            `    const result = await ${modelName}Service.client.createRecordAsync<${datatypeForCreate}, ${extendedInterfaceName}>(`,
            `      ${getServiceName(modelName)},`,
            `      record`,
            `    );`,
            `    return result;`,
            `  }`,
        ];
}
/*
 * The generateDataverseUpdateMethod method modifies an existing record using the base interface for Dataverse entities.
 */
function generateDataverseUpdateMethod(modelName, primaryKey, hasLookupColumns, multiSelectPicklistFields) {
    const baseInterfaceName = hasLookupColumns ? `${modelName}Base` : modelName;
    const extendedInterfaceName = modelName;
    const datatypeForUpdate = primaryKey !== '' ? `Partial<Omit<${baseInterfaceName}, '${primaryKey}'>>` : baseInterfaceName;
    const hasMultiSelect = multiSelectPicklistFields.length > 0;
    return hasMultiSelect
        ? [
            `  public static async update(id: string, changedFields: ${datatypeForUpdate}): Promise<IOperationResult<${extendedInterfaceName}>> {`,
            `    const result = await ${modelName}Service.client.updateRecordAsync<Record<string, unknown>, ${extendedInterfaceName}>(`,
            `      ${getServiceName(modelName)},`,
            `      id,`,
            `      serializeMultiSelectPicklistFields(changedFields as Record<string, unknown>, ${modelName}Service.multiSelectPicklistFields)`,
            `    );`,
            `    if (result.data) {`,
            `      deserializeMultiSelectPicklistFields(result.data as Record<string, unknown>, ${modelName}Service.multiSelectPicklistFields);`,
            `    }`,
            `    return result;`,
            `  }`,
        ]
        : [
            `  public static async update(id: string, changedFields: ${datatypeForUpdate}): Promise<IOperationResult<${extendedInterfaceName}>> {`,
            `    const result = await ${modelName}Service.client.updateRecordAsync<${datatypeForUpdate}, ${extendedInterfaceName}>(`,
            `      ${getServiceName(modelName)},`,
            `      id,`,
            `      changedFields`,
            `    );`,
            `    return result;`,
            `  }`,
        ];
}
/*
 * The generateDataverseGetMethod method retrieves a record by its ID using the extended interface for Dataverse entities.
 */
function generateDataverseGetMethod(modelName, multiSelectPicklistFields) {
    const extendedInterfaceName = modelName;
    const hasMultiSelect = multiSelectPicklistFields.length > 0;
    return hasMultiSelect
        ? [
            `  public static async get(id: string, options?: IGetOptions): Promise<IOperationResult<${extendedInterfaceName}>> {`,
            `    const result = await ${modelName}Service.client.retrieveRecordAsync<${extendedInterfaceName}>(`,
            `      ${getServiceName(modelName)},`,
            `      id,`,
            `      options`,
            `    );`,
            `    if (result.data) {`,
            `      deserializeMultiSelectPicklistFields(result.data as Record<string, unknown>, ${modelName}Service.multiSelectPicklistFields);`,
            `    }`,
            `    return result;`,
            `  }`,
        ]
        : [
            `  public static async get(id: string, options?: IGetOptions): Promise<IOperationResult<${extendedInterfaceName}>> {`,
            `    const result = await ${modelName}Service.client.retrieveRecordAsync<${extendedInterfaceName}>(`,
            `      ${getServiceName(modelName)},`,
            `      id,`,
            `      options`,
            `    );`,
            `    return result;`,
            `  }`,
        ];
}
/*
 * The generateDataverseGetAllMethod method retrieves all records using the extended interface for Dataverse entities.
 */
function generateDataverseGetAllMethod(modelName, multiSelectPicklistFields) {
    const extendedInterfaceName = modelName;
    const hasMultiSelect = multiSelectPicklistFields.length > 0;
    return hasMultiSelect
        ? [
            `  public static async getAll(options?: IGetAllOptions): Promise<IOperationResult<${extendedInterfaceName}[]>> {`,
            `    const result = await ${modelName}Service.client.retrieveMultipleRecordsAsync<${extendedInterfaceName}>(`,
            `      ${getServiceName(modelName)},`,
            `      options`,
            `    );`,
            `    result.data?.forEach(record => deserializeMultiSelectPicklistFields(record as Record<string, unknown>, ${modelName}Service.multiSelectPicklistFields));`,
            `    return result;`,
            `  }`,
        ]
        : [
            `  public static async getAll(options?: IGetAllOptions): Promise<IOperationResult<${extendedInterfaceName}[]>> {`,
            `    const result = await ${modelName}Service.client.retrieveMultipleRecordsAsync<${extendedInterfaceName}>(`,
            `      ${getServiceName(modelName)},`,
            `      options`,
            `    );`,
            `    return result;`,
            `  }`,
        ];
}
/*
 * The generateDataverseGetMetadataMethod method gets metadata using the extended interface for Dataverse entities.
 */
function generateDataverseGetMetadataMethod(modelName) {
    const extendedInterfaceName = modelName;
    return [
        `  public static getMetadata(`,
        `    options: GetEntityMetadataOptions<${extendedInterfaceName}> = {}`,
        `  ): Promise<IOperationResult<Partial<EntityMetadata>>> {`,
        `    return ${modelName}Service.client.executeAsync({`,
        `      dataverseRequest: {`,
        `        action: "getEntityMetadata",`,
        `        parameters: {`,
        `          tableName: ${getServiceName(modelName)},`,
        `          options: options as GetEntityMetadataOptions,`,
        `        },`,
        `      },`,
        `    });`,
        `  }`,
    ];
}
/*
 * The upload method uploads a file to the dataverse entity.
 *
 * Parameters:
 *   id - the primary key of the record to which the file is being uploaded
 *   columnName - the name of the column to which the file is being uploaded
 *   file - the file to be uploaded
 *   fileDisplayName - optional display name for the file
 *
 * Note: max possible size: 128 MB for files if configured in Dataverse (see MaxSizeInKb property in Dataverse Settings),
 *       max possible size for images is 30 MB.
 */
function generateDataverseUploadMethod(modelName) {
    return [
        `  public static async upload(id: string, columnName: ${convertToValidIdentifier(modelName)}UploadColumnName, file: File, fileDisplayName?: string): Promise<IOperationResult<void>> {`,
        `    const arrayBuffer = await file.arrayBuffer();`,
        `    const data = new Uint8Array(arrayBuffer);`,
        `    const result = await ${modelName}Service.client.uploadFileToRecord(`,
        `      ${getServiceName(modelName)},`,
        `      id,`,
        `      columnName,`,
        `      fileDisplayName || file.name,`,
        `      data,`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
/*
 * The download method downloads a file from the dataverse entity.
 *
 * Parameters:
 *   id - the primary key of the record from which the file is being downloaded
 *   columnName - the name of the column from which the file is being downloaded
 *
 * Note: max possible size: 128 MB for files if configured in Dataverse (see MaxSizeInKb property in Dataverse Settings),
 *       max possible size for images is 30 MB.
 */
function generateDataverseDownloadFileMethod(modelName) {
    return [
        `  public static async downloadFile(id: string, columnName: ${convertToValidIdentifier(modelName)}FileColumnName): Promise<IOperationResult<Uint8Array>> {`,
        `    const result = await ${modelName}Service.client.downloadFileFromRecord(`,
        `      ${getServiceName(modelName)},`,
        `      id,`,
        `      columnName,`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
/*
 * The downloadImage method downloads an image from the Dataverse entity via the Web API $value endpoint.
 *
 * Parameters:
 *   id - the primary key of the record from which the image is being downloaded
 *   columnName - the name of the image column from which the image is being downloaded
 *   fullSize - when true, requests the full-size image (?size=full); defaults to false (thumbnail/smaller size)
 */
function generateDataverseDownloadImageMethod(modelName) {
    return [
        `  public static async downloadImage(id: string, columnName: ${convertToValidIdentifier(modelName)}ImageColumnName, fullSize: boolean = false): Promise<IOperationResult<Uint8Array>> {`,
        `    const result = await ${modelName}Service.client.downloadImageFromRecord(`,
        `      ${getServiceName(modelName)},`,
        `      id,`,
        `      columnName,`,
        `      fullSize,`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
function generateDataverseDeleteFileOrImageMethod(modelName) {
    return [
        `  public static async deleteFileOrImage(id: string, columnName: ${convertToValidIdentifier(modelName)}UploadColumnName): Promise<IOperationResult<void>> {`,
        `    const result = await ${modelName}Service.client.deleteFileOrImageFromRecord(`,
        `      ${getServiceName(modelName)},`,
        `      id,`,
        `      columnName,`,
        `    );`,
        `    return result;`,
        `  }`,
    ];
}
/**
 * The generateIndexFile method creates an index.ts file that exports all models and services.
 */
async function generateIndexFile(outputDir, project) {
    const indexPath = project.join(outputDir, GENERATED_FOLDER, 'index.ts');
    const lines = [];
    // Add copyright notice
    addCopyrightNotice(lines);
    // Export all models from generated/models
    const modelsDir = project.join(outputDir, MODELS_FOLDER);
    if (await project.directoryExists(modelsDir)) {
        const modelFiles = await project.getSourceFilesInDirectory(modelsDir);
        if (modelFiles.length > 0) {
            lines.push('// Models');
            for (const file of modelFiles.sort()) {
                const fileName = file.replace('.ts', '');
                const alias = kebabToCamelCase(fileName);
                lines.push(`export * as ${alias} from './models/${fileName}';`);
            }
            lines.push('');
        }
    }
    // Export all services from generated/services
    const servicesDir = project.join(outputDir, SERVICES_FOLDER);
    if (await project.directoryExists(servicesDir)) {
        const serviceFiles = await project.getSourceFilesInDirectory(servicesDir);
        if (serviceFiles.length > 0) {
            lines.push('// Services');
            for (const file of serviceFiles.sort()) {
                const fileName = file.replace('.ts', '');
                lines.push(`export * from './services/${fileName}';`);
            }
            lines.push('');
        }
    }
    // Write the index file
    await project.createFile('index', indexPath, lines.join('\n'));
}
/**
 * Check if the propertyNode is a sharepoint choice or multi select choice
 *
 * @param {JsonValue} propNode
 * @returns {boolean}
 */
function isSharepointChoiceColumn(propNode) {
    try {
        if (isJsonObject(propNode) &&
            isJsonObject(propNode['x-ms-capabilities']) &&
            isJsonObject(propNode['x-ms-capabilities']['x-ms-sp']) &&
            propNode['x-ms-capabilities']['x-ms-sp'].IsChoice === true) {
            return true;
        }
    }
    catch {
        // Ignore errors and return false so we fall back to default behavior
    }
    return false;
}
