// @microsoft/power-apps-common/dist/plugins/PluginConstants.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { GetAppConnectionsKey } from './definitions/GetAppConnectionsPlugin.js';
import { GetCdsKey } from './definitions/GetCdsPlugin.js';
import { GetContextKey } from './definitions/GetContextPlugin.js';
import { HttpSendKey } from './definitions/HttpSendPlugin.js';
import { MessagingGetVersionInfoKey, MessagingSendKey } from './definitions/MessagingPlugin.js';
import { NotifyAppSdkLoadedKey } from './definitions/NotifyAppSdkLoadedPlugin.js';
import { TelemetryTrackEventKey, TelemetryTrackExceptionKey } from './definitions/TelemetryPlugin.js';
export const PluginAction = {
    GetAppConnections: GetAppConnectionsKey,
    GetCds: GetCdsKey,
    GetContext: GetContextKey,
    HttpSend: HttpSendKey,
    MessagingGetVersionInfo: MessagingGetVersionInfoKey,
    MessagingSend: MessagingSendKey,
    NotifyAppSdkLoaded: NotifyAppSdkLoadedKey,
    TelemetryTrackEvent: TelemetryTrackEventKey,
    TelemetryTrackException: TelemetryTrackExceptionKey,
};
